// oidc-spa/esm/tanstack-start/react/utilsBuilder.mjs
import { assert } from "../../tools/tsafe/assert.mjs";
import { createCreateValidateAndGetAccessTokenClaims_rfc9068 } from "./accessTokenValidation_rfc9068.mjs";
import { createOidcSpaUtils } from "./createOidcSpaUtils.mjs";
function createOidcSpaUtilsBuilder(params) {
    return {
        withAutoLogin: () => createOidcSpaUtilsBuilder({
            autoLogin: true,
            decodedIdTokenSchema: params.decodedIdTokenSchema,
            decodedIdToken_mock: params.decodedIdToken_mock,
            createValidateAndGetAccessTokenClaims: params.createValidateAndGetAccessTokenClaims
        }),
        withExpectedDecodedIdTokenShape: ({ decodedIdTokenSchema, decodedIdToken_mock }) => createOidcSpaUtilsBuilder({
            autoLogin: params.autoLogin,
            decodedIdTokenSchema,
            decodedIdToken_mock: decodedIdToken_mock,
            createValidateAndGetAccessTokenClaims: params.createValidateAndGetAccessTokenClaims
        }),
        withAccessTokenValidation: params_scope => createOidcSpaUtilsBuilder({
            autoLogin: params.autoLogin,
            decodedIdTokenSchema: params.decodedIdTokenSchema,
            decodedIdToken_mock: params.decodedIdToken_mock,
            createValidateAndGetAccessTokenClaims: (() => {
                switch (params_scope.type) {
                    case "RFC 9068: JSON Web Token (JWT) Profile for OAuth 2.0 Access Tokens": {
                        const { accessTokenClaimsSchema, accessTokenClaims_mock, expectedAudience } = params_scope;
                        const { createValidateAndGetAccessTokenClaims } = createCreateValidateAndGetAccessTokenClaims_rfc9068({
                            // @ts-expect-error
                            accessTokenClaims_mock,
                            // @ts-expect-error
                            accessTokenClaimsSchema,
                            expectedAudience
                        });
                        return createValidateAndGetAccessTokenClaims;
                    }
                    case "custom": {
                        const { createValidateAndGetAccessTokenClaims } = params_scope;
                        return createValidateAndGetAccessTokenClaims;
                    }
                    default:
                        assert(false);
                }
            })()
        }),
        createUtils: () => createOidcSpaUtils({
            autoLogin: params.autoLogin,
            decodedIdTokenSchema: params.decodedIdTokenSchema,
            decodedIdToken_mock: params.decodedIdToken_mock,
            createValidateAndGetAccessTokenClaims: params.createValidateAndGetAccessTokenClaims
        })
    };
}
export const oidcSpaUtilsBuilder = createOidcSpaUtilsBuilder({
    autoLogin: false,
    createValidateAndGetAccessTokenClaims: undefined,
    decodedIdToken_mock: undefined,
    decodedIdTokenSchema: undefined
});
