// oidc-spa/esm/angular.mjs
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f)
        throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
        throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m")
        throw new TypeError("Private method is not writable");
    if (kind === "a" && !f)
        throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
        throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var _AbstractOidcService_instances, _AbstractOidcService_warnUserSecondsBeforeAutoLogout, _AbstractOidcService_isRunningGetParams, _AbstractOidcService_deadlockDetectionTimer, _AbstractOidcService_dState, _AbstractOidcService_initialize, _AbstractOidcService_getPrInitializedNotResolvedErrorMessage, _AbstractOidcService_getState, _AbstractOidcService_getAutoLoginAndInitializationErrorAccessErrorMessage, _AbstractOidcService_getOidc, _AbstractOidcService_isUserLoggedIn_override, _AbstractOidcService_decodedIdToken$, _AbstractOidcService_$decodedIdToken, _AbstractOidcService_isNewBrowserSession_override;
import { isPlatformBrowser } from "@angular/common";
import { BehaviorSubject, from, switchMap, Subject } from "rxjs";
import { Deferred } from "./tools/Deferred.mjs";
import { assert, is } from "./tools/tsafe/assert.mjs";
import { uncapitalize } from "./tools/tsafe/uncapitalize.mjs";
import { createObjectThatThrowsIfAccessed, AccessError } from "./tools/createObjectThatThrowsIfAccessed.mjs";
import { inject, makeEnvironmentProviders, provideAppInitializer, PLATFORM_ID } from "@angular/core";
import { toSignal } from "@angular/core/rxjs-interop";
import { Router } from "@angular/router";
import { getBaseHref } from "./tools/getBaseHref.mjs";
import { setDesiredPostLoginRedirectUrl } from "./core/desiredPostLoginRedirectUrl.mjs";
assert;
export class AbstractOidcService {
    constructor() {
        _AbstractOidcService_instances.add(this);
        this.autoLogin = false;
        this.providerAwaitsInitialization = true;
        this.decodedIdTokenSchema = undefined;
        this.mockDecodedIdToken = undefined;
        _AbstractOidcService_warnUserSecondsBeforeAutoLogout.set(this, 60);
        _AbstractOidcService_isRunningGetParams.set(this, false);
        this.allowDecodedIdTokenAccessInShouldInjectAccessToken = false;
        _AbstractOidcService_deadlockDetectionTimer.set(this, undefined);
        _AbstractOidcService_dState.set(this, new Deferred());
        this.prInitialized = __classPrivateFieldGet(this, _AbstractOidcService_dState, "f").pr.then(() => true);
        _AbstractOidcService_isUserLoggedIn_override.set(this, undefined);
        this.decodedIdTokenAccess = undefined;
        _AbstractOidcService_decodedIdToken$.set(this, (() => {
            const decodedIdToken$ = new BehaviorSubject(createObjectThatThrowsIfAccessed({
                debugMessage: __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getPrInitializedNotResolvedErrorMessage).call(this, {
                    callerName: "decodedIdToken"
                })
            }));
            (async () => {
                const { initializationError, oidc } = await __classPrivateFieldGet(this, _AbstractOidcService_dState, "f").pr;
                if (initializationError !== undefined) {
                    decodedIdToken$.next(createObjectThatThrowsIfAccessed({
                        debugMessage: __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getAutoLoginAndInitializationErrorAccessErrorMessage).call(this, {
                            callerName: "decodedIdToken"
                        })
                    }));
                    return;
                }
                assert(oidc !== undefined);
                if (!oidc.isUserLoggedIn) {
                    decodedIdToken$.next(createObjectThatThrowsIfAccessed({
                        debugMessage: [
                            `oidc-spa: Trying to read properties of decodedIdToken, the user`,
                            `isn't currently logged in, this does not make sense.`,
                            `You are responsible for controlling the flow of your app and`,
                            `not try to read the decodedIdToken when oidc.isUserLoggedIn is false.`
                        ].join(" ")
                    }));
                    return;
                }
                decodedIdToken$.next(oidc.getDecodedIdToken());
                oidc.subscribeToTokensChange(() => {
                    const value_new = oidc.getDecodedIdToken();
                    const value_current = decodedIdToken$.getValue();
                    if (value_new === value_current) {
                        return;
                    }
                    decodedIdToken$.next(value_new);
                });
            })();
            return decodedIdToken$;
        })());
        _AbstractOidcService_$decodedIdToken.set(this, toSignal(this.decodedIdToken$, { requireSync: true }));
        this.accessTokenRotation$ = (() => {
            const accessTokenRotation$ = new Subject();
            (async () => {
                const { oidc } = await __classPrivateFieldGet(this, _AbstractOidcService_dState, "f").pr;
                if (oidc === undefined) {
                    return;
                }
                if (!oidc.isUserLoggedIn) {
                    return;
                }
                oidc.subscribeToTokensChange(({ accessToken }) => {
                    accessTokenRotation$.next(accessToken);
                });
            })();
            return accessTokenRotation$;
        })();
        this.$secondsLeftBeforeAutoLogout = (() => {
            const secondsLeftBeforeAutoLogout$ = new BehaviorSubject(null);
            (async () => {
                const { oidc } = await __classPrivateFieldGet(this, _AbstractOidcService_dState, "f").pr;
                if (oidc === undefined) {
                    return;
                }
                if (!oidc.isUserLoggedIn) {
                    return;
                }
                oidc.subscribeToAutoLogoutCountdown(({ secondsLeft }) => {
                    if (secondsLeft === undefined || secondsLeft > __classPrivateFieldGet(this, _AbstractOidcService_warnUserSecondsBeforeAutoLogout, "f")) {
                        if (secondsLeftBeforeAutoLogout$.getValue() !== null) {
                            secondsLeftBeforeAutoLogout$.next(null);
                        }
                        return;
                    }
                    secondsLeftBeforeAutoLogout$.next(secondsLeft);
                });
            })();
            return toSignal(secondsLeftBeforeAutoLogout$, { requireSync: true });
        })();
        _AbstractOidcService_isNewBrowserSession_override.set(this, undefined);
    }
    static provide(params) {
        const paramsOrGetParams = params;
        assert(is(this));
        return makeEnvironmentProviders([
            this,
            provideAppInitializer(async () => {
                {
                    const platformId = inject(PLATFORM_ID);
                    if (!isPlatformBrowser(platformId)) {
                        return;
                    }
                }
                const instance = inject(this);
                __classPrivateFieldGet(instance, _AbstractOidcService_instances, "m", _AbstractOidcService_initialize).call(instance, {
                    prOidcOrInitializationError: (async () => {
                        const [{ createOidc }, { warnUserSecondsBeforeAutoLogout, ...params }] = await Promise.all([
                            import("./core/index.mjs"),
                            typeof paramsOrGetParams === "function"
                                ? (async () => {
                                    __classPrivateFieldSet(instance, _AbstractOidcService_isRunningGetParams, true, "f");
                                    const params = await paramsOrGetParams();
                                    __classPrivateFieldSet(instance, _AbstractOidcService_isRunningGetParams, false, "f");
                                    return params;
                                })()
                                : paramsOrGetParams
                        ]);
                        if (warnUserSecondsBeforeAutoLogout !== undefined) {
                            __classPrivateFieldSet(instance, _AbstractOidcService_warnUserSecondsBeforeAutoLogout, warnUserSecondsBeforeAutoLogout, "f");
                        }
                        try {
                            return createOidc({
                                BASE_URL: getBaseHref(),
                                autoLogin: instance.autoLogin,
                                decodedIdTokenSchema: instance.decodedIdTokenSchema,
                                ...params
                            });
                        }
                        catch (initializationError) {
                            assert(initializationError instanceof Error);
                            assert(is(initializationError));
                            return initializationError;
                        }
                    })()
                });
                if (instance.providerAwaitsInitialization) {
                    await instance.prInitialized;
                }
            })
        ]);
    }
    static provideMock(params = {}) {
        assert(is(this));
        return makeEnvironmentProviders([
            this,
            provideAppInitializer(async () => {
                {
                    const platformId = inject(PLATFORM_ID);
                    if (!isPlatformBrowser(platformId)) {
                        return;
                    }
                }
                const instance = inject(this);
                __classPrivateFieldGet(instance, _AbstractOidcService_instances, "m", _AbstractOidcService_initialize).call(instance, {
                    prOidcOrInitializationError: (async () => {
                        const { createMockOidc } = await import("./core/createMockOidc.mjs");
                        return createMockOidc({
                            BASE_URL: getBaseHref(),
                            autoLogin: instance.autoLogin,
                            isUserInitiallyLoggedIn: instance.autoLogin
                                ? true
                                : params.isUserInitiallyLoggedIn,
                            mockedParams: {
                                issuerUri: params.mockIssuerUri,
                                clientId: params.mockClientId
                            },
                            mockedTokens: {
                                accessToken: params.mockAccessToken,
                                decodedIdToken: await (() => {
                                    if (instance.mockDecodedIdToken === undefined) {
                                        return undefined;
                                    }
                                    if (typeof instance.mockDecodedIdToken === "function") {
                                        return instance.mockDecodedIdToken();
                                    }
                                })()
                            }
                        });
                    })()
                });
                await instance.prInitialized;
            })
        ]);
    }
    static createBearerInterceptor(params) {
        const { shouldInjectAccessToken: getShouldInjectAccessToken } = params;
        return (req, next) => {
            const instance = inject(this);
            let shouldInjectAccessToken = undefined;
            const shouldInjectAccessTokenByState = [];
            let isDecodedIdTokenAccessed = false;
            (instance.decodedIdTokenAccess = new Subject()).subscribe(() => (isDecodedIdTokenAccessed = true));
            try {
                shouldInjectAccessToken = getShouldInjectAccessToken(req);
            }
            catch (error) {
                if (!(error instanceof OidcAccessedTooEarlyError) && !(error instanceof AccessError)) {
                    throw error;
                }
            }
            if (shouldInjectAccessToken === undefined) {
                for (const isUserLoggedIn of [false, true]) {
                    for (const isNewBrowserSession of isUserLoggedIn ? [false, true] : [undefined]) {
                        let shouldInjectAccessTokenOrError;
                        __classPrivateFieldSet(instance, _AbstractOidcService_isUserLoggedIn_override, isUserLoggedIn, "f");
                        __classPrivateFieldSet(instance, _AbstractOidcService_isNewBrowserSession_override, isNewBrowserSession, "f");
                        try {
                            shouldInjectAccessTokenOrError = getShouldInjectAccessToken(req);
                        }
                        catch (error) {
                            shouldInjectAccessTokenOrError = { error };
                        }
                        __classPrivateFieldSet(instance, _AbstractOidcService_isUserLoggedIn_override, undefined, "f");
                        __classPrivateFieldSet(instance, _AbstractOidcService_isNewBrowserSession_override, undefined, "f");
                        shouldInjectAccessTokenByState.push({
                            isUserLoggedIn,
                            isNewBrowserSession,
                            shouldInjectAccessTokenOrError
                        });
                    }
                }
            }
            instance.decodedIdTokenAccess = undefined;
            if (!instance.providerAwaitsInitialization &&
                isDecodedIdTokenAccessed &&
                !instance.allowDecodedIdTokenAccessInShouldInjectAccessToken) {
                throw new Error("oidc-spa: See https://docs.oidc-spa.dev/release-notes/reading-decodedaccesstoken-within-shouldinjectaccesstoken");
            }
            if (shouldInjectAccessToken === false ||
                (shouldInjectAccessToken === undefined &&
                    shouldInjectAccessTokenByState.every(({ shouldInjectAccessTokenOrError }) => shouldInjectAccessTokenOrError === false))) {
                return next(req);
            }
            if (__classPrivateFieldGet(instance, _AbstractOidcService_isRunningGetParams, "f") && __classPrivateFieldGet(instance, _AbstractOidcService_deadlockDetectionTimer, "f") === undefined) {
                __classPrivateFieldSet(instance, _AbstractOidcService_deadlockDetectionTimer, setTimeout(() => {
                    const name = this.name.replace(/^_/, "");
                    switch (shouldInjectAccessToken) {
                        case false:
                            assert(false);
                        case true:
                            console.warn([
                                "oidc-spa: Probable deadlock detected!",
                                `Request ${req.method} ${req.urlWithParams} requires an access token,`,
                                `but this request is probably being made inside the async callback of`,
                                `${name}.provide(async () => {...}).`,
                                "At this point the access token cannot exist yet, initialization depends on this request itself."
                            ].join(" "));
                            break;
                        case undefined:
                            console.warn([
                                "oidc-spa: Probable deadlock detected!",
                                `While evaluating shouldInjectAccessToken(req) for ${req.method} ${req.urlWithParams},`,
                                "you accessed synchronous properties of",
                                `\`${uncapitalize(name)} = inject(${name})\` that`,
                                "are only available after initialization completes.",
                                `Requests made inside the ${name}.provide(async () => {...}) callback`,
                                `(typically to fetch the params), should not trigger a read of `,
                                `${uncapitalize(name)}.isUserLoggedIn or other properties inside shouldInjectAccessToken()`,
                                "This creates a causality violation.",
                                "Reorganize your shouldInjectAccessToken implementation to exit early with `false` in those case,",
                                "before accessing the property."
                            ].join(" "));
                            break;
                    }
                }, 4000), "f");
            }
            return from(instance.getAccessToken()).pipe(switchMap(({ accessToken }) => {
                if (__classPrivateFieldGet(instance, _AbstractOidcService_deadlockDetectionTimer, "f") !== undefined) {
                    clearTimeout(__classPrivateFieldGet(instance, _AbstractOidcService_deadlockDetectionTimer, "f"));
                    __classPrivateFieldSet(instance, _AbstractOidcService_deadlockDetectionTimer, undefined, "f");
                }
                if (shouldInjectAccessToken === undefined) {
                    const match = shouldInjectAccessTokenByState.find(({ isUserLoggedIn, isNewBrowserSession }) => isUserLoggedIn === instance.isUserLoggedIn &&
                        isNewBrowserSession ===
                            (instance.isUserLoggedIn ? instance.isNewBrowserSession : undefined));
                    assert(match !== undefined);
                    const { shouldInjectAccessTokenOrError } = match;
                    if (typeof shouldInjectAccessTokenOrError !== "boolean") {
                        const { error } = shouldInjectAccessTokenOrError;
                        throw error;
                    }
                    shouldInjectAccessToken = shouldInjectAccessTokenOrError;
                }
                if (!shouldInjectAccessToken) {
                    return next(req);
                }
                if (!instance.isUserLoggedIn) {
                    throw new OidcAccessedTooEarlyError([
                        `oidc-spa: attempted to inject an Authorization bearer token`,
                        `for ${req.method} ${req.urlWithParams} but the user is not logged in.`,
                        `You shouldn't be attempting to call API endpoint that require authentication`,
                        `when the user isn't logged in.`
                    ].join(" "));
                }
                assert(accessToken !== undefined);
                return next(req.clone({ setHeaders: { Authorization: `Bearer ${accessToken}` } }));
            }));
        };
    }
    static get enforceLoginGuard() {
        const canActivateFn = (async (route) => {
            const instance = inject(this);
            const router = inject(Router);
            await instance.prInitialized;
            const redirectUrl = router.serializeUrl(router.createUrlTree(route.url.map(u => u.path), {
                queryParams: route.queryParams,
                fragment: route.fragment ?? undefined
            }));
            const oidc = __classPrivateFieldGet(instance, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(instance, { callerName: "enforceLoginGuard" });
            const isUrlAlreadyReplaced = window.location.href.replace(/\/$/, "") === redirectUrl.replace(/\/$/, "");
            if (!oidc.isUserLoggedIn) {
                await oidc.login({
                    doesCurrentHrefRequiresAuth: isUrlAlreadyReplaced,
                    redirectUrl
                });
            }
            define_temporary_postLoginRedirectUrl: {
                if (isUrlAlreadyReplaced) {
                    break define_temporary_postLoginRedirectUrl;
                }
                setDesiredPostLoginRedirectUrl({ postLoginRedirectUrl: redirectUrl });
                const history_pushState = history.pushState;
                const history_replaceState = history.replaceState;
                const onNavigated = () => {
                    history.pushState = history_pushState;
                    history.replaceState = history_replaceState;
                    setDesiredPostLoginRedirectUrl({ postLoginRedirectUrl: undefined });
                };
                history.pushState = function pushState(...args) {
                    onNavigated();
                    return history_pushState.call(history, ...args);
                };
                history.replaceState = function replaceState(...args) {
                    onNavigated();
                    return history_replaceState.call(history, ...args);
                };
            }
            return true;
        });
        return canActivateFn;
    }
    get initializationError() {
        const state = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getState).call(this, { callerName: "initializationError" });
        return state.initializationError;
    }
    get issuerUri() {
        return __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "issuerUri" }).issuerUri;
    }
    get clientId() {
        return __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "clientId" }).clientId;
    }
    get validRedirectUri() {
        return __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "validRedirectUri" }).validRedirectUri;
    }
    get backFromAuthServer() {
        const oidc = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "backFromAuthServer" });
        if (!oidc.isUserLoggedIn) {
            throw new Error("oidc-spa: backFromAuthServer called but the user is not logged in.");
        }
        return oidc.backFromAuthServer;
    }
    get isUserLoggedIn() {
        return (__classPrivateFieldGet(this, _AbstractOidcService_isUserLoggedIn_override, "f") ??
            __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "isUserLoggedIn" }).isUserLoggedIn);
    }
    async login(params) {
        await this.prInitialized;
        const oidc = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "login" });
        if (oidc.isUserLoggedIn) {
            throw new Error([
                "oidc-spa: login() called but the user is already logged in.",
                "If you wish to send the user to the login page for some update",
                "use oidc.goToAuthServer() instead"
            ].join(" "));
        }
        return oidc.login({
            ...params,
            doesCurrentHrefRequiresAuth: false
        });
    }
    async renewTokens(params) {
        await this.prInitialized;
        const oidc = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "renewTokens" });
        if (!oidc.isUserLoggedIn) {
            throw new Error("oidc-spa: renewTokens() called but the user is not logged in.");
        }
        return oidc.renewTokens(params);
    }
    async logout(params) {
        await this.prInitialized;
        const oidc = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "logout" });
        if (!oidc.isUserLoggedIn) {
            throw new Error("oidc-spa: logout() called but the user is not logged in.");
        }
        return oidc.logout(params);
    }
    async goToAuthServer(params) {
        await this.prInitialized;
        const oidc = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "goToAuthServer" });
        if (!oidc.isUserLoggedIn) {
            throw new Error("oidc-spa: goToAuthServer() called but the user is not logged in.");
        }
        return oidc.goToAuthServer(params);
    }
    get decodedIdToken$() {
        this.decodedIdTokenAccess?.next();
        return __classPrivateFieldGet(this, _AbstractOidcService_decodedIdToken$, "f");
    }
    get $decodedIdToken() {
        this.decodedIdTokenAccess?.next();
        return __classPrivateFieldGet(this, _AbstractOidcService_$decodedIdToken, "f");
    }
    async getAccessToken() {
        await this.prInitialized;
        const oidc = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "getAccessToken" });
        return oidc.isUserLoggedIn
            ? { isUserLoggedIn: true, accessToken: (await oidc.getTokens()).accessToken }
            : {
                isUserLoggedIn: false
            };
    }
    get isNewBrowserSession() {
        if (__classPrivateFieldGet(this, _AbstractOidcService_isNewBrowserSession_override, "f") !== undefined) {
            return __classPrivateFieldGet(this, _AbstractOidcService_isNewBrowserSession_override, "f");
        }
        const oidc = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getOidc).call(this, { callerName: "isNewBrowserSession" });
        if (!oidc.isUserLoggedIn) {
            throw new Error("oidc-spa: isNewBrowserSession was used but the used is not logged in");
        }
        return oidc.isNewBrowserSession;
    }
}
_AbstractOidcService_warnUserSecondsBeforeAutoLogout = new WeakMap(), _AbstractOidcService_isRunningGetParams = new WeakMap(), _AbstractOidcService_deadlockDetectionTimer = new WeakMap(), _AbstractOidcService_dState = new WeakMap(), _AbstractOidcService_isUserLoggedIn_override = new WeakMap(), _AbstractOidcService_decodedIdToken$ = new WeakMap(), _AbstractOidcService_$decodedIdToken = new WeakMap(), _AbstractOidcService_isNewBrowserSession_override = new WeakMap(), _AbstractOidcService_instances = new WeakSet(), _AbstractOidcService_initialize = function _AbstractOidcService_initialize(params) {
    const { prOidcOrInitializationError } = params;
    prOidcOrInitializationError.then(oidcOrInitializationError => {
        let initializationError = undefined;
        let oidc = undefined;
        if (oidcOrInitializationError instanceof Error) {
            initializationError = oidcOrInitializationError;
        }
        else {
            oidc = oidcOrInitializationError;
            initializationError = oidc.isUserLoggedIn ? undefined : oidc.initializationError;
        }
        __classPrivateFieldGet(this, _AbstractOidcService_dState, "f").resolve({
            oidc,
            initializationError
        });
    });
}, _AbstractOidcService_getPrInitializedNotResolvedErrorMessage = function _AbstractOidcService_getPrInitializedNotResolvedErrorMessage(params) {
    const { callerName } = params;
    return [
        `oidc-spa: ${callerName} called/accessed before`,
        "`oidc.prInitialized` resolved.",
        "You are using `awaitInitialization: false`.",
        "In your template you should wrap your usage of",
        "oidc.isUserLoggedIn, oidc.$decodedIdToken() ect. into",
        "@defer (when oidc.prInitialized | async) { } @placeholder { Loading... }"
    ].join(" ");
}, _AbstractOidcService_getState = function _AbstractOidcService_getState(params) {
    const { callerName } = params;
    const { hasResolved, value } = __classPrivateFieldGet(this, _AbstractOidcService_dState, "f").getState();
    if (!hasResolved) {
        throw new OidcAccessedTooEarlyError(__classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getPrInitializedNotResolvedErrorMessage).call(this, { callerName }));
    }
    return value;
}, _AbstractOidcService_getAutoLoginAndInitializationErrorAccessErrorMessage = function _AbstractOidcService_getAutoLoginAndInitializationErrorAccessErrorMessage(params) {
    const { callerName } = params;
    return [
        `oidc-spa: ${callerName} was accessed but initialization failed.`,
        "You are using `autoLogin: true`, so there is no anonymous state.",
        "Handle this by gating your UI:",
        "if (oidc.initializationError) show an error/fallback."
    ].join(" ");
}, _AbstractOidcService_getOidc = function _AbstractOidcService_getOidc(params) {
    const { callerName } = params;
    const state = __classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getState).call(this, { callerName });
    if (state.oidc === undefined) {
        // initialization failed
        assert(state.initializationError !== undefined);
        throw new Error(__classPrivateFieldGet(this, _AbstractOidcService_instances, "m", _AbstractOidcService_getAutoLoginAndInitializationErrorAccessErrorMessage).call(this, { callerName }));
    }
    return state.oidc;
};
export class OidcAccessedTooEarlyError extends Error {
    constructor(message) {
        super(message);
        Object.setPrototypeOf(this, new.target.prototype);
    }
}
