// oidc-spa/core/createOidc.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.registerExports_earlyInit = registerExports_earlyInit;
exports.registerExports_tokenSubstitution = registerExports_tokenSubstitution;
exports.registerExports_DPoP = registerExports_DPoP;
exports.createOidc = createOidc;
exports.createOidc_nonMemoized = createOidc_nonMemoized;
const oidc_client_ts_1 = require("../vendor/frontend/oidc-client-ts");
const OidcMetadata_1 = require("./OidcMetadata");
const assert_1 = require("../tools/tsafe/assert");
const id_1 = require("../tools/tsafe/id");
const Deferred_1 = require("../tools/Deferred");
const evtIsUserActive_1 = require("./evtIsUserActive");
const startCountdown_1 = require("../tools/startCountdown");
const toHumanReadableDuration_1 = require("../tools/toHumanReadableDuration");
const toFullyQualifiedUrl_1 = require("../tools/toFullyQualifiedUrl");
const OidcInitializationError_1 = require("./OidcInitializationError");
const StateData_1 = require("./StateData");
const logoutPropagationToOtherTabs_1 = require("./logoutPropagationToOtherTabs");
const loginPropagationToOtherTabs_1 = require("./loginPropagationToOtherTabs");
const configId_1 = require("./configId");
const oidcClientTsUserToTokens_1 = require("./oidcClientTsUserToTokens");
const loginSilent_1 = require("./loginSilent");
const AuthResponse_1 = require("./AuthResponse");
const persistedAuthState_1 = require("./persistedAuthState");
const Evt_1 = require("../tools/Evt");
const haveSharedParentDomain_1 = require("../tools/haveSharedParentDomain");
const loginOrGoToAuthServer_1 = require("./loginOrGoToAuthServer");
const lazySessionStorage_1 = require("../tools/lazySessionStorage");
const ongoingLoginOrRefreshProcesses_1 = require("./ongoingLoginOrRefreshProcesses");
const isNewBrowserSession_1 = require("./isNewBrowserSession");
const getIsOnline_1 = require("../tools/getIsOnline");
const isKeycloak_1 = require("../keycloak/isKeycloak");
const INFINITY_TIME_1 = require("../tools/INFINITY_TIME");
const earlyInit_rootRelativeOriginalLocationHref_1 = require("./earlyInit_rootRelativeOriginalLocationHref");
const isLikelyDevServer_1 = require("../tools/isLikelyDevServer");
const createObjectThatThrowsIfAccessed_1 = require("../tools/createObjectThatThrowsIfAccessed");
const instancesThatCantUseIframes_1 = require("./instancesThatCantUseIframes");
const desiredPostLoginRedirectUrl_1 = require("./desiredPostLoginRedirectUrl");
const homeAndRedirectUri_1 = require("./homeAndRedirectUri");
const ensureNonBlankPaint_1 = require("../tools/ensureNonBlankPaint");
const StateDataCookie_1 = require("./StateDataCookie");
const loadWebcryptoLinerShim_1 = require("../tools/loadWebcryptoLinerShim");
const lazySessionStorage_2 = require("../tools/lazySessionStorage");
// NOTE: Replaced at build time
const VERSION = "10.2.6";
const globalContext = {
    prOidcByConfigId: new Map(),
    hasLogoutBeenCalled: (0, id_1.id)(false),
    dExports_earlyInit: new Deferred_1.Deferred(),
    dExports_tokenSubstitution: new Deferred_1.Deferred(),
    dExports_DPoP: new Deferred_1.Deferred()
};
function registerExports_earlyInit(exports) {
    globalContext.dExports_earlyInit.resolve(exports);
}
function registerExports_tokenSubstitution(exports) {
    globalContext.dExports_tokenSubstitution.resolve(exports);
}
function registerExports_DPoP(exports) {
    globalContext.dExports_DPoP.resolve(exports);
}
/** @see: https://docs.oidc-spa.dev/v/v10/usage */
async function createOidc(params) {
    for (const name of ["issuerUri", "clientId"]) {
        const value = params[name];
        if (!value) {
            throw new Error(`The parameter "${name}" is required, you provided: ${value}. (Forgot a .env variable?)`);
        }
    }
    const { issuerUri: issuerUri_params, clientId, debugLogs, ...rest } = params;
    const issuerUri = (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
        urlish: issuerUri_params,
        doAssertNoQueryParams: true,
        doOutputWithTrailingSlash: false
    });
    const log = (() => {
        if (!debugLogs) {
            return undefined;
        }
        return (0, id_1.id)((...[first, ...rest]) => {
            const label = "oidc-spa";
            if (typeof first === "string") {
                console.log(...[`${label}: ${first}`, ...rest]);
            }
            else {
                console.log(...[`${label}:`, first, ...rest]);
            }
        });
    })();
    const configId = (0, configId_1.getConfigId)({ issuerUri, clientId });
    const { prOidcByConfigId } = globalContext;
    use_previous_instance: {
        const prOidc = prOidcByConfigId.get(configId);
        if (prOidc === undefined) {
            break use_previous_instance;
        }
        log?.([
            `createOidc was called again with the same config (${JSON.stringify({
                issuerUri,
                clientId
            })})`,
            `Returning the previous instance. All potential different parameters are ignored.`
        ].join(" "));
        // @ts-expect-error: We know what we're doing
        return prOidc;
    }
    const dOidc = new Deferred_1.Deferred();
    prOidcByConfigId.set(configId, dOidc.pr);
    const oidc = await createOidc_nonMemoized(rest, {
        issuerUri,
        clientId,
        configId,
        log
    });
    dOidc.resolve(oidc);
    return oidc;
}
async function createOidc_nonMemoized(params, preProcessedParams) {
    const { transformUrlBeforeRedirect, extraQueryParams: extraQueryParamsOrGetter, extraTokenParams: extraTokenParamsOrGetter, decodedIdTokenSchema, idleSessionLifetimeInSeconds, autoLogoutParams = { redirectTo: "current page" }, autoLogin = false, postLoginRedirectUrl: postLoginRedirectUrl_default, __unsafe_clientSecret, __unsafe_useIdTokenAsAccessToken = false, __metadata, disableDPoP: disableDPoP_params = false, sessionRestorationMethod: sessionRestorationMethod_params, BASE_URL: BASE_URL_params } = params;
    const exports_earlyInit = await (async () => {
        const timer = window.setTimeout(() => {
            console.warn([
                "oidc-spa: Setup error.",
                "oidcEarlyInit() wasn't called.",
                "This is supposed to be handled by the oidc-spa Vite plugin",
                "or manually in other environments."
            ].join(" "));
        }, 3000);
        const exports_earlyInit = await globalContext.dExports_earlyInit.pr;
        window.clearTimeout(timer);
        return exports_earlyInit;
    })();
    if (!exports_earlyInit.shouldLoadApp) {
        return new Promise(() => { });
    }
    const { getEvtIframeAuthResponse, getRedirectAuthResponse, sessionRestorationMethod: sessionRestorationMethod_earlyInit } = exports_earlyInit;
    const sessionRestorationMethod = sessionRestorationMethod_params ?? sessionRestorationMethod_earlyInit ?? "auto";
    const { value: exports_tokenSubstitution } = globalContext.dExports_tokenSubstitution.getState();
    const { value: exports_DPoP } = globalContext.dExports_DPoP.getState();
    const scopes = Array.from(new Set(["openid", ...(params.scopes ?? ["profile"])]));
    const { issuerUri, clientId, configId, log } = preProcessedParams;
    if (window.crypto.subtle === undefined) {
        log?.("window.crypto.subtle not present, lazily loading polyfills.");
        await (0, loadWebcryptoLinerShim_1.loadWebcryptoLinerShim)();
    }
    const getExtraQueryParams = (() => {
        if (extraQueryParamsOrGetter === undefined) {
            return undefined;
        }
        if (typeof extraQueryParamsOrGetter !== "function") {
            return () => extraQueryParamsOrGetter;
        }
        return extraQueryParamsOrGetter;
    })();
    const getExtraTokenParams = (() => {
        if (extraTokenParamsOrGetter === undefined) {
            return undefined;
        }
        if (typeof extraTokenParamsOrGetter !== "function") {
            return () => extraTokenParamsOrGetter;
        }
        return extraTokenParamsOrGetter;
    })();
    const { homeUrlAndRedirectUri } = (0, homeAndRedirectUri_1.getHomeAndRedirectUri)({ BASE_URL_params });
    log?.(`Calling createOidc v${VERSION} ${JSON.stringify({
        issuerUri,
        clientId,
        scopes,
        validRedirectUri: homeUrlAndRedirectUri
    }, null, 2)}`);
    if (exports_tokenSubstitution !== undefined) {
        log?.([
            "Token substitution successfully enabled.",
            "",
            "→ Tokens exposed to the application layer are unusable for resource server calls.",
            "→ They remain structurally valid JWTs, but their signature segment is replaced.",
            "→ Before any request leaves the app (fetch, XHR, WebSocket, beacon),",
            "   the real tokens are restored inside a fully hardened, sandboxed interceptor.",
            "",
            "This means that even with an XSS vulnerability or a compromised dependency,",
            "an attacker cannot extract valid tokens, unless they have also compromised",
            "your application's build pipeline."
        ].join("\n"));
    }
    const stateUrlParamValue_instance = (0, StateData_1.generateStateUrlParamValue)();
    const oidcMetadata = __metadata ?? (await (0, OidcMetadata_1.fetchOidcMetadata)({ issuerUri }));
    const shouldEnableDPoP = (() => {
        if (disableDPoP_params) {
            log?.("DPoP explicitly disabled for this instance");
            return false;
        }
        if (exports_DPoP === undefined) {
            log?.("DPoP disabled, to enable it see: https://docs.oidc-spa.dev/features/dpop");
            return false;
        }
        if (__unsafe_useIdTokenAsAccessToken) {
            if (exports_DPoP.isEnforced) {
                throw new Error([
                    "oidc-spa: Cannot enable DPoP when",
                    "__unsafe_useIdTokenAsAccessToken is set to true"
                ].join(" "));
            }
            log?.("DPoP Disabled due to __unsafe_useIdTokenAsAccessToken: true");
            return false;
        }
        if (oidcMetadata === undefined) {
            return false;
        }
        const isSupported = (() => {
            const { dpop_signing_alg_values_supported } = oidcMetadata;
            if (dpop_signing_alg_values_supported === undefined) {
                return false;
            }
            return dpop_signing_alg_values_supported.includes("ES256");
        })();
        if (!isSupported) {
            if (exports_DPoP.isEnforced) {
                throw new Error("oidc-spa: The IdP does not support DPoP");
            }
            log?.("DPoP disabled because it's not supported by the IdP");
        }
        else {
            log?.("DPoP enabled");
        }
        return isSupported;
    })();
    const canUseIframe = (() => {
        switch (sessionRestorationMethod) {
            case "auto":
                break;
            case "full page redirect":
                return false;
            case "iframe":
                return true;
            default:
                assert_1.assert;
        }
        third_party_cookies: {
            if (oidcMetadata === undefined) {
                return false;
            }
            const { authorization_endpoint } = oidcMetadata;
            (0, assert_1.assert)(authorization_endpoint !== undefined, "Missing authorization_endpoint on the provided __metadata");
            const isOidcServerThirdPartyRelativeToApp = !(0, haveSharedParentDomain_1.getHaveSharedParentDomain)({
                url1: window.location.origin,
                // TODO: No, here we should test against the authorization endpoint!
                url2: authorization_endpoint
            });
            if (!isOidcServerThirdPartyRelativeToApp) {
                break third_party_cookies;
            }
            const isLikelyDevServer = (0, isLikelyDevServer_1.getIsLikelyDevServer)();
            const domain_auth = new URL(authorization_endpoint).origin.split("//")[1];
            (0, assert_1.assert)(domain_auth !== undefined, "33921384");
            const domain_here = window.location.origin.split("//")[1];
            let isWellKnownProviderDomain = false;
            let isIp = false;
            const suggestedDeployments = (() => {
                if (/^(?:\d{1,3}\.){3}\d{1,3}$|^\[?[A-Fa-f0-9:]+\]?$/.test(domain_auth)) {
                    isIp = true;
                    return [];
                }
                const baseDomain = (() => {
                    const segments = domain_auth.split(".");
                    if (segments.length >= 3) {
                        segments.shift();
                    }
                    return segments.join(".");
                })();
                {
                    const baseDomain_low = baseDomain.toLowerCase();
                    if (baseDomain_low.includes("auth0") ||
                        baseDomain_low.includes("clerk") ||
                        baseDomain_low.includes("microsoft") ||
                        baseDomain_low.includes("okta") ||
                        baseDomain_low.includes("aws")) {
                        isWellKnownProviderDomain = true;
                        return [];
                    }
                }
                const baseUrl = new URL(homeUrlAndRedirectUri).pathname;
                return [
                    `myapp.${baseDomain}`,
                    baseDomain === domain_auth ? undefined : baseDomain,
                    `${baseDomain}/${baseUrl === "/" ? "dashboard" : baseUrl}`
                ].filter(x => x !== undefined);
            })();
            if (isLikelyDevServer) {
                log?.([
                    "Detected localhost environment.",
                    "\nWhen reloading while logged in, you will briefly see",
                    "some URL params appear in the address bar.",
                    "\nThis happens because session restore via iframe is disabled,",
                    "the browser treats your auth server as a third party.",
                    `\nAuth server: ${domain_auth}`,
                    `\nApp domain:  ${domain_here}`,
                    ...(() => {
                        if (isIp) {
                            return [];
                        }
                        if (isWellKnownProviderDomain) {
                            return [
                                "\nYou seem to be using a well-known auth provider.",
                                "Check your provider's docs, some allow configuring",
                                `a your custom domain at least for the authorization endpoint.`,
                                "\nIf configured, oidc-spa will restore sessions silently",
                                "and improve the user experience."
                            ];
                        }
                        return [
                            "\nOnce deployed under the same root domain as your auth server,",
                            "oidc-spa will use iframes to restore sessions silently.",
                            "\nSuggested deployments:",
                            ...suggestedDeployments.map(d => `\n  • ${d}`)
                        ];
                    })(),
                    "\n\nMore info:",
                    "https://docs.oidc-spa.dev/v/v10/resources/third-party-cookies-and-session-restoration"
                ].join(" "));
            }
            else {
                log?.([
                    "Silent session restore via iframe is disabled.",
                    `\nAuth server: ${domain_auth}`,
                    `App domain:  ${domain_here}`,
                    "\nThey do not share a common root domain.",
                    ...(() => {
                        if (isIp) {
                            return [];
                        }
                        if (isWellKnownProviderDomain) {
                            return [
                                "\nYou seem to be using a well-known auth provider.",
                                "Check if you can configure a custom auth domain.",
                                "\nIf so, oidc-spa can restore sessions silently",
                                "and improve the user experience."
                            ];
                        }
                        return [
                            "\nTo improve the experience, here are some examples of deployment for your app:",
                            ...suggestedDeployments.map(d => `\n  • ${d}`)
                        ];
                    })(),
                    "\nMore info:",
                    "https://docs.oidc-spa.dev/v/v10/resources/third-party-cookies-and-session-restoration"
                ].join(" "));
            }
            return false;
        }
        return true;
    })();
    if (!canUseIframe) {
        (0, instancesThatCantUseIframes_1.notifyNewInstanceThatCantUseIframes)();
        if (instancesThatCantUseIframes_1.evtIsThereMoreThanOneInstanceThatCantUserIframes.current) {
            log?.([
                "More than one oidc instance can't use iframe",
                "falling back to persisting tokens in session storage"
            ].join(" "));
        }
    }
    const oidcClientTsUserManager = oidcMetadata === undefined
        ? (0, createObjectThatThrowsIfAccessed_1.createObjectThatThrowsIfAccessed)({
            debugMessage: "oidc-spa: Wrong assertion 43943"
        })
        : new oidc_client_ts_1.UserManager({
            stateUrlParamValue: stateUrlParamValue_instance,
            authority: issuerUri,
            client_id: clientId,
            redirect_uri: homeUrlAndRedirectUri,
            silent_redirect_uri: homeUrlAndRedirectUri,
            post_logout_redirect_uri: homeUrlAndRedirectUri,
            response_mode: (0, isKeycloak_1.isKeycloak)({ issuerUri }) && !(0, StateDataCookie_1.getIsStateDataCookieEnabled)() ? "fragment" : "query",
            response_type: "code",
            scope: scopes.join(" "),
            userStore: new oidc_client_ts_1.WebStorageStateStore({
                store: (() => {
                    if (canUseIframe) {
                        return new oidc_client_ts_1.InMemoryWebStorage();
                    }
                    const storage = (0, lazySessionStorage_1.createLazySessionStorage)({ storageId: configId });
                    if (instancesThatCantUseIframes_1.evtIsThereMoreThanOneInstanceThatCantUserIframes.current) {
                        storage.persistCurrentStateAndSubsequentChanges();
                    }
                    else {
                        instancesThatCantUseIframes_1.evtIsThereMoreThanOneInstanceThatCantUserIframes.subscribe(() => {
                            storage.persistCurrentStateAndSubsequentChanges();
                        });
                    }
                    return storage;
                })()
            }),
            stateStore: new oidc_client_ts_1.WebStorageStateStore({
                store: localStorage,
                prefix: StateData_1.STATE_STORE_KEY_PREFIX
            }),
            client_secret: __unsafe_clientSecret,
            metadata: oidcMetadata,
            dpop: (() => {
                if (!shouldEnableDPoP) {
                    return undefined;
                }
                (0, assert_1.assert)(exports_DPoP !== undefined, "49240");
                return {
                    store: exports_DPoP.createDPoPStore({
                        implementation: (0, loadWebcryptoLinerShim_1.hasLoadWebcryptoLinerShimBeenCalled)()
                            ? "in memory"
                            : "indexedDB",
                        configId,
                        clientId
                    })
                };
            })()
        });
    const evtInitializationOutcomeUserNotLoggedIn = (0, Evt_1.createEvt)();
    const { loginOrGoToAuthServer } = (0, loginOrGoToAuthServer_1.createLoginOrGoToAuthServer)({
        configId,
        oidcClientTsUserManager,
        transformUrlBeforeRedirect,
        getExtraQueryParams,
        getExtraTokenParams,
        homeUrl: homeUrlAndRedirectUri,
        stateUrlParamValue_instance,
        evtInitializationOutcomeUserNotLoggedIn,
        log
    });
    const { loginSilent } = (0, loginSilent_1.createLoginSilent)({
        getEvtIframeAuthResponse,
        oidcClientTsUserManager,
        stateUrlParamValue_instance,
        configId,
        transformUrlBeforeRedirect,
        getExtraQueryParams,
        getExtraTokenParams,
        autoLogin,
        log
    });
    const { getIsNewBrowserSession } = (0, isNewBrowserSession_1.createGetIsNewBrowserSession)({
        configId,
        evtInitializationOutcomeUserNotLoggedIn
    });
    const { completeLoginOrRefreshProcess } = await (0, ongoingLoginOrRefreshProcesses_1.startLoginOrRefreshProcess)();
    const resultOfLoginProcess = await (async () => {
        if (oidcMetadata === undefined) {
            return (await Promise.resolve().then(() => __importStar(require("./diagnostic")))).createWellKnownOidcConfigurationEndpointUnreachableInitializationError({
                issuerUri
            });
        }
        restore_from_session_storage: {
            if (canUseIframe) {
                break restore_from_session_storage;
            }
            if (!instancesThatCantUseIframes_1.evtIsThereMoreThanOneInstanceThatCantUserIframes.current) {
                break restore_from_session_storage;
            }
            let oidcClientTsUser;
            try {
                oidcClientTsUser = await oidcClientTsUserManager.getUser();
            }
            catch {
                // NOTE: Not sure if it can throw, but let's be safe.
                oidcClientTsUser = null;
                try {
                    await oidcClientTsUserManager.removeUser();
                }
                catch { }
            }
            if (oidcClientTsUser === null) {
                break restore_from_session_storage;
            }
            log?.("Session was restored from session storage");
            return {
                oidcClientTsUser,
                backFromAuthServer: undefined,
                isRestoredFromSessionStorage: true
            };
        }
        handle_redirect_auth_response: {
            let stateDataAndAuthResponse = undefined;
            {
                const { authResponse, clearAuthResponse } = getRedirectAuthResponse();
                if (authResponse === undefined) {
                    break handle_redirect_auth_response;
                }
                const stateData = (0, StateData_1.getStateData)({ stateUrlParamValue: authResponse.state });
                if (stateData === undefined) {
                    clearAuthResponse();
                    break handle_redirect_auth_response;
                }
                if (stateData.configId !== configId) {
                    break handle_redirect_auth_response;
                }
                (0, assert_1.assert)(stateData.context === "redirect", "3229492");
                clearAuthResponse();
                stateDataAndAuthResponse = { stateData, authResponse };
            }
            if (stateDataAndAuthResponse === undefined) {
                break handle_redirect_auth_response;
            }
            const { stateData, authResponse } = stateDataAndAuthResponse;
            switch (stateData.action) {
                case "login":
                    {
                        log?.(`Handling login redirect auth response ${JSON.stringify({
                            ...authResponse,
                            ...(authResponse.code === undefined
                                ? undefined
                                : {
                                    code: authResponse.code.slice(0, 20) + "..."
                                })
                        }, null, 2)}`);
                        const authResponseUrl = (0, AuthResponse_1.authResponseToUrl)(authResponse);
                        (0, StateDataCookie_1.clearStateDataCookie)({ stateUrlParamValue: authResponse.state });
                        let oidcClientTsUser = undefined;
                        try {
                            oidcClientTsUser = await oidcClientTsUserManager.signinRedirectCallback(authResponseUrl);
                        }
                        catch (error) {
                            (0, assert_1.assert)(error instanceof Error, "741947");
                            if (error.message === "Failed to fetch") {
                                return (await Promise.resolve().then(() => __importStar(require("./diagnostic")))).createFailedToFetchTokenEndpointInitializationError({
                                    clientId,
                                    issuerUri
                                });
                            }
                            if (error.message === "Invalid client or Invalid client credentials") {
                                return (await Promise.resolve().then(() => __importStar(require("./diagnostic")))).createInvalidClientCredentialsInitializationError({
                                    clientId,
                                    issuerUri
                                });
                            }
                            {
                                const authResponse_error = authResponse.error;
                                if (authResponse_error !== undefined) {
                                    log?.([
                                        `The auth server responded with: ${authResponse_error},`,
                                        `trying to restore session as if we didn't had a auth response.`
                                    ].join(" "));
                                    break handle_redirect_auth_response;
                                }
                            }
                            return error;
                        }
                        (0, loginPropagationToOtherTabs_1.notifyOtherTabsOfLogin)({ configId });
                        return {
                            oidcClientTsUser,
                            isRestoredFromSessionStorage: false,
                            backFromAuthServer: {
                                extraQueryParams: stateData.extraQueryParams,
                                result: Object.fromEntries(Object.entries(authResponse)
                                    .map(([name, value]) => {
                                    if (name === "state" ||
                                        name === "session_state" ||
                                        name === "iss" ||
                                        name === "code") {
                                        return undefined;
                                    }
                                    if (value === undefined) {
                                        return undefined;
                                    }
                                    return [name, value];
                                })
                                    .filter(entry => entry !== undefined))
                            }
                        };
                    }
                    break;
                case "logout":
                    {
                        log?.("Handling logout redirect auth response", authResponse);
                        const authResponseUrl = (0, AuthResponse_1.authResponseToUrl)(authResponse);
                        try {
                            await oidcClientTsUserManager.signoutRedirectCallback(authResponseUrl);
                        }
                        catch { }
                        (0, logoutPropagationToOtherTabs_1.notifyOtherTabsOfLogout)({
                            configId,
                            sessionId: stateData.sessionId
                        });
                        if (autoLogin ||
                            location.pathname.endsWith(".html") ||
                            location.pathname.endsWith(".htm")) {
                            location.reload();
                            await new Promise(() => { });
                        }
                        // NOTE: The user is no longer logged in.
                        return undefined;
                    }
                    break;
                default:
                    (0, assert_1.assert)(false);
            }
        }
        silent_login_if_possible_and_auto_login: {
            const persistedAuthState = (0, persistedAuthState_1.getPersistedAuthState)({ configId });
            if (persistedAuthState === "explicitly logged out" && !autoLogin) {
                log?.("Skipping session restoration completely, the user is explicitly logged out.");
                break silent_login_if_possible_and_auto_login;
            }
            {
                const { isOnline, prOnline } = (0, getIsOnline_1.getIsOnline)();
                if (!isOnline) {
                    if (autoLogin) {
                        log?.([
                            "The browser is currently offline",
                            "Since autoLogin is enabled we wait until it comes back online",
                            "to continue with authentication"
                        ].join(" "));
                        await prOnline;
                    }
                    else {
                        log?.([
                            "The browser is not currently online so we proceed with initialization",
                            "assuming the user isn't authenticated"
                        ].join(" "));
                        break silent_login_if_possible_and_auto_login;
                    }
                }
            }
            let authResponse_error = undefined;
            let oidcClientTsUser = undefined;
            actual_silent_signin: {
                if (persistedAuthState === "explicitly logged out") {
                    log?.("Skipping session restoration via iframe (silent signin), the user is explicitly logged out.");
                    break actual_silent_signin;
                }
                if (!canUseIframe) {
                    log?.("Skipping session restoration via iframe (can't use iframe)");
                    break actual_silent_signin;
                }
                log?.("Performing session restoration via iframe (silent signin)");
                const result_loginSilent = await loginSilent({
                    extraTokenParams: undefined
                });
                (0, assert_1.assert)(result_loginSilent.outcome !== "token refreshed using refresh token", "876995");
                (0, assert_1.assert)(result_loginSilent.outcome !== "got error auth response using refresh token", "876996");
                if (result_loginSilent.outcome === "timeout" ||
                    result_loginSilent.outcome === "other error") {
                    const { authorization_endpoint } = oidcMetadata;
                    (0, assert_1.assert)(authorization_endpoint !== undefined, "39447394");
                    return (await Promise.resolve().then(() => __importStar(require("./diagnostic")))).createIframeTimeoutInitializationError({
                        redirectUri: homeUrlAndRedirectUri,
                        clientId,
                        issuerUri,
                        authorizationEndpointUrl: authorization_endpoint
                    });
                }
                assert_1.assert;
                const { authResponse } = result_loginSilent;
                log?.(`Silent signin auth response ${JSON.stringify({
                    ...authResponse,
                    ...(authResponse.code === undefined
                        ? undefined
                        : {
                            code: authResponse.code.slice(0, 20) + "..."
                        })
                }, null, 2)}`);
                (0, StateDataCookie_1.clearStateDataCookie)({ stateUrlParamValue: authResponse.state });
                authResponse_error = authResponse.error;
                try {
                    oidcClientTsUser = await oidcClientTsUserManager.signinRedirectCallback((0, AuthResponse_1.authResponseToUrl)(authResponse));
                }
                catch (error) {
                    (0, assert_1.assert)(error instanceof Error, "433344");
                    if (error.message === "Failed to fetch") {
                        return (await Promise.resolve().then(() => __importStar(require("./diagnostic")))).createFailedToFetchTokenEndpointInitializationError({
                            clientId,
                            issuerUri
                        });
                    }
                    if (error.message === "Invalid client or Invalid client credentials") {
                        return (await Promise.resolve().then(() => __importStar(require("./diagnostic")))).createInvalidClientCredentialsInitializationError({
                            clientId,
                            issuerUri
                        });
                    }
                    if (authResponse_error === undefined) {
                        return error;
                    }
                }
            }
            if (oidcClientTsUser === undefined) {
                if (autoLogin ||
                    (persistedAuthState === "logged in" &&
                        (authResponse_error === undefined ||
                            authResponse_error === "interaction_required" ||
                            authResponse_error === "login_required" ||
                            authResponse_error === "consent_required" ||
                            authResponse_error === "account_selection_required"))) {
                    log?.("Performing session restoration (or autoLogin) with full page redirect");
                    completeLoginOrRefreshProcess();
                    if (autoLogin && persistedAuthState !== "logged in") {
                        evtInitializationOutcomeUserNotLoggedIn.post();
                    }
                    (0, ensureNonBlankPaint_1.ensureNonBlankPaint)();
                    await (0, ongoingLoginOrRefreshProcesses_1.waitForAllOtherOngoingLoginOrRefreshProcessesToComplete)({
                        prUnlock: (0, loginOrGoToAuthServer_1.getPrSafelyRestoredFromBfCacheAfterLoginBackNavigationOrInitializationError)()
                    });
                    await loginOrGoToAuthServer({
                        action: "login",
                        doForceReloadOnBfCache: true,
                        redirectUrl: (() => {
                            if (postLoginRedirectUrl_default) {
                                return postLoginRedirectUrl_default;
                            }
                            if (!instancesThatCantUseIframes_1.evtIsThereMoreThanOneInstanceThatCantUserIframes.current) {
                                return (0, earlyInit_rootRelativeOriginalLocationHref_1.getRootRelativeOriginalLocationHref_earlyInit)();
                            }
                            return (0, desiredPostLoginRedirectUrl_1.getDesiredPostLoginRedirectUrl)() ?? window.location.href;
                        })(),
                        doNavigateBackToLastPublicUrlIfTheTheUserNavigateBack: true,
                        extraQueryParams_local: undefined,
                        transformUrlBeforeRedirect_local: undefined,
                        interaction: (() => {
                            if (persistedAuthState === "explicitly logged out") {
                                return "ensure interaction";
                            }
                            if (autoLogin) {
                                return "directly redirect if active session show login otherwise";
                            }
                            return "ensure no interaction";
                        })(),
                        preRedirectHook: () => {
                            (0, persistedAuthState_1.persistAuthState)({ configId, state: undefined });
                        }
                    });
                }
                if (authResponse_error !== undefined) {
                    log?.([
                        `The auth server responded with: ${authResponse_error} `,
                        "login_required" === authResponse_error
                            ? `(login_required just means that there's no active session for the user)`
                            : ""
                    ].join(""));
                }
                break silent_login_if_possible_and_auto_login;
            }
            log?.("Successful session restoration via iframe (silent signin)");
            return {
                oidcClientTsUser,
                backFromAuthServer: undefined,
                isRestoredFromSessionStorage: false
            };
        }
        // NOTE: The user is not logged in.
        return undefined;
    })();
    completeLoginOrRefreshProcess();
    await (0, ongoingLoginOrRefreshProcesses_1.waitForAllOtherOngoingLoginOrRefreshProcessesToComplete)({
        prUnlock: Promise.resolve()
    });
    const oidc_common = {
        issuerUri,
        clientId,
        validRedirectUri: homeUrlAndRedirectUri
    };
    not_loggedIn_case: {
        if (!(resultOfLoginProcess instanceof Error) && resultOfLoginProcess !== undefined) {
            break not_loggedIn_case;
        }
        evtInitializationOutcomeUserNotLoggedIn.post();
        if ((0, persistedAuthState_1.getPersistedAuthState)({ configId }) !== "explicitly logged out") {
            (0, persistedAuthState_1.persistAuthState)({ configId, state: undefined });
        }
        const oidc_notLoggedIn = (() => {
            if (resultOfLoginProcess instanceof Error) {
                log?.("User not logged in and there was an initialization error");
                const error = resultOfLoginProcess;
                const initializationError = error instanceof OidcInitializationError_1.OidcInitializationError
                    ? error
                    : new OidcInitializationError_1.OidcInitializationError({
                        isAuthServerLikelyDown: false,
                        messageOrCause: error
                    });
                if (autoLogin) {
                    throw initializationError;
                }
                console.error([
                    `oidc-spa Initialization Error: `,
                    `isAuthServerLikelyDown: ${initializationError.isAuthServerLikelyDown}`,
                    ``,
                    initializationError.message
                ].join("\n"));
                return (0, id_1.id)({
                    ...oidc_common,
                    isUserLoggedIn: false,
                    login: async () => {
                        alert("Authentication is currently unavailable. Please try again later.");
                        return new Promise(() => { });
                    },
                    initializationError
                });
            }
            if (resultOfLoginProcess === undefined) {
                log?.("User not logged in");
                return (0, id_1.id)({
                    ...oidc_common,
                    isUserLoggedIn: false,
                    login: async ({ doesCurrentHrefRequiresAuth = false, extraQueryParams, redirectUrl, transformUrlBeforeRedirect } = {}) => {
                        await (0, ongoingLoginOrRefreshProcesses_1.waitForAllOtherOngoingLoginOrRefreshProcessesToComplete)({
                            prUnlock: (0, loginOrGoToAuthServer_1.getPrSafelyRestoredFromBfCacheAfterLoginBackNavigationOrInitializationError)()
                        });
                        if (!canUseIframe) {
                            log?.([
                                "IMPORTANT DEBUG INFO:",
                                "\nWe are about to redirect to your Identity Provider (IdP).",
                                "\nIf you see an 'Invalid Redirect URI' error on the IdP page, make sure you've added:",
                                `\n${homeUrlAndRedirectUri}`,
                                "\nto the list of valid redirect URIs in your IdP configuration.",
                                "\nIf you see a 'Client not found' error make sure you've created the flowing OIDC client:",
                                `\n${clientId}`
                            ].join(" "));
                        }
                        return loginOrGoToAuthServer({
                            action: "login",
                            doNavigateBackToLastPublicUrlIfTheTheUserNavigateBack: doesCurrentHrefRequiresAuth,
                            doForceReloadOnBfCache: false,
                            redirectUrl: redirectUrl ?? postLoginRedirectUrl_default ?? window.location.href,
                            extraQueryParams_local: extraQueryParams,
                            transformUrlBeforeRedirect_local: transformUrlBeforeRedirect,
                            interaction: (0, persistedAuthState_1.getPersistedAuthState)({ configId }) === "explicitly logged out"
                                ? "ensure interaction"
                                : "directly redirect if active session show login otherwise",
                            preRedirectHook: undefined
                        });
                    },
                    initializationError: undefined
                });
            }
            (0, assert_1.assert)(false);
        })();
        {
            const { prOtherTabLogin } = (0, loginPropagationToOtherTabs_1.getPrOtherTabLogin)({
                configId
            });
            prOtherTabLogin.then(async () => {
                log?.(`Other tab has logged in, reloading this tab`);
                await (0, ongoingLoginOrRefreshProcesses_1.waitForAllOtherOngoingLoginOrRefreshProcessesToComplete)({
                    prUnlock: new Promise(() => { })
                });
                window.location.reload();
            });
        }
        // @ts-expect-error: We know what we're doing
        return oidc_notLoggedIn;
    }
    log?.("User is logged in");
    (0, assert_1.assert)(oidcMetadata !== undefined, "30483403");
    const { oidcClientTsUserToTokens } = (0, oidcClientTsUserToTokens_1.createOidcClientTsUserToTokens)({
        configId,
        decodedIdTokenSchema,
        __unsafe_useIdTokenAsAccessToken,
        exports_DPoP: shouldEnableDPoP ? exports_DPoP : undefined,
        exports_tokenSubstitution,
        log
    });
    let currentTokens = oidcClientTsUserToTokens({
        oidcClientTsUser: resultOfLoginProcess.oidcClientTsUser,
        decodedIdToken_previous: undefined
    });
    detect_useless_idleSessionLifetimeInSeconds: {
        if (idleSessionLifetimeInSeconds === undefined) {
            break detect_useless_idleSessionLifetimeInSeconds;
        }
        if (idleSessionLifetimeInSeconds <= 30) {
            throw new Error([
                `oidc-spa: Setting idleSessionLifetimeInSeconds to anything bellow 5 minutes (300)`,
                `is probably a misunderstanding of what this parameter does.`,
                `See: https://docs.oidc-spa.dev/v/v10/auto-logout`
            ].join(" "));
        }
        if (currentTokens.refreshTokenExpirationTime === undefined) {
            break detect_useless_idleSessionLifetimeInSeconds;
        }
        console.warn([
            "oidc-spa: You've specified idleSessionLifetimeInSeconds,",
            "but your auth server issues a refresh_token with a known expiration time.",
            "idleSessionLifetimeInSeconds should only be used as a fallback",
            "for auth servers that don't specify when an inactive session expires (Auth0, Clerk, WorkOS).",
            "The auth server, not your code, is the source of truth.",
            "See: https://docs.oidc-spa.dev/v/v10/auto-logout"
        ].join(" "));
    }
    {
        if ((0, persistedAuthState_1.getPersistedAuthState)({ configId }) !== undefined) {
            (0, persistedAuthState_1.persistAuthState)({ configId, state: undefined });
        }
        if (!canUseIframe) {
            (0, persistedAuthState_1.persistAuthState)({
                configId,
                state: {
                    stateDescription: "logged in",
                    refreshTokenExpirationTime: currentTokens.refreshTokenExpirationTime,
                    serverDateNow: currentTokens.getServerDateNow(),
                    idleSessionLifetimeInSeconds
                }
            });
        }
    }
    const autoLogoutCountdownTickCallbacks = new Set();
    const onTokenChanges = new Set();
    const { sid: sessionId, sub: subjectId } = currentTokens.decodedIdToken_original;
    (0, assert_1.assert)(subjectId !== undefined, "The 'sub' claim is missing from the id token");
    (0, assert_1.assert)(sessionId === undefined || typeof sessionId === "string");
    let wouldHaveAutoLoggedOutIfBrowserWasOnline = false;
    let prOngoingTokenRenewal = undefined;
    const oidc_loggedIn = (0, id_1.id)({
        ...oidc_common,
        isUserLoggedIn: true,
        getTokens: async () => {
            if (wouldHaveAutoLoggedOutIfBrowserWasOnline) {
                await oidc_loggedIn.logout(autoLogoutParams);
                (0, assert_1.assert)(false);
            }
            if (prOngoingTokenRenewal === undefined) {
                // NOTE: Give a chance to renewOnLocalTimeShift to do it's job
                await new Promise(resolve => setTimeout(resolve, 0));
            }
            if (prOngoingTokenRenewal !== undefined) {
                await prOngoingTokenRenewal;
            }
            renew_tokens: {
                {
                    const msBeforeExpirationOfTheAccessToken = currentTokens.accessTokenExpirationTime - currentTokens.getServerDateNow();
                    if (msBeforeExpirationOfTheAccessToken > 30000) {
                        break renew_tokens;
                    }
                }
                {
                    const msElapsedSinceCurrentTokenWereIssued = currentTokens.getServerDateNow() - currentTokens.issuedAtTime;
                    if (msElapsedSinceCurrentTokenWereIssued < 5000) {
                        break renew_tokens;
                    }
                }
                await oidc_loggedIn.renewTokens();
            }
            return currentTokens;
        },
        getDecodedIdToken: () => currentTokens.decodedIdToken,
        logout: async (params) => {
            if (globalContext.hasLogoutBeenCalled) {
                log?.("logout() has already been called, ignoring the call");
                return new Promise(() => { });
            }
            globalContext.hasLogoutBeenCalled = true;
            const rootRelativePostLogoutRedirectUrl = (() => {
                switch (params.redirectTo) {
                    case "current page":
                        return window.location.href;
                    case "home":
                        return homeUrlAndRedirectUri;
                    case "specific url":
                        return (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                            urlish: params.url,
                            doAssertNoQueryParams: false
                        });
                }
            })().slice(window.location.origin.length);
            await (0, ongoingLoginOrRefreshProcesses_1.waitForAllOtherOngoingLoginOrRefreshProcessesToComplete)({
                prUnlock: new Promise(() => { })
            });
            if (!oidcMetadata.end_session_endpoint) {
                log?.("No end session endpoint, managing logging state locally");
                (0, persistedAuthState_1.persistAuthState)({ configId, state: { stateDescription: "explicitly logged out" } });
                try {
                    await oidcClientTsUserManager.removeUser();
                }
                catch {
                    // NOTE: Not sure if it can throw
                }
                (0, logoutPropagationToOtherTabs_1.notifyOtherTabsOfLogout)({
                    configId,
                    sessionId
                });
                window.location.href = rootRelativePostLogoutRedirectUrl;
                return new Promise(() => { });
            }
            log?.([
                "IMPORTANT DEBUG INFO:",
                "\nWe are about to redirect to your Identity Provider (IdP).",
                "\nIf you see an 'Invalid Redirect URI' error on the IdP page, make sure you've added:",
                `\n${homeUrlAndRedirectUri}`,
                "\nto the list of valid post logout redirect URIs in your IdP configuration."
            ].join(" "));
            window.addEventListener("pageshow", event => {
                if (!event.persisted) {
                    return;
                }
                location.reload();
            });
            (0, StateDataCookie_1.setStateDataCookieIfEnabled)({
                homeUrl: homeUrlAndRedirectUri,
                stateUrlParamValue_instance,
                stateDataCookie: {
                    action: "logout",
                    rootRelativeRedirectUrl: rootRelativePostLogoutRedirectUrl
                }
            });
            try {
                await oidcClientTsUserManager.signoutRedirect({
                    state: (0, id_1.id)({
                        configId,
                        context: "redirect",
                        rootRelativeRedirectUrl: rootRelativePostLogoutRedirectUrl,
                        action: "logout",
                        sessionId
                    }),
                    redirectMethod: "assign"
                });
            }
            catch (error) {
                (0, assert_1.assert)(false, `signoutRedirect() is not expected to throw but it did: ${String(error)}`);
            }
            return new Promise(() => { });
        },
        renewTokens: (() => {
            // NOTE: Cannot throw (or if it does it's our fault)
            async function renewTokens_nonMutexed(params) {
                const { extraTokenParams } = params;
                const fallbackToFullPageReload = async () => {
                    (0, persistedAuthState_1.persistAuthState)({ configId, state: undefined });
                    await (0, ongoingLoginOrRefreshProcesses_1.waitForAllOtherOngoingLoginOrRefreshProcessesToComplete)({
                        prUnlock: new Promise(() => { })
                    });
                    await loginOrGoToAuthServer({
                        action: "login",
                        redirectUrl: window.location.href,
                        doForceReloadOnBfCache: true,
                        extraQueryParams_local: undefined,
                        transformUrlBeforeRedirect_local: undefined,
                        doNavigateBackToLastPublicUrlIfTheTheUserNavigateBack: true,
                        interaction: "directly redirect if active session show login otherwise",
                        preRedirectHook: undefined
                    });
                    (0, assert_1.assert)(false, "136134");
                };
                if (!currentTokens.hasRefreshToken && !canUseIframe) {
                    log?.([
                        "Unable to refresh tokens without a full app reload,",
                        "because no refresh token is available",
                        "and your app setup prevents silent sign-in via iframe.",
                        "Your only option to refresh tokens is to call `window.location.reload()`"
                    ].join(" "));
                    await fallbackToFullPageReload();
                    (0, assert_1.assert)(false, "136135");
                }
                log?.("Renewing tokens");
                const { completeLoginOrRefreshProcess } = await (0, ongoingLoginOrRefreshProcesses_1.startLoginOrRefreshProcess)();
                const result_loginSilent = await loginSilent({
                    extraTokenParams
                });
                if (result_loginSilent.outcome === "timeout") {
                    log?.([
                        `Silent refresh of the token failed the iframe didn't post a response (timeout).`,
                        `This isn't recoverable, reloading the page.`
                    ].join(" "));
                    window.location.reload();
                    await new Promise(() => { });
                    (0, assert_1.assert)(false);
                }
                const clearPersistedTokensIfSessionStorageIfAny = () => {
                    let hasRemoved = false;
                    for (let i = sessionStorage.length - 1; i >= 0; i--) {
                        const key = sessionStorage.key(i);
                        (0, assert_1.assert)(key !== null, "323303");
                        if (key.startsWith(lazySessionStorage_2.SESSION_STORAGE_GLOBAL_PREFIX)) {
                            hasRemoved = true;
                            sessionStorage.removeItem(key);
                        }
                    }
                    if (hasRemoved) {
                        log?.("The persisted session in sessionStorage was probably no longer valid");
                    }
                };
                let oidcClientTsUser;
                switch (result_loginSilent.outcome) {
                    case "token refreshed using refresh token":
                        {
                            log?.("Refresh token used");
                            oidcClientTsUser = result_loginSilent.oidcClientTsUser;
                        }
                        break;
                    case "got error auth response using refresh token":
                    case "other error":
                        {
                            switch (result_loginSilent.outcome) {
                                case "got error auth response using refresh token":
                                    {
                                        const { authResponse } = result_loginSilent;
                                        log?.([
                                            "Got error response trying to refresh tokens using the refresh token,",
                                            "token endpoint response:",
                                            JSON.stringify(authResponse, null, 2)
                                        ].join(" "));
                                    }
                                    break;
                                case "other error":
                                    {
                                        const { error } = result_loginSilent;
                                        log?.(`Got an unexpected error trying to refresh token: ${error.message}`);
                                    }
                                    break;
                                default:
                                    (0, assert_1.assert)(false);
                                    break;
                            }
                            clearPersistedTokensIfSessionStorageIfAny();
                            completeLoginOrRefreshProcess();
                            await fallbackToFullPageReload();
                            (0, assert_1.assert)(false, "136135");
                        }
                        break;
                    case "got auth response from iframe":
                        {
                            const { authResponse } = result_loginSilent;
                            (0, StateDataCookie_1.clearStateDataCookie)({ stateUrlParamValue: authResponse.state });
                            const authResponse_error = authResponse.error;
                            if (authResponse_error === undefined) {
                                log?.([
                                    "Tokens refreshed using iframe, authorization endpoint response: ",
                                    JSON.stringify(authResponse, null, 2)
                                ].join(" "));
                            }
                            else {
                                log?.([
                                    "Got error response trying to refresh tokens using iframe,",
                                    "Authorization endpoint response:",
                                    JSON.stringify(authResponse, null, 2)
                                ].join(" "));
                            }
                            let oidcClientTsUser_scope = undefined;
                            try {
                                oidcClientTsUser_scope =
                                    await oidcClientTsUserManager.signinRedirectCallback((0, AuthResponse_1.authResponseToUrl)(authResponse));
                            }
                            catch (error) {
                                if (authResponse_error === undefined) {
                                    console.error(error);
                                    (0, assert_1.assert)(false, `This is a bug in oidc-spa, please report.`);
                                }
                            }
                            if (oidcClientTsUser_scope === undefined) {
                                clearPersistedTokensIfSessionStorageIfAny();
                                completeLoginOrRefreshProcess();
                                log?.([
                                    "The user is probably not logged in anymore,",
                                    "need to redirect to login pages"
                                ].join(" "));
                                await fallbackToFullPageReload();
                                (0, assert_1.assert)(false, "136135");
                            }
                            oidcClientTsUser = oidcClientTsUser_scope;
                        }
                        break;
                    default:
                        (0, assert_1.assert)(false);
                        break;
                }
                currentTokens = oidcClientTsUserToTokens({
                    oidcClientTsUser,
                    decodedIdToken_previous: currentTokens.decodedIdToken
                });
                if ((0, persistedAuthState_1.getPersistedAuthState)({ configId }) !== undefined) {
                    (0, persistedAuthState_1.persistAuthState)({
                        configId,
                        state: {
                            stateDescription: "logged in",
                            refreshTokenExpirationTime: currentTokens.refreshTokenExpirationTime,
                            serverDateNow: currentTokens.getServerDateNow(),
                            idleSessionLifetimeInSeconds
                        }
                    });
                }
                Array.from(onTokenChanges).forEach(onTokenChange => onTokenChange(currentTokens));
                completeLoginOrRefreshProcess();
            }
            let ongoingCall = undefined;
            function handleThen() {
                (0, assert_1.assert)(ongoingCall !== undefined, "131276");
                const { pr } = ongoingCall;
                pr.then(() => {
                    (0, assert_1.assert)(ongoingCall !== undefined, "549462");
                    if (ongoingCall.pr !== pr) {
                        return;
                    }
                    ongoingCall = undefined;
                });
            }
            async function renewTokens_mutexed(params) {
                const { extraTokenParams: extraTokenParams_local } = params;
                const extraTokenParams = {
                    ...getExtraTokenParams?.(),
                    ...extraTokenParams_local
                };
                if (ongoingCall === undefined) {
                    ongoingCall = {
                        pr: renewTokens_nonMutexed({ extraTokenParams }),
                        extraTokenParams
                    };
                    handleThen();
                    return ongoingCall.pr;
                }
                if (JSON.stringify(extraTokenParams) === JSON.stringify(ongoingCall.extraTokenParams)) {
                    return ongoingCall.pr;
                }
                ongoingCall = {
                    pr: (async () => {
                        await ongoingCall.pr;
                        return renewTokens_nonMutexed({ extraTokenParams });
                    })(),
                    extraTokenParams
                };
                handleThen();
                return ongoingCall.pr;
            }
            return params => {
                const { extraTokenParams } = params ?? {};
                prOngoingTokenRenewal = renewTokens_mutexed({ extraTokenParams });
                prOngoingTokenRenewal.then(() => {
                    prOngoingTokenRenewal = undefined;
                });
                return prOngoingTokenRenewal;
            };
        })(),
        subscribeToTokensChange: onTokenChange => {
            onTokenChanges.add(onTokenChange);
            const unsubscribeFromTokensChange = () => {
                onTokenChanges.delete(onTokenChange);
            };
            return {
                unsubscribe: unsubscribeFromTokensChange,
                unsubscribeFromTokensChange
            };
        },
        subscribeToAutoLogoutCountdown: tickCallback => {
            autoLogoutCountdownTickCallbacks.add(tickCallback);
            const unsubscribeFromAutoLogoutCountdown = () => {
                autoLogoutCountdownTickCallbacks.delete(tickCallback);
            };
            return { unsubscribeFromAutoLogoutCountdown };
        },
        goToAuthServer: ({ extraQueryParams, redirectUrl, transformUrlBeforeRedirect }) => loginOrGoToAuthServer({
            action: "go to auth server",
            redirectUrl: redirectUrl ?? window.location.href,
            extraQueryParams_local: extraQueryParams,
            transformUrlBeforeRedirect_local: transformUrlBeforeRedirect
        }),
        backFromAuthServer: resultOfLoginProcess.backFromAuthServer,
        isNewBrowserSession: (() => {
            const value = getIsNewBrowserSession({ subjectId });
            log?.(`isNewBrowserSession: ${value}`);
            return value;
        })()
    });
    if (resultOfLoginProcess.isRestoredFromSessionStorage) {
        const { isOnline, prOnline } = (0, getIsOnline_1.getIsOnline)();
        if (isOnline) {
            await oidc_loggedIn.getTokens();
        }
        else {
            prOnline.then(() => oidc_loggedIn.getTokens());
        }
    }
    {
        const { prOtherTabLogout } = (0, logoutPropagationToOtherTabs_1.getPrOtherTabLogout)({
            configId,
            sessionId
        });
        prOtherTabLogout.then(async () => {
            log?.(`Other tab has logged out, refreshing current tab`);
            await (0, ongoingLoginOrRefreshProcesses_1.waitForAllOtherOngoingLoginOrRefreshProcessesToComplete)({
                prUnlock: new Promise(() => { })
            });
            location.reload();
        });
    }
    // NOTE: Pessimistic renewal of token on potential clock adjustment.
    (function renewOnLocalTimeShift() {
        // NOTE: If we can't confidently silently refresh tokens we won't risk reloading
        // the page just to cover the local time drift edge case.
        if (!currentTokens.hasRefreshToken && !canUseIframe) {
            return;
        }
        const CHECK_DELAY_MS = 5000;
        const DRIFT_THRESHOLD_MS = 1000;
        // NOTE: performance.now() is monotonic and not subject to system clock changes,
        // so the difference between Date.now() and performance.now() lets us detect
        // local clock adjustments even if timers are throttled in background tabs.
        const getClockOffset = () => Date.now() - performance.now();
        let referenceClockOffset = getClockOffset();
        let timer = undefined;
        const scheduleCheck = () => {
            timer = setTimeout(() => {
                const currentClockOffset = getClockOffset();
                if (Math.abs(currentClockOffset - referenceClockOffset) > DRIFT_THRESHOLD_MS) {
                    log?.("Renewing token now as local time might have shifted");
                    oidc_loggedIn.renewTokens();
                    return;
                }
                referenceClockOffset = currentClockOffset;
                scheduleCheck();
            }, CHECK_DELAY_MS);
        };
        scheduleCheck();
        const { unsubscribeFromTokensChange } = oidc_loggedIn.subscribeToTokensChange(() => {
            if (timer !== undefined) {
                clearTimeout(timer);
            }
            unsubscribeFromTokensChange();
            renewOnLocalTimeShift();
        });
    })();
    (function scheduleTokenRefreshToKeepSessionAlive() {
        if (!currentTokens.hasRefreshToken && !canUseIframe) {
            log?.([
                "Session keep-alive disabled: no refresh token and no iframe support. ",
                "Result: once tokens expire, continuing requires full reload."
            ].join(" "));
            return;
        }
        const msBeforeExpiration_idleSessionLifetimeInSeconds = idleSessionLifetimeInSeconds === undefined ? undefined : idleSessionLifetimeInSeconds * 1000;
        const msBeforeExpiration_refreshToken = currentTokens.refreshTokenExpirationTime === undefined
            ? undefined
            : currentTokens.refreshTokenExpirationTime - currentTokens.getServerDateNow();
        const msBeforeExpiration_accessToken = currentTokens.accessTokenExpirationTime - currentTokens.getServerDateNow();
        let isRefreshTokenNeverExpiring = false;
        if (currentTokens.refreshTokenExpirationTime !== undefined &&
            currentTokens.refreshTokenExpirationTime >= INFINITY_TIME_1.INFINITY_TIME) {
            const warningLines = [];
            if (scopes.includes("offline_access")) {
                warningLines.push("offline_access scope was explicitly requested.");
            }
            else if ((0, isKeycloak_1.isKeycloak)({ issuerUri })) {
                warningLines.push("Keycloak likely enabled offline_access by default.");
            }
            if (warningLines.length > 0) {
                warningLines.push(...[
                    "Misconfiguration: offline_access is for native apps, not for web apps like yours. ",
                    "You lose SSO and users must log in after every reload."
                ]);
                console.warn(`oidc-spa: ${warningLines.join(" ")}`);
                return;
            }
            isRefreshTokenNeverExpiring = true;
        }
        const RENEW_MS_BEFORE_EXPIRES = 30000;
        const MIN_ACCEPTABLE_MS_BEFORE_EXPIRATION = RENEW_MS_BEFORE_EXPIRES + 15000;
        detect_session_reached_max_life: {
            if (msBeforeExpiration_refreshToken === undefined) {
                break detect_session_reached_max_life;
            }
            if (msBeforeExpiration_refreshToken > MIN_ACCEPTABLE_MS_BEFORE_EXPIRATION) {
                break detect_session_reached_max_life;
            }
            console.warn([
                "oidc-spa: The session is nearing its maximum lifetime, and the user will soon need to log in again,",
                `or you've configured a refresh_token with a TTL of ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(msBeforeExpiration_refreshToken)}.`,
                `If it's the latter, the TTL is too short, it must be at least ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(MIN_ACCEPTABLE_MS_BEFORE_EXPIRATION)} for reliable operation.`,
                "Shorter lifetimes can cause unpredictable session expirations and are usually a misconfiguration.",
                "\nIn either case, oidc-spa will not ping the auth server to keep the session alive."
            ].join(" "));
            return;
        }
        let msBeforeExpiration = (() => {
            if (msBeforeExpiration_refreshToken !== undefined && !isRefreshTokenNeverExpiring) {
                log?.([
                    (0, toHumanReadableDuration_1.toHumanReadableDuration)(msBeforeExpiration_refreshToken),
                    `before expiration of the refresh_token.`,
                    `Scheduling renewal of the tokens ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(RENEW_MS_BEFORE_EXPIRES)} before expiration as a way to keep the session alive on the OIDC server.`
                ].join(" "));
                return msBeforeExpiration_refreshToken;
            }
            if (msBeforeExpiration_idleSessionLifetimeInSeconds !== undefined) {
                if (msBeforeExpiration_idleSessionLifetimeInSeconds < MIN_ACCEPTABLE_MS_BEFORE_EXPIRATION) {
                    throw new Error([
                        `oidc-spa: The configured idleSessionLifetimeInSeconds (${(0, toHumanReadableDuration_1.toHumanReadableDuration)(msBeforeExpiration_idleSessionLifetimeInSeconds)}) is too short.`,
                        `For reliability, it must be at least ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(MIN_ACCEPTABLE_MS_BEFORE_EXPIRATION)}.`,
                        "Very short session idle lifetimes are usually a misconfiguration, even for ultra sensitive apps."
                    ].join(" "));
                }
                log?.([
                    `You've set idleSessionLifetimeInSeconds to ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(msBeforeExpiration_idleSessionLifetimeInSeconds)}.`,
                    `This means the user session will expire after ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(msBeforeExpiration_idleSessionLifetimeInSeconds)} of inactivity (assuming you're right).`,
                    `Scheduling token renewal ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(RENEW_MS_BEFORE_EXPIRES)} before expiration to keep the session active on the OIDC server.`
                ].join(" "));
                return msBeforeExpiration_idleSessionLifetimeInSeconds;
            }
            const msBeforeExpiration = msBeforeExpiration_accessToken > MIN_ACCEPTABLE_MS_BEFORE_EXPIRATION
                ? msBeforeExpiration_accessToken
                : 3600000;
            log?.([
                "The auth server's idle session timeout is unknown.",
                isRefreshTokenNeverExpiring && "(The refresh token never expires)",
                `Assuming a default idle session TTL of ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(msBeforeExpiration)}.`,
                `Scheduling token renewal ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(RENEW_MS_BEFORE_EXPIRES)} before expiration to keep the session active on the OIDC server.`
            ]
                .filter(line => typeof line === "string")
                .join(" "));
            return msBeforeExpiration;
        })();
        const timer = setTimeout(async () => {
            {
                const { isOnline, prOnline } = (0, getIsOnline_1.getIsOnline)();
                if (!isOnline) {
                    const didCameBackOnlineInTime = await Promise.race([
                        new Promise(resolve => setTimeout(() => resolve(false), RENEW_MS_BEFORE_EXPIRES - 1000)),
                        prOnline.then(() => true)
                    ]);
                    if (!didCameBackOnlineInTime) {
                        log?.([
                            "The session expired on the OIDC server.",
                            "We couldn't keep it alive because the browser was offline.",
                            "We are not redirecting to the login page to support PWAs with offline features.",
                            "However, the next getTokens() call will trigger a redirect to the Auth server login page."
                        ].join(" "));
                        return;
                    }
                }
            }
            log?.(`Renewing the tokens now as otherwise the session will be terminated by the auth server in ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(RENEW_MS_BEFORE_EXPIRES)}`);
            await oidc_loggedIn.renewTokens();
        }, Math.min(msBeforeExpiration - RENEW_MS_BEFORE_EXPIRES, 
        // NOTE: We want to make sure we do not overflow the setTimeout
        // that must be a 32 bit unsigned integer.
        // This can happen if the tokenExpirationTime is more than 24.8 days in the future.
        Math.pow(2, 31) - 1));
        const { unsubscribeFromTokensChange } = oidc_loggedIn.subscribeToTokensChange(() => {
            clearTimeout(timer);
            unsubscribeFromTokensChange();
            scheduleTokenRefreshToKeepSessionAlive();
        });
    })();
    auto_logout: {
        const getCurrentRefreshTokenTtlInSeconds = () => {
            if (currentTokens.refreshTokenExpirationTime === undefined) {
                return idleSessionLifetimeInSeconds;
            }
            if (currentTokens.refreshTokenExpirationTime >= INFINITY_TIME_1.INFINITY_TIME) {
                return idleSessionLifetimeInSeconds ?? 0;
            }
            const ttlInSeconds = (currentTokens.refreshTokenExpirationTime - currentTokens.issuedAtTime) / 1000;
            if (idleSessionLifetimeInSeconds !== undefined) {
                return Math.min(idleSessionLifetimeInSeconds, ttlInSeconds);
            }
            return ttlInSeconds;
        };
        if (getCurrentRefreshTokenTtlInSeconds() === 0) {
            log?.("The refresh_token never expires, disabling auto logout mechanism.");
            break auto_logout;
        }
        if (getCurrentRefreshTokenTtlInSeconds() === undefined) {
            log?.(`${currentTokens.hasRefreshToken
                ? "The refresh token is opaque, we can't read it's expiration time"
                : "No refresh token"}, and idleSessionLifetimeInSeconds was not set, can't implement auto logout mechanism.`);
            break auto_logout;
        }
        const { startCountdown } = (0, startCountdown_1.createStartCountdown)({
            tickCallback: async ({ secondsLeft }) => {
                const invokeAllCallbacks = (params) => {
                    const { secondsLeft } = params;
                    Array.from(autoLogoutCountdownTickCallbacks).forEach(tickCallback => tickCallback({ secondsLeft }));
                };
                if (secondsLeft === 0) {
                    cancel_if_offline: {
                        const { isOnline, prOnline } = (0, getIsOnline_1.getIsOnline)();
                        if (isOnline) {
                            break cancel_if_offline;
                        }
                        const didCameBackOnline = await Promise.race([
                            new Promise(resolve => setTimeout(() => resolve(false), 10000)),
                            prOnline.then(() => true)
                        ]);
                        if (didCameBackOnline) {
                            break cancel_if_offline;
                        }
                        log?.([
                            "Normally now we should auto logout.",
                            "However since the browser is currently offline",
                            "we avoid calling logout() now to play nice in case",
                            "this app is a PWA.",
                            "Next getTokens() is called logout will be called"
                        ].join(" "));
                        unsubscribeFromIsUserActive();
                        invokeAllCallbacks({ secondsLeft: undefined });
                        wouldHaveAutoLoggedOutIfBrowserWasOnline = true;
                        return;
                    }
                    await oidc_loggedIn.logout(autoLogoutParams);
                }
                invokeAllCallbacks({ secondsLeft });
            }
        });
        let stopCountdown = undefined;
        const evtIsUserActive = (0, evtIsUserActive_1.createEvtIsUserActive)({
            configId,
            sessionId
        });
        const { unsubscribe: unsubscribeFromIsUserActive } = evtIsUserActive.subscribe(eventData => {
            if (eventData.isUserActive) {
                if (stopCountdown !== undefined) {
                    stopCountdown();
                    stopCountdown = undefined;
                }
            }
            else {
                (0, assert_1.assert)(stopCountdown === undefined, "902992");
                const currentRefreshTokenTtlInSeconds = getCurrentRefreshTokenTtlInSeconds();
                (0, assert_1.assert)(currentRefreshTokenTtlInSeconds !== undefined, "902992326");
                stopCountdown = startCountdown({
                    countDownFromSeconds: Math.floor((currentRefreshTokenTtlInSeconds * 1000 -
                        eventData.hasBeenInactiveForHowLongMs) /
                        1000)
                }).stopCountdown;
            }
        });
        {
            const currentRefreshTokenTtlInSeconds = getCurrentRefreshTokenTtlInSeconds();
            (0, assert_1.assert)(currentRefreshTokenTtlInSeconds !== undefined, "9029923253");
            log?.([
                `The user will be automatically logged out after ${(0, toHumanReadableDuration_1.toHumanReadableDuration)(currentRefreshTokenTtlInSeconds * 1000)} of inactivity.`,
                idleSessionLifetimeInSeconds === undefined
                    ? undefined
                    : `It was artificially defined by using the idleSessionLifetimeInSeconds param.`
            ]
                .filter(x => x !== undefined)
                .join("\n"));
        }
    }
    return oidc_loggedIn;
}
