// oidc-spa/core/createMockOidc.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createMockOidc = createMockOidc;
const createObjectThatThrowsIfAccessed_1 = require("../tools/createObjectThatThrowsIfAccessed");
const id_1 = require("../tools/tsafe/id");
const toFullyQualifiedUrl_1 = require("../tools/toFullyQualifiedUrl");
const urlSearchParams_1 = require("../tools/urlSearchParams");
const earlyInit_rootRelativeOriginalLocationHref_1 = require("../core/earlyInit_rootRelativeOriginalLocationHref");
const INFINITY_TIME_1 = require("../tools/INFINITY_TIME");
const earlyInit_BASE_URL_1 = require("../core/earlyInit_BASE_URL");
const URL_SEARCH_PARAM_NAME = "isUserLoggedIn";
const locationHref_moduleEvalTime = location.href;
async function createMockOidc(params) {
    const { isUserInitiallyLoggedIn = true, mockedParams = {}, mockedTokens = {}, autoLogin = false, postLoginRedirectUrl } = params;
    const BASE_URL_params = params.BASE_URL;
    const isUserLoggedIn = (() => {
        const { wasPresent, value } = (0, urlSearchParams_1.getSearchParam)({
            url: (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                urlish: (() => {
                    try {
                        return (0, earlyInit_rootRelativeOriginalLocationHref_1.getRootRelativeOriginalLocationHref_earlyInit)();
                    }
                    catch {
                        return locationHref_moduleEvalTime;
                    }
                })(),
                doAssertNoQueryParams: false
            }),
            name: URL_SEARCH_PARAM_NAME
        });
        if (!wasPresent) {
            return isUserInitiallyLoggedIn;
        }
        remove_from_url: {
            const { wasPresent, url_withoutTheParam } = (0, urlSearchParams_1.getSearchParam)({
                url: window.location.href,
                name: URL_SEARCH_PARAM_NAME
            });
            if (!wasPresent) {
                break remove_from_url;
            }
            window.history.replaceState({}, "", url_withoutTheParam);
        }
        return value === "true";
    })();
    const homeUrl = (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
        urlish: BASE_URL_params ?? (0, earlyInit_BASE_URL_1.getBASE_URL_earlyInit)() ?? "/",
        doAssertNoQueryParams: true,
        doOutputWithTrailingSlash: true
    });
    const common = {
        clientId: mockedParams.clientId ?? "mymockclient",
        issuerUri: mockedParams.issuerUri ?? "https://my-mock-oidc-server.net/realms/mymockrealm",
        validRedirectUri: homeUrl
    };
    const loginOrGoToAuthServer = async (params) => {
        const { redirectUrl: redirectUrl_params } = params;
        const redirectUrl = (0, urlSearchParams_1.addOrUpdateSearchParam)({
            url: (() => {
                if (redirectUrl_params === undefined) {
                    return window.location.href;
                }
                return (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                    urlish: redirectUrl_params,
                    doAssertNoQueryParams: false
                });
            })(),
            name: URL_SEARCH_PARAM_NAME,
            value: "true",
            encodeMethod: "www-form"
        });
        window.location.href = redirectUrl;
        return new Promise(() => { });
    };
    if (!isUserLoggedIn) {
        const oidc = (0, id_1.id)({
            ...common,
            isUserLoggedIn: false,
            login: ({ redirectUrl } = {}) => loginOrGoToAuthServer({ redirectUrl }),
            initializationError: undefined
        });
        if (autoLogin) {
            await oidc.login({
                redirectUrl: postLoginRedirectUrl,
                doesCurrentHrefRequiresAuth: true
            });
            // Never here
        }
        // @ts-expect-error: We know what we are doing
        return oidc;
    }
    const oidc = {
        ...common,
        isUserLoggedIn: true,
        renewTokens: async () => { },
        ...(() => {
            const tokens_common = {
                accessToken: mockedTokens.accessToken ?? "mocked-access-token",
                accessTokenExpirationTime: mockedTokens.accessTokenExpirationTime ?? INFINITY_TIME_1.INFINITY_TIME,
                idToken: mockedTokens.idToken ?? "mocked-id-token",
                decodedIdToken: mockedTokens.decodedIdToken ??
                    (0, createObjectThatThrowsIfAccessed_1.createObjectThatThrowsIfAccessed)({
                        debugMessage: [
                            "You haven't provided a mocked decodedIdToken",
                            "See https://docs.oidc-spa.dev/v/v10/integration-guides/usage#mock-adapter"
                        ].join("\n")
                    }),
                decodedIdToken_original: mockedTokens.decodedIdToken_original ??
                    (0, createObjectThatThrowsIfAccessed_1.createObjectThatThrowsIfAccessed)({
                        debugMessage: [
                            "You haven't provided a mocked decodedIdToken_original",
                            "See https://docs.oidc-spa.dev/v/v10/integration-guides/usage#mock-adapter"
                        ].join("\n")
                    }),
                issuedAtTime: Date.now(),
                getServerDateNow: () => Date.now()
            };
            const tokens = mockedTokens.refreshToken !== undefined || mockedTokens.hasRefreshToken === true
                ? (0, id_1.id)({
                    ...tokens_common,
                    hasRefreshToken: true,
                    refreshToken: mockedTokens.refreshToken ?? "mocked-refresh-token",
                    refreshTokenExpirationTime: mockedTokens.refreshTokenExpirationTime
                })
                : (0, id_1.id)({
                    ...tokens_common,
                    hasRefreshToken: false
                });
            return {
                getTokens: () => Promise.resolve(tokens),
                getDecodedIdToken: () => tokens_common.decodedIdToken
            };
        })(),
        subscribeToTokensChange: () => {
            const unsubscribeFromTokensChange = () => { };
            return {
                unsubscribeFromTokensChange,
                unsubscribe: unsubscribeFromTokensChange
            };
        },
        logout: params => {
            const redirectUrl = (0, urlSearchParams_1.addOrUpdateSearchParam)({
                url: (() => {
                    switch (params.redirectTo) {
                        case "current page":
                            return window.location.href;
                        case "home":
                            return homeUrl;
                        case "specific url":
                            return (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                                urlish: params.url,
                                doAssertNoQueryParams: false
                            });
                    }
                })(),
                name: URL_SEARCH_PARAM_NAME,
                value: "false",
                encodeMethod: "www-form"
            });
            window.location.href = redirectUrl;
            return new Promise(() => { });
        },
        subscribeToAutoLogoutCountdown: () => ({
            unsubscribeFromAutoLogoutCountdown: () => { }
        }),
        goToAuthServer: async ({ redirectUrl }) => loginOrGoToAuthServer({ redirectUrl }),
        isNewBrowserSession: false,
        backFromAuthServer: undefined
    };
    return oidc;
}
