// oidc-spa/core/earlyInit.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oidcEarlyInit = oidcEarlyInit;
const StateData_1 = require("./StateData");
const assert_1 = require("../tools/tsafe/assert");
const earlyInit_BASE_URL_1 = require("./earlyInit_BASE_URL");
const isBrowser_1 = require("../tools/isBrowser");
const Evt_1 = require("../tools/Evt");
const earlyInit_rootRelativeOriginalLocationHref_1 = require("./earlyInit_rootRelativeOriginalLocationHref");
const earlyInit_prModuleCreateOidc_1 = require("./earlyInit_prModuleCreateOidc");
const IFRAME_MESSAGE_PREFIX = "oidc-spa:cross-window-messaging:";
let shouldLoadApp = undefined;
function oidcEarlyInit(params) {
    if (shouldLoadApp !== undefined) {
        return { shouldLoadApp };
    }
    shouldLoadApp = oidcEarlyInit_nonMemoized(params).shouldLoadApp;
    return { shouldLoadApp };
}
function oidcEarlyInit_nonMemoized(params) {
    const { BASE_URL, sessionRestorationMethod, securityDefenses = {} } = params ?? {};
    if (!isBrowser_1.isBrowser) {
        return { shouldLoadApp: true };
    }
    const { shouldLoadApp } = handleOidcCallback();
    let exports_earlyInit;
    if (shouldLoadApp) {
        let evtIframeAuthResponse = undefined;
        {
            const _MessageEvent_prototype_data_get = (() => {
                const pd = Object.getOwnPropertyDescriptor(MessageEvent.prototype, "data");
                (0, assert_1.assert)(pd !== undefined);
                const { get } = pd;
                (0, assert_1.assert)(get !== undefined);
                return get;
            })();
            const _MessageEvent_prototype_origin_get = (() => {
                const pd = Object.getOwnPropertyDescriptor(MessageEvent.prototype, "origin");
                (0, assert_1.assert)(pd !== undefined);
                const { get } = pd;
                (0, assert_1.assert)(get !== undefined);
                return get;
            })();
            const _Event_prototype_stopImmediatePropagation_value = Event.prototype.stopImmediatePropagation;
            const origin = window.location.origin;
            window.addEventListener("message", event => {
                if (_MessageEvent_prototype_origin_get.call(event) !== origin) {
                    return;
                }
                const eventData = _MessageEvent_prototype_data_get.call(event);
                if (typeof eventData !== "string") {
                    return;
                }
                if (!eventData.startsWith(IFRAME_MESSAGE_PREFIX)) {
                    return;
                }
                _Event_prototype_stopImmediatePropagation_value.call(event);
                const authResponse = JSON.parse(eventData.slice(IFRAME_MESSAGE_PREFIX.length));
                (evtIframeAuthResponse ?? (evtIframeAuthResponse = (0, Evt_1.createEvt)())).post(authResponse);
            }, {
                capture: true,
                once: false,
                passive: false
            });
        }
        if (BASE_URL !== undefined) {
            (0, earlyInit_BASE_URL_1.setBASE_URL_earlyInit)({ BASE_URL });
        }
        {
            const { enableBrowserRuntimeFreeze, enableDPoP, enableTokenSubstitution } = securityDefenses;
            enableDPoP?.();
            enableTokenSubstitution?.();
            enableBrowserRuntimeFreeze?.();
        }
        exports_earlyInit = {
            shouldLoadApp: true,
            getEvtIframeAuthResponse: () => {
                return (evtIframeAuthResponse ?? (evtIframeAuthResponse = (0, Evt_1.createEvt)()));
            },
            getRedirectAuthResponse: () => {
                return redirectAuthResponse === undefined
                    ? { authResponse: undefined }
                    : {
                        authResponse: redirectAuthResponse,
                        clearAuthResponse: () => {
                            redirectAuthResponse = undefined;
                        }
                    };
            },
            sessionRestorationMethod
        };
    }
    else {
        exports_earlyInit = {
            shouldLoadApp: false
        };
    }
    earlyInit_prModuleCreateOidc_1.prModuleCreateOidc.then(({ registerExports_earlyInit }) => {
        registerExports_earlyInit(exports_earlyInit);
    });
    return { shouldLoadApp };
}
let redirectAuthResponse = undefined;
function handleOidcCallback() {
    const location_urlObj = new URL(window.location.href);
    const locationHrefAssessment = (() => {
        fragment: {
            const stateUrlParamValue = new URLSearchParams(location_urlObj.hash.replace(/^#/, "")).get("state");
            if (stateUrlParamValue === null) {
                break fragment;
            }
            if (!(0, StateData_1.getIsStatQueryParamValue)({ maybeStateUrlParamValue: stateUrlParamValue })) {
                break fragment;
            }
            return { hasAuthResponseInUrl: true, responseMode: "fragment" };
        }
        query: {
            const stateUrlParamValue = location_urlObj.searchParams.get("state");
            if (stateUrlParamValue === null) {
                break query;
            }
            if (!(0, StateData_1.getIsStatQueryParamValue)({ maybeStateUrlParamValue: stateUrlParamValue })) {
                break query;
            }
            if (location_urlObj.searchParams.get("client_id") !== null &&
                location_urlObj.searchParams.get("response_type") !== null &&
                location_urlObj.searchParams.get("redirect_uri") !== null) {
                // NOTE: We are probably in a Keycloakify theme and oidc-spa was loaded by mistake.
                break query;
            }
            return { hasAuthResponseInUrl: true, responseMode: "query" };
        }
        return { hasAuthResponseInUrl: false };
    })();
    if (!locationHrefAssessment.hasAuthResponseInUrl) {
        (0, earlyInit_rootRelativeOriginalLocationHref_1.setGetRootRelativeOriginalLocationHref_earlyInit)({
            rootRelativeOriginalLocationHref: location_urlObj.href.slice(location_urlObj.origin.length)
        });
        return { shouldLoadApp: true };
    }
    (0, earlyInit_rootRelativeOriginalLocationHref_1.setGetRootRelativeOriginalLocationHref_earlyInit)({
        rootRelativeOriginalLocationHref: location_urlObj.pathname
    });
    const { authResponse } = (() => {
        const authResponse = { state: "" };
        const searchParams = (() => {
            switch (locationHrefAssessment.responseMode) {
                case "fragment":
                    return new URLSearchParams(location_urlObj.hash.replace(/^#/, ""));
                case "query":
                    return location_urlObj.searchParams;
                default:
                    (0, assert_1.assert)(false);
            }
        })();
        for (const [key, value] of searchParams) {
            authResponse[key] = value;
        }
        (0, assert_1.assert)(authResponse.state !== "", "063965");
        return { authResponse };
    })();
    const stateData = (0, StateData_1.getStateData)({ stateUrlParamValue: authResponse.state });
    if (stateData === undefined) {
        history.replaceState({}, "", (0, earlyInit_rootRelativeOriginalLocationHref_1.getRootRelativeOriginalLocationHref_earlyInit)());
        return { shouldLoadApp: true };
    }
    switch (stateData.context) {
        case "iframe":
            if (parent !== top) {
                const errorMessage = [
                    "oidc-spa: For security reasons, refusing to post the auth response.",
                    "If you want your app to be framable use sessionRestorationMethod: 'full page redirect'."
                ].join(" ");
                alert(errorMessage);
                throw new Error(errorMessage);
            }
            parent.postMessage(`${IFRAME_MESSAGE_PREFIX}${JSON.stringify(authResponse)}`, location.origin);
            return { shouldLoadApp: false };
        case "redirect": {
            redirectAuthResponse = authResponse;
            const rootRelativeRedirectUrl = (() => {
                if (stateData.action === "login" && authResponse.error === "consent_required") {
                    return stateData.rootRelativeRedirectUrl_consentRequiredCase;
                }
                return stateData.rootRelativeRedirectUrl;
            })();
            history.replaceState({}, "", rootRelativeRedirectUrl);
            return { shouldLoadApp: true };
        }
        default:
            (0, assert_1.assert)(false);
    }
}
