// oidc-spa/core/evtIsUserActive.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEvtIsUserActive = createEvtIsUserActive;
const Evt_1 = require("../tools/Evt");
const subscribeToUserInteraction_1 = require("../tools/subscribeToUserInteraction");
const assert_1 = require("../tools/tsafe/assert");
const id_1 = require("../tools/tsafe/id");
const globalContext = {
    appInstanceId: Math.random().toString(36).slice(2),
    evtIsUserActiveBySessionId: new Map()
};
function createEvtIsUserActive(params) {
    const { configId, sessionId } = params;
    use_existing_instance: {
        if (sessionId === undefined) {
            break use_existing_instance;
        }
        const evtIsUserActive = globalContext.evtIsUserActiveBySessionId.get(sessionId);
        if (evtIsUserActive === undefined) {
            break use_existing_instance;
        }
        return evtIsUserActive;
    }
    const { notifyOtherTabsOfUserInteraction, subscribeToUserInteractionOnOtherTabs } = (() => {
        const channelName = `oidc-spa:user-interaction-tracker:${sessionId ?? configId}`;
        function notifyOtherTabsOfUserInteraction() {
            new BroadcastChannel(channelName).postMessage((0, id_1.id)({
                appInstanceId: globalContext.appInstanceId
            }));
        }
        function subscribeToUserInteractionOnOtherTabs(callback) {
            const channel = new BroadcastChannel(channelName);
            channel.onmessage = ({ data: message }) => {
                (0, assert_1.assert)((0, assert_1.is)(message));
                if (message.appInstanceId === globalContext.appInstanceId) {
                    return;
                }
                callback();
            };
        }
        return { notifyOtherTabsOfUserInteraction, subscribeToUserInteractionOnOtherTabs };
    })();
    const evtIsUserActive = (0, Evt_1.createEvt)();
    let isUserActive = true;
    const scheduleSetInactive = () => {
        const start = performance.now();
        const timer = setTimeout(() => {
            const elapsed = performance.now() - start;
            (0, assert_1.assert)(isUserActive, "011507");
            isUserActive = false;
            evtIsUserActive.post({ isUserActive, hasBeenInactiveForHowLongMs: elapsed });
        }, 5000);
        return () => {
            clearTimeout(timer);
            clearScheduledSetInactive = undefined;
        };
    };
    let clearScheduledSetInactive = scheduleSetInactive();
    const onUserActivity = (params) => {
        const { isInteractionOnCurrentTab } = params;
        clearScheduledSetInactive?.();
        clearScheduledSetInactive = scheduleSetInactive();
        if (isInteractionOnCurrentTab) {
            notifyOtherTabsOfUserInteraction();
        }
        if (!isUserActive) {
            isUserActive = true;
            evtIsUserActive.post({ isUserActive });
        }
    };
    (0, subscribeToUserInteraction_1.subscribeToUserInteraction)({
        throttleMs: 1000,
        callback: () => onUserActivity({ isInteractionOnCurrentTab: true })
    });
    subscribeToUserInteractionOnOtherTabs(() => onUserActivity({ isInteractionOnCurrentTab: false }));
    if (sessionId !== undefined) {
        globalContext.evtIsUserActiveBySessionId.set(sessionId, evtIsUserActive);
    }
    return evtIsUserActive;
}
