// oidc-spa/core/loginOrGoToAuthServer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPrSafelyRestoredFromBfCacheAfterLoginBackNavigationOrInitializationError = getPrSafelyRestoredFromBfCacheAfterLoginBackNavigationOrInitializationError;
exports.createLoginOrGoToAuthServer = createLoginOrGoToAuthServer;
const toFullyQualifiedUrl_1 = require("../tools/toFullyQualifiedUrl");
const assert_1 = require("../tools/tsafe/assert");
const noUndefined_1 = require("../tools/tsafe/noUndefined");
const StatefulEvt_1 = require("../tools/StatefulEvt");
const Deferred_1 = require("../tools/Deferred");
const urlSearchParams_1 = require("../tools/urlSearchParams");
const getIsOnline_1 = require("../tools/getIsOnline");
const StateDataCookie_1 = require("./StateDataCookie");
const globalContext = {
    evtHasLoginBeenCalled: (0, StatefulEvt_1.createStatefulEvt)(() => false)
};
function getPrSafelyRestoredFromBfCacheAfterLoginBackNavigationOrInitializationError() {
    const dOut = new Deferred_1.Deferred();
    const { unsubscribe } = globalContext.evtHasLoginBeenCalled.subscribe(hasLoginBeenCalled => {
        if (!hasLoginBeenCalled) {
            unsubscribe();
            dOut.resolve();
        }
    });
    return dOut.pr;
}
function createLoginOrGoToAuthServer(params) {
    const { configId, oidcClientTsUserManager, transformUrlBeforeRedirect, getExtraQueryParams, getExtraTokenParams, homeUrl, stateUrlParamValue_instance, evtInitializationOutcomeUserNotLoggedIn, log } = params;
    let lastPublicUrl = undefined;
    async function loginOrGoToAuthServer(params) {
        const { redirectUrl: redirectUrl_params, extraQueryParams_local, transformUrlBeforeRedirect_local, ...rest } = params;
        log?.(`Calling loginOrGoToAuthServer ${JSON.stringify(params, null, 2)}`);
        delay_until_online: {
            const { isOnline, prOnline } = (0, getIsOnline_1.getIsOnline)();
            if (isOnline) {
                break delay_until_online;
            }
            log?.("The browser seem offline, waiting to get back a connection before proceeding to login");
            await prOnline;
        }
        login_specific_handling: {
            if (rest.action !== "login") {
                break login_specific_handling;
            }
            if (globalContext.evtHasLoginBeenCalled.current) {
                log?.("login() has already been called, ignoring the call");
                return new Promise(() => { });
            }
            globalContext.evtHasLoginBeenCalled.current = true;
            if (document.visibilityState !== "visible") {
                rest.interaction === "ensure no interaction";
                const dVisible = new Deferred_1.Deferred();
                const onVisible = () => {
                    if (document.visibilityState !== "visible") {
                        return;
                    }
                    document.removeEventListener("visibilitychange", onVisible);
                    dVisible.resolve();
                };
                document.addEventListener("visibilitychange", onVisible);
                await dVisible.pr;
            }
            bf_cache_handling: {
                if (rest.doForceReloadOnBfCache) {
                    const callback = (event) => {
                        if (!event.persisted) {
                            return;
                        }
                        location.reload();
                    };
                    window.addEventListener("pageshow", callback);
                    break bf_cache_handling;
                }
                const callback = (event) => {
                    if (!event.persisted) {
                        return;
                    }
                    window.removeEventListener("pageshow", callback);
                    log?.("We came back from the login pages and the state of the app has been restored");
                    if (rest.doNavigateBackToLastPublicUrlIfTheTheUserNavigateBack) {
                        if (lastPublicUrl !== undefined) {
                            log?.(`Loading last public route: ${lastPublicUrl}`);
                            window.location.href = lastPublicUrl;
                        }
                        else {
                            log?.("We don't know the last public route, navigating back in history");
                            window.history.back();
                        }
                    }
                    else {
                        // NOTE: We know the user is not logged in because login can only be called when not logged in.
                        log?.("The current page doesn't require auth, avoiding reloading the page");
                        globalContext.evtHasLoginBeenCalled.current = false;
                    }
                };
                window.addEventListener("pageshow", callback);
            }
        }
        const redirectUrl = (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
            urlish: redirectUrl_params,
            doAssertNoQueryParams: false
        });
        {
            const redirectUrl_obj = new URL(redirectUrl);
            const redirectUrl_originAndPath = `${redirectUrl_obj.origin}${redirectUrl_obj.pathname}`;
            if (!redirectUrl_originAndPath.replace(/\/?$/, "/").startsWith(homeUrl)) {
                throw new Error([
                    `oidc-spa: redirect target ${redirectUrl_originAndPath} is outside of your application.`,
                    `The homeUrl that you have provided defines the root where your app is hosted: ${homeUrl}.\n`,
                    `This usually means one of the following:\n`,
                    `1) The homeUrl is not set correctly. It must be the actual hosting root (for Vite, typically \`import.meta.env.BASE_URL\`).\n`,
                    `2) You are trying to redirect outside of your application, which is not allowed by OIDC.`
                ].join(" "));
            }
        }
        const rootRelativeRedirectUrl = redirectUrl.slice(window.location.origin.length);
        log?.(`redirectUrl: ${rootRelativeRedirectUrl}`);
        const rootRelativeRedirectUrl_consentRequiredCase = (() => {
            switch (rest.action) {
                case "login":
                    return (lastPublicUrl ?? homeUrl).slice(window.location.origin.length);
                case "go to auth server":
                    return rootRelativeRedirectUrl;
            }
        })();
        (0, StateDataCookie_1.setStateDataCookieIfEnabled)({
            homeUrl,
            stateUrlParamValue_instance,
            stateDataCookie: {
                action: "login",
                rootRelativeRedirectUrl,
                rootRelativeRedirectUrl_consentRequiredCase
            }
        });
        const stateData = {
            context: "redirect",
            rootRelativeRedirectUrl,
            extraQueryParams: {},
            configId,
            action: "login",
            rootRelativeRedirectUrl_consentRequiredCase
        };
        const isSilent = rest.action === "login" && rest.interaction === "ensure no interaction";
        const transformUrl_oidcClientTs = (url) => {
            [
                [
                    getExtraQueryParams,
                    transformUrlBeforeRedirect === undefined
                        ? undefined
                        : (url) => transformUrlBeforeRedirect({
                            isSilent,
                            authorizationUrl: url
                        })
                ],
                [extraQueryParams_local, transformUrlBeforeRedirect_local]
            ].forEach(([extraQueryParamsMaybeGetter, transformUrlBeforeRedirect], i, arr) => {
                const url_before = i !== arr.length - 1 ? undefined : url;
                add_extra_query_params: {
                    if (extraQueryParamsMaybeGetter === undefined) {
                        break add_extra_query_params;
                    }
                    const extraQueryParams = typeof extraQueryParamsMaybeGetter === "function"
                        ? extraQueryParamsMaybeGetter({ isSilent, url })
                        : extraQueryParamsMaybeGetter;
                    for (const [name, value] of Object.entries(extraQueryParams)) {
                        if (value === undefined) {
                            continue;
                        }
                        url = (0, urlSearchParams_1.addOrUpdateSearchParam)({
                            url,
                            name,
                            value,
                            encodeMethod: "www-form"
                        });
                    }
                }
                apply_transform_url: {
                    if (transformUrlBeforeRedirect === undefined) {
                        break apply_transform_url;
                    }
                    url = transformUrlBeforeRedirect(url);
                }
                update_state: {
                    if (url_before === undefined) {
                        break update_state;
                    }
                    const paramValueByName_current = (0, urlSearchParams_1.getAllSearchParams)(url);
                    const paramValueByName_before = (0, urlSearchParams_1.getAllSearchParams)(url_before);
                    for (const [name, value_current] of Object.entries(paramValueByName_current)) {
                        const value_before = paramValueByName_before[name];
                        if (value_before === value_current) {
                            continue;
                        }
                        stateData.extraQueryParams[name] = value_current;
                    }
                }
            });
            return url;
        };
        const redirectMethod = (() => {
            switch (rest.action) {
                case "login":
                    return rest.doNavigateBackToLastPublicUrlIfTheTheUserNavigateBack
                        ? "replace"
                        : "assign";
                case "go to auth server":
                    return "assign";
            }
        })();
        log?.(`redirectMethod: ${redirectMethod}`);
        if (rest.action === "login") {
            rest.preRedirectHook?.();
        }
        return oidcClientTsUserManager
            .signinRedirect({
            state: stateData,
            redirectMethod,
            prompt: (() => {
                switch (rest.action) {
                    case "go to auth server":
                        return undefined;
                    case "login":
                        switch (rest.interaction) {
                            case "ensure no interaction":
                                return "none";
                            case "ensure interaction":
                                return "login";
                            case "directly redirect if active session show login otherwise":
                                return undefined;
                        }
                        assert_1.assert;
                }
                assert_1.assert;
            })(),
            transformUrl: transformUrl_oidcClientTs,
            extraTokenParams: getExtraTokenParams === undefined ? undefined : (0, noUndefined_1.noUndefined)(getExtraTokenParams())
        })
            .then(() => new Promise(() => { }), error => {
            (0, assert_1.assert)(error instanceof Error, "393430");
            (0, assert_1.assert)(false, `This is a bug in oidc-spa (loginOrGoToAuthServer), please report: ${error.message}`);
        });
    }
    const { unsubscribe } = evtInitializationOutcomeUserNotLoggedIn.subscribe(() => {
        unsubscribe();
        const realPushState = history.pushState.bind(history);
        history.pushState = function pushState(...args) {
            lastPublicUrl = window.location.href;
            return realPushState(...args);
        };
    });
    return {
        loginOrGoToAuthServer
    };
}
