// oidc-spa/core/oidcClientTsUserToTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOidcClientTsUserToTokens = createOidcClientTsUserToTokens;
const assert_1 = require("../tools/tsafe/assert");
const id_1 = require("../tools/tsafe/id");
const readExpirationTimeInJwt_1 = require("../tools/readExpirationTimeInJwt");
const decodeJwt_1 = require("../tools/decodeJwt");
const INFINITY_TIME_1 = require("../tools/INFINITY_TIME");
const getServerDateNow_1 = require("../tools/getServerDateNow");
function createOidcClientTsUserToTokens(params) {
    const { configId, decodedIdTokenSchema, __unsafe_useIdTokenAsAccessToken, exports_DPoP, exports_tokenSubstitution, log } = params;
    function oidcClientTsUserToTokens(params) {
        const { oidcClientTsUser, decodedIdToken_previous } = params;
        const isFirstInit = decodedIdToken_previous === undefined;
        const accessToken = oidcClientTsUser.access_token;
        const refreshToken = oidcClientTsUser.refresh_token;
        const idToken = oidcClientTsUser.id_token;
        (0, assert_1.assert)(idToken !== undefined, "No id token provided by the oidc server");
        const decodedIdToken_original = (0, decodeJwt_1.decodeJwt)(idToken);
        if (isFirstInit) {
            log?.([
                `Decoded ID token`,
                decodedIdTokenSchema === undefined ? "" : " before `decodedIdTokenSchema.parse()`\n",
                JSON.stringify(decodedIdToken_original, null, 2)
            ].join(""));
        }
        const decodedIdToken = (() => {
            let decodedIdToken;
            if (decodedIdTokenSchema !== undefined) {
                decodedIdToken = decodedIdTokenSchema.parse(decodedIdToken_original);
                if (isFirstInit) {
                    log?.([
                        "Decoded ID token after `decodedIdTokenSchema.parse()`\n",
                        JSON.stringify(decodedIdToken, null, 2)
                    ].join(""));
                }
            }
            else {
                // @ts-expect-error
                decodedIdToken = decodedIdToken_original;
            }
            if (decodedIdToken_previous !== undefined &&
                JSON.stringify(decodedIdToken) === JSON.stringify(decodedIdToken_previous)) {
                // NOTE: For stable ref, prevent re-render for component that would memoize
                return decodedIdToken_previous;
            }
            return decodedIdToken;
        })();
        const issuedAtTime = (() => {
            let id_token_iat = (() => {
                let iat;
                try {
                    const iat_claimValue = toNumber(decodedIdToken_original.iat);
                    iat = iat_claimValue;
                }
                catch {
                    iat = undefined;
                }
                if (iat === undefined) {
                    return undefined;
                }
                return iat;
            })();
            if (id_token_iat === undefined) {
                return oidcClientTsUser.__oidc_spa_localTimeWhenTokenIssued;
            }
            correct_entra_builtin_skew: {
                // See: https://github.com/keycloakify/oidc-spa/issues/162
                const expires_in = toNumber(oidcClientTsUser.__oidc_spa_tokenResponse["expires_in"]);
                if (expires_in === undefined) {
                    break correct_entra_builtin_skew;
                }
                let access_token_iat;
                let access_token_exp;
                try {
                    const decodedAccessToken = (0, decodeJwt_1.decodeJwt)(accessToken);
                    (0, assert_1.assert)(decodedAccessToken instanceof Object);
                    const iat = toNumber(decodedAccessToken["iat"]);
                    const exp = toNumber(decodedAccessToken["exp"]);
                    (0, assert_1.assert)(iat !== undefined);
                    (0, assert_1.assert)(exp !== undefined);
                    access_token_iat = iat;
                    access_token_exp = exp;
                }
                catch {
                    break correct_entra_builtin_skew;
                }
                const access_token_expires_in = access_token_exp - access_token_iat;
                const builtin_skew_sec = access_token_expires_in - expires_in;
                id_token_iat += builtin_skew_sec;
            }
            return id_token_iat * 1000;
        })();
        const paramsOfCreateGetServerDateNow = {
            issuedAtTime_local: oidcClientTsUser.__oidc_spa_localTimeWhenTokenIssued,
            issuedAtTime
        };
        const tokens_common = {
            ...(__unsafe_useIdTokenAsAccessToken
                ? {
                    accessToken: idToken,
                    accessTokenExpirationTime: (() => {
                        const expirationTime = (0, readExpirationTimeInJwt_1.readExpirationTimeInJwt)(idToken);
                        (0, assert_1.assert)(expirationTime !== undefined, "Failed to get id token expiration time while trying to substitute the access token by the id token");
                        return expirationTime;
                    })()
                }
                : {
                    accessToken,
                    accessTokenExpirationTime: (() => {
                        read_from_jwt: {
                            const expirationTime = (0, readExpirationTimeInJwt_1.readExpirationTimeInJwt)(accessToken);
                            if (expirationTime === undefined) {
                                break read_from_jwt;
                            }
                            return expirationTime;
                        }
                        read_from_token_response_expires_at: {
                            const expires_at = toNumber(oidcClientTsUser.__oidc_spa_tokenResponse["expires_at"]);
                            if (expires_at === undefined) {
                                break read_from_token_response_expires_at;
                            }
                            return expires_at * 1000;
                        }
                        read_from_token_response_expires_in: {
                            const expires_in = toNumber(oidcClientTsUser.__oidc_spa_tokenResponse["expires_in"]);
                            if (expires_in === undefined) {
                                break read_from_token_response_expires_in;
                            }
                            return issuedAtTime + expires_in * 1000;
                        }
                        (0, assert_1.assert)(false, "Failed to get access token expiration time");
                    })()
                }),
            idToken,
            decodedIdToken,
            decodedIdToken_original,
            issuedAtTime,
            getServerDateNow: (0, getServerDateNow_1.createGetServerDateNow)(paramsOfCreateGetServerDateNow)
        };
        const tokens = refreshToken === undefined
            ? (0, id_1.id)({
                ...tokens_common,
                hasRefreshToken: false
            })
            : (0, id_1.id)({
                ...tokens_common,
                hasRefreshToken: true,
                refreshToken,
                refreshTokenExpirationTime: (() => {
                    for (const propertyName of [
                        "refresh_expires_at",
                        "refresh_token_expires_at"
                    ]) {
                        const expiresAt = toNumber(oidcClientTsUser.__oidc_spa_tokenResponse[propertyName]);
                        if (expiresAt === undefined) {
                            continue;
                        }
                        if (expiresAt === 0) {
                            return INFINITY_TIME_1.INFINITY_TIME;
                        }
                        return expiresAt * 1000;
                    }
                    for (const propertyName of [
                        "refresh_expires_in",
                        "refresh_token_expires_in"
                    ]) {
                        const expiresIn = toNumber(oidcClientTsUser.__oidc_spa_tokenResponse[propertyName]);
                        if (expiresIn === undefined) {
                            continue;
                        }
                        if (expiresIn === 0) {
                            return INFINITY_TIME_1.INFINITY_TIME;
                        }
                        return issuedAtTime + expiresIn * 1000;
                    }
                    read_from_jwt: {
                        const expirationTime = (0, readExpirationTimeInJwt_1.readExpirationTimeInJwt)(refreshToken);
                        if (expirationTime === undefined) {
                            break read_from_jwt;
                        }
                        return expirationTime;
                    }
                    return undefined;
                })()
            });
        if (exports_DPoP !== undefined) {
            exports_DPoP.registerAccessTokenForDPoP({
                configId,
                accessToken: tokens.accessToken,
                paramsOfCreateGetServerDateNow
            });
        }
        if (exports_tokenSubstitution !== undefined) {
            const placeholders = exports_tokenSubstitution.getTokensPlaceholders({
                configId,
                tokens
            });
            tokens.accessToken = placeholders.accessToken;
            tokens.idToken = placeholders.idToken;
            tokens.refreshToken = placeholders.refreshToken;
        }
        if (isFirstInit &&
            tokens.hasRefreshToken &&
            tokens.refreshTokenExpirationTime !== undefined &&
            tokens.refreshTokenExpirationTime < tokens.accessTokenExpirationTime) {
            console.warn([
                "The OIDC refresh token expirationTime is shorter than the one of the access token.",
                "This is very unusual and probably a misconfiguration."
            ].join(" "));
        }
        return tokens;
    }
    return { oidcClientTsUserToTokens };
}
function toNumber(v) {
    if (v === undefined || v === null) {
        return undefined;
    }
    if (typeof v === "string") {
        const v_n = parseFloat(v);
        (0, assert_1.assert)(!isNaN(v_n), `3922033 ${v}`);
        return v_n;
    }
    (0, assert_1.assert)(typeof v === "number", `2932202 ${v}`);
    return v;
}
