// oidc-spa/core/StateData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.STATE_STORE_KEY_PREFIX = void 0;
exports.generateStateUrlParamValue = generateStateUrlParamValue;
exports.getIsStatQueryParamValue = getIsStatQueryParamValue;
exports.clearStateStore = clearStateStore;
exports.getStateData = getStateData;
const assert_1 = require("../tools/tsafe/assert");
const typeGuard_1 = require("../tools/tsafe/typeGuard");
const generateUrlSafeRandom_1 = require("../tools/generateUrlSafeRandom");
const STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX = "b2lkYy1zcGEu";
const RANDOM_STRING_LENGTH = 32 - STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX.length;
function generateStateUrlParamValue() {
    return `${STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX}${(0, generateUrlSafeRandom_1.generateUrlSafeRandom)({
        length: RANDOM_STRING_LENGTH
    })}`;
}
function getIsStatQueryParamValue(params) {
    const { maybeStateUrlParamValue } = params;
    return (maybeStateUrlParamValue.startsWith(STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX) &&
        maybeStateUrlParamValue.length ===
            STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX.length + RANDOM_STRING_LENGTH);
}
exports.STATE_STORE_KEY_PREFIX = "oidc.";
function getKey(params) {
    const { stateUrlParamValue } = params;
    return `${exports.STATE_STORE_KEY_PREFIX}${stateUrlParamValue}`;
}
function getStateStore(params) {
    const { stateUrlParamValue } = params;
    const item = localStorage.getItem(getKey({ stateUrlParamValue }));
    if (item === null) {
        return undefined;
    }
    const obj = JSON.parse(item);
    (0, assert_1.assert)((0, typeGuard_1.typeGuard)(obj, obj instanceof Object && obj.data instanceof Object && typeof obj.data.context === "string"));
    return obj;
}
function clearStateStore(params) {
    const { stateUrlParamValue } = params;
    localStorage.removeItem(getKey({ stateUrlParamValue }));
}
function getStateData(params) {
    const { stateUrlParamValue } = params;
    const stateStore = getStateStore({ stateUrlParamValue });
    if (stateStore === undefined) {
        return undefined;
    }
    return stateStore.data;
}
