// oidc-spa/server/extractRequestAuthContext.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AnyRequest = void 0;
exports.extractRequestAuthContext = extractRequestAuthContext;
const tsafe_1 = require("../vendor/server/tsafe");
var AnyRequest;
(function (AnyRequest) {
    tsafe_1.assert;
    tsafe_1.assert;
    tsafe_1.assert;
    tsafe_1.assert;
})(AnyRequest || (exports.AnyRequest = AnyRequest = {}));
function isUnified(req) {
    if ("type" in req && req.type === "unified") {
        tsafe_1.assert;
        return true;
    }
    return false;
}
function isRequest(req) {
    if ("type" in req || "raw" in req || "socket" in req) {
        return false;
    }
    tsafe_1.assert;
    return true;
}
function isIncomingMessage(req) {
    if ("socket" in req) {
        tsafe_1.assert;
        return true;
    }
    return false;
}
function isHonoRequest(req) {
    if ("raw" in req && isRequest(req.raw)) {
        //assert<Equals<typeof req, AnyRequest.RequestLike>>;
        return true;
    }
    return false;
}
function isFastifyRequest(req) {
    if ("raw" in req && isIncomingMessage(req.raw)) {
        //assert<Equals<typeof req, AnyRequest.FastifyRequestLike>>;
        return true;
    }
    return false;
}
function anyRequestToUnified(req) {
    if (isUnified(req)) {
        return req;
    }
    if (isRequest(req)) {
        const url = new URL(req.url);
        return {
            type: "unified",
            method: req.method,
            pseudoHeaders: {
                ":authority": url.host,
                ":path": `${url.pathname}${url.search}`,
                ":scheme": url.protocol.replace(/:$/, "")
            },
            headers: (0, tsafe_1.objectFromEntries)([
                "Authorization",
                "DPoP",
                "Forwarded",
                "X-Forwarded-Proto",
                "X-Forwarded-Host"
            ].map(name => [name, req.headers.get(name) ?? undefined]))
        };
    }
    if (isIncomingMessage(req)) {
        const getHeaderValue = (name) => {
            const value = req.headers[name.toLowerCase()];
            if (typeof value === "string" && value.length > 0) {
                return value.split(",")[0].trim();
            }
        };
        if (req.method === undefined || req.url === undefined) {
            throw new Error(`oidc-spa: Request not ready`);
        }
        return {
            type: "unified",
            method: req.method,
            pseudoHeaders: {
                ":scheme": getHeaderValue(":scheme") || (req.socket.encrypted === true ? "https" : "http"),
                ":authority": getHeaderValue(":authority") || getHeaderValue("host") || "localhost",
                ":path": req.url
            },
            headers: (0, tsafe_1.objectFromEntries)([
                "Authorization",
                "DPoP",
                "Forwarded",
                "X-Forwarded-Proto",
                "X-Forwarded-Host"
            ].map(name => [name, getHeaderValue(name)]))
        };
    }
    if (isHonoRequest(req) || isFastifyRequest(req)) {
        return anyRequestToUnified(req.raw);
    }
    (0, tsafe_1.assert)(false, [
        "oidc-spa: The request object provided wasn't parsed correctly",
        "this is a bug in oidc-spa, please open an issue about it",
        "https://github.com/keycloakify/oidc-spa"
    ].join(" "));
}
function extractRequestAuthContext(params) {
    const { request, trustProxy } = params;
    const request_unified = anyRequestToUnified(request);
    if (request_unified.headers.Authorization === undefined) {
        return undefined;
    }
    const match = request_unified.headers.Authorization.trim().match(/^((?:Bearer)|(?:DPoP))\s+(.+)$/);
    if (match === null) {
        return (0, tsafe_1.id)({
            isWellFormed: false,
            debugErrorMessage: "Malformed Authorization header"
        });
    }
    const [, scheme, accessToken] = match;
    if (!(0, tsafe_1.isAmong)(["Bearer", "DPoP"], scheme)) {
        return (0, tsafe_1.id)({
            isWellFormed: false,
            debugErrorMessage: `Unsupported scheme ${scheme}, expected Bearer or DPoP`
        });
    }
    if (scheme === "Bearer") {
        return (0, tsafe_1.id)({
            isWellFormed: true,
            accessTokenAndMetadata: (0, tsafe_1.id)({
                scheme: "Bearer",
                accessToken,
                rejectIfAccessTokenDPoPBound: true
            })
        });
    }
    tsafe_1.assert;
    if (request_unified.headers.DPoP === undefined) {
        return (0, tsafe_1.id)({
            isWellFormed: false,
            debugErrorMessage: "Scheme DPoP was specified but the DPoP header is missing"
        });
    }
    const expectedHtu = (() => {
        const { pseudoHeaders: ps } = request_unified;
        const url = new URL(`${ps[":scheme"]}://${ps[":authority"]}${ps[":path"]}`);
        read_proxy_forwarded_values: {
            if (!trustProxy) {
                break read_proxy_forwarded_values;
            }
            const forwardedParams = (() => {
                // Reverse proxies may terminate TLS; honor forwarded headers to rebuild the externally visible URL.
                const forwardedHeader = request_unified.headers["Forwarded"];
                if (forwardedHeader === undefined) {
                    return { proto: undefined, host: undefined };
                }
                const [firstEntry] = forwardedHeader.split(",");
                const tokens = firstEntry.trim().split(";");
                const getTokenValue = (name) => {
                    const token = tokens.find(token => token.trim().toLowerCase().startsWith(`${name.toLowerCase()}=`));
                    if (token === undefined) {
                        return undefined;
                    }
                    const [, rawValue] = token.split("=");
                    return rawValue?.replace(/^"|"$/g, "").trim();
                };
                return {
                    proto: getTokenValue("proto"),
                    host: getTokenValue("host")
                };
            })();
            {
                const forwardedProto = (() => {
                    if (forwardedParams.proto) {
                        return forwardedParams.proto;
                    }
                    const value = request_unified.headers["X-Forwarded-Proto"];
                    if (value === undefined) {
                        return undefined;
                    }
                    return value.split(",")[0]?.trim();
                })();
                const forwardedProto_normalized = (() => {
                    if (forwardedProto === undefined) {
                        return undefined;
                    }
                    const proto = forwardedProto.replace(/:$/, "").toLowerCase();
                    return proto === "http" || proto === "https" ? proto : undefined;
                })();
                if (forwardedProto_normalized !== undefined) {
                    url.protocol = `${forwardedProto_normalized}:`;
                }
            }
            {
                const forwardedHost = (() => {
                    if (forwardedParams.host) {
                        return forwardedParams.host;
                    }
                    const value = request_unified.headers["X-Forwarded-Host"];
                    if (!value) {
                        return undefined;
                    }
                    return value.split(",")[0]?.trim();
                })();
                if (forwardedHost) {
                    url.host = forwardedHost;
                }
            }
        }
        return `${url.origin}${url.pathname}`;
    })();
    return (0, tsafe_1.id)({
        isWellFormed: true,
        accessTokenAndMetadata: (0, tsafe_1.id)({
            scheme: "DPoP",
            accessToken,
            dpopProof: request_unified.headers.DPoP,
            expectedHtu,
            expectedHtm: request_unified.method
        })
    });
}
