// oidc-spa/esm/vite-plugin/vite-plugin.mjs
import { assert } from "../tools/tsafe/assert.mjs";
import { createHandleClientEntrypoint } from "./handleClientEntrypoint.mjs";
import { createHandleServerEntrypoint } from "./handleServerEntrypoint.mjs";
import { manageOptimizedDeps } from "./manageOptimizedDeps.mjs";
import { transformCreateFileRoute } from "./transformTanstackRouterCreateFileRoute.mjs";
import { getProjectType } from "./projectType.mjs";
export function oidcSpa(params = {}) {
    let load_handleClientEntrypoint = undefined;
    let load_handleServerEntrypoint = undefined;
    let projectType = undefined;
    const plugin = {
        name: "oidc-spa",
        enforce: "pre",
        config(userConfig) {
            const projectType = getProjectType({
                pluginNames: userConfig.plugins
                    ?.flat()
                    .filter(plugin => plugin instanceof Object)
                    .filter(plugin => "name" in plugin)
                    .map(plugin => plugin.name) ?? []
            });
            userConfig = manageOptimizedDeps({ userConfig, projectType });
            return userConfig;
        },
        configResolved(resolvedConfig) {
            projectType = getProjectType({
                pluginNames: resolvedConfig.plugins.map(({ name }) => name)
            });
            load_handleClientEntrypoint = createHandleClientEntrypoint({
                oidcSpaVitePluginParams: params,
                resolvedConfig,
                projectType
            }).load_handleClientEntrypoint;
            load_handleServerEntrypoint = createHandleServerEntrypoint({
                resolvedConfig,
                projectType
            }).load_handleServerEntrypoint;
        },
        transform(code, id) {
            let transformed = null;
            assert(projectType !== undefined);
            tanstack_start_specific_transformations: {
                if (projectType !== "tanstack-start") {
                    break tanstack_start_specific_transformations;
                }
                transformed = transformCreateFileRoute({
                    code,
                    id
                });
            }
            return transformed;
        },
        async load(id) {
            {
                assert(load_handleClientEntrypoint !== undefined);
                const r = await load_handleClientEntrypoint({
                    id,
                    pluginContext: this
                });
                if (r !== null) {
                    return r;
                }
            }
            {
                assert(load_handleServerEntrypoint !== undefined);
                const r = await load_handleServerEntrypoint({
                    id,
                    pluginContext: this
                });
                if (r !== null) {
                    return r;
                }
            }
            return null;
        }
    };
    return plugin;
}
