// oidc-spa/react-spa/createOidcSpaUtils.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || (function () {
    var ownKeys = function(o) {
        ownKeys = Object.getOwnPropertyNames || function (o) {
            var ar = [];
            for (var k in o) if (Object.prototype.hasOwnProperty.call(o, k)) ar[ar.length] = k;
            return ar;
        };
        return ownKeys(o);
    };
    return function (mod) {
        if (mod && mod.__esModule) return mod;
        var result = {};
        if (mod != null) for (var k = ownKeys(mod), i = 0; i < k.length; i++) if (k[i] !== "default") __createBinding(result, mod, k[i]);
        __setModuleDefault(result, mod);
        return result;
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.createOidcSpaUtils = createOidcSpaUtils;
const react_1 = require("react");
const OidcInitializationError_1 = require("../core/OidcInitializationError");
const Deferred_1 = require("../tools/Deferred");
const isBrowser_1 = require("../tools/isBrowser");
const assert_1 = require("../tools/tsafe/assert");
const createObjectThatThrowsIfAccessed_1 = require("../tools/createObjectThatThrowsIfAccessed");
const StatefulEvt_1 = require("../tools/StatefulEvt");
const id_1 = require("../tools/tsafe/id");
const toFullyQualifiedUrl_1 = require("../tools/toFullyQualifiedUrl");
const desiredPostLoginRedirectUrl_1 = require("../core/desiredPostLoginRedirectUrl");
function createOidcSpaUtils(params) {
    const { autoLogin, decodedIdTokenSchema, decodedIdToken_mock } = params;
    const dParamsOfBootstrap = new Deferred_1.Deferred();
    const dOidcCoreOrInitializationError = new Deferred_1.Deferred();
    const evtAutoLogoutState = (0, StatefulEvt_1.createStatefulEvt)(() => ({
        shouldDisplayWarning: false
    }));
    dOidcCoreOrInitializationError.pr.then(oidcCoreOrInitializationError => {
        const { hasResolved, value: paramsOfBootstrap } = dParamsOfBootstrap.getState();
        (0, assert_1.assert)(hasResolved);
        if (paramsOfBootstrap.implementation === "mock") {
            return;
        }
        assert_1.assert;
        const { warnUserSecondsBeforeAutoLogout = 60 } = paramsOfBootstrap;
        if (oidcCoreOrInitializationError === undefined ||
            oidcCoreOrInitializationError instanceof OidcInitializationError_1.OidcInitializationError) {
            return;
        }
        const oidcCore = oidcCoreOrInitializationError;
        if (!oidcCore.isUserLoggedIn) {
            return;
        }
        oidcCore.subscribeToAutoLogoutCountdown(({ secondsLeft }) => {
            const newState = (() => {
                if (secondsLeft === undefined) {
                    return {
                        shouldDisplayWarning: false
                    };
                }
                if (secondsLeft > warnUserSecondsBeforeAutoLogout) {
                    return {
                        shouldDisplayWarning: false
                    };
                }
                return {
                    shouldDisplayWarning: true,
                    secondsLeftBeforeAutoLogout: secondsLeft
                };
            })();
            if (!newState.shouldDisplayWarning && !evtAutoLogoutState.current.shouldDisplayWarning) {
                return;
            }
            evtAutoLogoutState.current = newState;
        });
    });
    function useOidc(params) {
        const { assert: assert_params } = params ?? {};
        if (!isBrowser_1.isBrowser) {
            throw new Error([
                "oidc-spa: useOidc() can't be used on the server.",
                "You can prevent this component from rendering on the server",
                "by wrapping it into <OidcInitializationGate />"
            ].join(" "));
        }
        const { hasResolved, value: oidcCore } = dOidcCoreOrInitializationError.getState();
        if (!hasResolved) {
            throw dOidcCoreOrInitializationError.pr;
        }
        if (oidcCore instanceof OidcInitializationError_1.OidcInitializationError) {
            throw oidcCore;
        }
        check_assertion: {
            if (assert_params === undefined) {
                break check_assertion;
            }
            const getMessage = (v) => [
                "oidc-spa: There is a logic error in the application.",
                `If this component is mounted the user is supposed ${v}.`,
                "An explicit assertion was made in this sense."
            ].join(" ");
            switch (assert_params) {
                case "user logged in":
                    if (!oidcCore.isUserLoggedIn) {
                        throw new Error(getMessage("to be logged in but currently they arn't"));
                    }
                    break;
                case "user not logged in":
                    if (oidcCore.isUserLoggedIn) {
                        throw new Error(getMessage("not to be logged in but currently they are"));
                    }
                    break;
                default:
                    (0, assert_1.assert)(false);
            }
        }
        const [, reRenderIfDecodedIdTokenChanged] = (0, react_1.useState)(() => {
            if (!oidcCore.isUserLoggedIn) {
                return undefined;
            }
            return oidcCore.getDecodedIdToken();
        });
        const [evtIsDecodedIdTokenUsed] = (0, react_1.useState)(() => (0, StatefulEvt_1.createStatefulEvt)(() => false));
        (0, react_1.useEffect)(() => {
            if (!oidcCore.isUserLoggedIn) {
                return;
            }
            let isActive = true;
            let unsubscribe = undefined;
            (async () => {
                if (!evtIsDecodedIdTokenUsed.current) {
                    const dDecodedIdTokenUsed = new Deferred_1.Deferred();
                    const { unsubscribe: unsubscribe_scope } = evtIsDecodedIdTokenUsed.subscribe(() => {
                        unsubscribe_scope();
                        dDecodedIdTokenUsed.resolve();
                    });
                    unsubscribe = unsubscribe_scope;
                    await dDecodedIdTokenUsed.pr;
                    if (!isActive) {
                        return;
                    }
                }
                reRenderIfDecodedIdTokenChanged(oidcCore.getDecodedIdToken());
                unsubscribe = oidcCore.subscribeToTokensChange(() => {
                    reRenderIfDecodedIdTokenChanged(oidcCore.getDecodedIdToken());
                }).unsubscribeFromTokensChange;
            })();
            return () => {
                isActive = false;
                unsubscribe?.();
            };
        }, []);
        const [evtIsAutoLogoutStateUsed] = (0, react_1.useState)(() => (0, StatefulEvt_1.createStatefulEvt)(() => false));
        const [, reRenderIfAutoLogoutStateChanged] = (0, react_1.useState)(() => evtAutoLogoutState.current);
        (0, react_1.useEffect)(() => {
            let isActive = true;
            let unsubscribe = undefined;
            (async () => {
                if (!evtIsAutoLogoutStateUsed.current) {
                    const dAutoLogoutStateUsed = new Deferred_1.Deferred();
                    const { unsubscribe: unsubscribe_scope } = evtIsAutoLogoutStateUsed.subscribe(() => {
                        unsubscribe_scope();
                        dAutoLogoutStateUsed.resolve();
                    });
                    unsubscribe = unsubscribe_scope;
                    await dAutoLogoutStateUsed.pr;
                    if (!isActive) {
                        return;
                    }
                }
                reRenderIfAutoLogoutStateChanged(evtAutoLogoutState.current);
                unsubscribe = evtAutoLogoutState.subscribe(reRenderIfAutoLogoutStateChanged).unsubscribe;
            })();
            return () => {
                isActive = false;
                unsubscribe?.();
            };
        }, []);
        if (!oidcCore.isUserLoggedIn) {
            return (0, id_1.id)({
                isUserLoggedIn: false,
                initializationError: oidcCore.initializationError,
                issuerUri: oidcCore.issuerUri,
                clientId: oidcCore.clientId,
                validRedirectUri: oidcCore.validRedirectUri,
                autoLogoutState: { shouldDisplayWarning: false },
                login: params => oidcCore.login({
                    doesCurrentHrefRequiresAuth: false,
                    ...params
                })
            });
        }
        return (0, id_1.id)({
            isUserLoggedIn: true,
            get decodedIdToken() {
                evtIsDecodedIdTokenUsed.current = true;
                return oidcCore.getDecodedIdToken();
            },
            logout: oidcCore.logout,
            renewTokens: oidcCore.renewTokens,
            goToAuthServer: oidcCore.goToAuthServer,
            backFromAuthServer: oidcCore.backFromAuthServer,
            isNewBrowserSession: oidcCore.isNewBrowserSession,
            get autoLogoutState() {
                evtIsAutoLogoutStateUsed.current = true;
                return evtAutoLogoutState.current;
            },
            issuerUri: oidcCore.issuerUri,
            clientId: oidcCore.clientId,
            validRedirectUri: oidcCore.validRedirectUri
        });
    }
    async function getOidc(params) {
        if (!isBrowser_1.isBrowser) {
            throw new Error("oidc-spa: getOidc() can't be used on the server");
        }
        const oidcCore = await dOidcCoreOrInitializationError.pr;
        if (oidcCore instanceof OidcInitializationError_1.OidcInitializationError) {
            return new Promise(() => { });
        }
        if (params?.assert === "user logged in" && !oidcCore.isUserLoggedIn) {
            throw new Error([
                "oidc-spa: Called getOidc({ assert: 'user logged in' })",
                "but the user is not currently logged in."
            ].join(" "));
        }
        if (params?.assert === "user not logged in" && oidcCore.isUserLoggedIn) {
            throw new Error([
                "oidc-spa: Called getOidc({ assert: 'user not logged in' })",
                "but the user is currently logged in."
            ].join(" "));
        }
        const common = {
            issuerUri: oidcCore.issuerUri,
            clientId: oidcCore.clientId,
            validRedirectUri: oidcCore.validRedirectUri
        };
        return oidcCore.isUserLoggedIn
            ? (0, id_1.id)({
                ...common,
                isUserLoggedIn: true,
                getAccessToken: async () => {
                    const { accessToken } = await oidcCore.getTokens();
                    return accessToken;
                },
                subscribeToAccessTokenRotation: next => {
                    const { unsubscribeFromTokensChange } = oidcCore.subscribeToTokensChange(({ accessToken }) => {
                        next(accessToken);
                    });
                    return { unsubscribeFromAccessTokenRotation: unsubscribeFromTokensChange };
                },
                getDecodedIdToken: oidcCore.getDecodedIdToken,
                subscribeToDecodedIdTokenChange: next => {
                    const current = oidcCore.getDecodedIdToken();
                    const { unsubscribeFromTokensChange } = oidcCore.subscribeToTokensChange(({ decodedIdToken }) => {
                        // NOTE: oidc-spa/core keeps the reference stable
                        // when structure hasn't changed.
                        if (current !== decodedIdToken) {
                            next(decodedIdToken);
                        }
                    });
                    return { unsubscribeFromDecodedIdTokenChange: unsubscribeFromTokensChange };
                },
                logout: oidcCore.logout,
                renewTokens: oidcCore.renewTokens,
                goToAuthServer: oidcCore.goToAuthServer,
                backFromAuthServer: oidcCore.backFromAuthServer,
                isNewBrowserSession: oidcCore.isNewBrowserSession,
                subscribeToAutoLogoutState: next => {
                    next(evtAutoLogoutState.current);
                    const { unsubscribe } = evtAutoLogoutState.subscribe(next);
                    return { unsubscribeFromAutoLogoutState: unsubscribe };
                }
            })
            : (0, id_1.id)({
                ...common,
                isUserLoggedIn: false,
                initializationError: oidcCore.initializationError,
                login: oidcCore.login
            });
    }
    const prModuleCore = !isBrowser_1.isBrowser ? undefined : Promise.resolve().then(() => __importStar(require("../core")));
    let bootstrapOidc_prResolved = undefined;
    const bootstrapOidc = (paramsOfBootstrap) => {
        if (bootstrapOidc_prResolved !== undefined) {
            return bootstrapOidc_prResolved;
        }
        bootstrapOidc_prResolved = dOidcCoreOrInitializationError.pr.then(() => { });
        (async () => {
            if (!isBrowser_1.isBrowser) {
                return;
            }
            (0, assert_1.assert)(prModuleCore !== undefined);
            dParamsOfBootstrap.resolve(paramsOfBootstrap);
            switch (paramsOfBootstrap.implementation) {
                case "mock":
                    {
                        const { createMockOidc } = await Promise.resolve().then(() => __importStar(require("../core/createMockOidc")));
                        const oidcCore = await createMockOidc({
                            BASE_URL: paramsOfBootstrap.BASE_URL,
                            // NOTE: The `as false` is lying here, it's just to preserve some level of type-safety.
                            autoLogin: autoLogin,
                            // NOTE: Same here, the nullish coalescing is lying.
                            isUserInitiallyLoggedIn: paramsOfBootstrap.isUserInitiallyLoggedIn,
                            mockedParams: {
                                clientId: paramsOfBootstrap.clientId_mock,
                                issuerUri: paramsOfBootstrap.issuerUri_mock
                            },
                            mockedTokens: {
                                decodedIdToken: paramsOfBootstrap.decodedIdToken_mock ??
                                    decodedIdToken_mock ??
                                    (0, createObjectThatThrowsIfAccessed_1.createObjectThatThrowsIfAccessed)({
                                        debugMessage: [
                                            "oidc-spa: You didn't provide any mock for the decodedIdToken",
                                            "Either provide a default one by specifying decodedIdToken_mock",
                                            "as parameter of .withExpectedDecodedIdTokenShape() or",
                                            "specify decodedIdToken_mock when calling bootstrapOidc()"
                                        ].join(" ")
                                    })
                            }
                        });
                        dOidcCoreOrInitializationError.resolve(oidcCore);
                    }
                    break;
                case "real":
                    {
                        const { createOidc } = await prModuleCore;
                        let oidcCoreOrInitializationError;
                        try {
                            oidcCoreOrInitializationError = await createOidc({
                                BASE_URL: paramsOfBootstrap.BASE_URL,
                                autoLogin,
                                decodedIdTokenSchema,
                                issuerUri: paramsOfBootstrap.issuerUri,
                                clientId: paramsOfBootstrap.clientId,
                                idleSessionLifetimeInSeconds: paramsOfBootstrap.idleSessionLifetimeInSeconds,
                                scopes: paramsOfBootstrap.scopes,
                                transformUrlBeforeRedirect: paramsOfBootstrap.transformUrlBeforeRedirect,
                                extraQueryParams: paramsOfBootstrap.extraQueryParams,
                                extraTokenParams: paramsOfBootstrap.extraTokenParams,
                                sessionRestorationMethod: paramsOfBootstrap.sessionRestorationMethod,
                                debugLogs: paramsOfBootstrap.debugLogs,
                                __unsafe_clientSecret: paramsOfBootstrap.__unsafe_clientSecret,
                                __metadata: paramsOfBootstrap.__metadata,
                                __unsafe_useIdTokenAsAccessToken: paramsOfBootstrap.__unsafe_useIdTokenAsAccessToken,
                                autoLogoutParams: paramsOfBootstrap.autoLogoutParams,
                                disableDPoP: paramsOfBootstrap.disableDPoP
                            });
                        }
                        catch (error) {
                            if (!(error instanceof OidcInitializationError_1.OidcInitializationError)) {
                                throw error;
                            }
                            dOidcCoreOrInitializationError.resolve(error);
                            return;
                        }
                        dOidcCoreOrInitializationError.resolve(oidcCoreOrInitializationError);
                    }
                    break;
            }
        })();
        return bootstrapOidc_prResolved;
    };
    async function enforceLogin(loaderContext) {
        if (!isBrowser_1.isBrowser) {
            throw new Error("oidc-spa: getOidc() can't be used on the server");
        }
        const { cause } = loaderContext;
        const redirectUrl = (() => {
            if (loaderContext.request?.url !== undefined) {
                return (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                    urlish: loaderContext.request.url,
                    doAssertNoQueryParams: false
                });
            }
            if (loaderContext.location?.publicHref !== undefined) {
                return (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                    urlish: loaderContext.location.publicHref,
                    doAssertNoQueryParams: false
                });
            }
            return window.location.href;
        })();
        const oidc = await getOidc();
        const isUrlAlreadyReplaced = window.location.href.replace(/\/$/, "") === redirectUrl.replace(/\/$/, "");
        if (!oidc.isUserLoggedIn) {
            if (cause === "preload") {
                throw new Error("oidc-spa: User is not yet logged in. This is an expected error, nothing to be addressed.");
            }
            await oidc.login({
                redirectUrl,
                doesCurrentHrefRequiresAuth: isUrlAlreadyReplaced
            });
        }
        define_temporary_postLoginRedirectUrl: {
            if (isUrlAlreadyReplaced) {
                break define_temporary_postLoginRedirectUrl;
            }
            (0, desiredPostLoginRedirectUrl_1.setDesiredPostLoginRedirectUrl)({ postLoginRedirectUrl: redirectUrl });
            const history_pushState = history.pushState;
            const history_replaceState = history.replaceState;
            const onNavigated = () => {
                history.pushState = history_pushState;
                history.replaceState = history_replaceState;
                (0, desiredPostLoginRedirectUrl_1.setDesiredPostLoginRedirectUrl)({ postLoginRedirectUrl: undefined });
            };
            history.pushState = function pushState(...args) {
                onNavigated();
                return history_pushState.call(history, ...args);
            };
            history.replaceState = function replaceState(...args) {
                onNavigated();
                return history_replaceState.call(history, ...args);
            };
        }
    }
    function OidcInitializationErrorGate(props) {
        const { errorComponent: ErrorComponent, children } = props;
        const { hasResolved, value: oidcCoreOrOidcInitializationError } = dOidcCoreOrInitializationError.getState();
        if (!hasResolved) {
            throw dOidcCoreOrInitializationError.pr;
        }
        if (oidcCoreOrOidcInitializationError instanceof OidcInitializationError_1.OidcInitializationError) {
            const oidcInitializationError = oidcCoreOrOidcInitializationError;
            return (0, react_1.createElement)(ErrorComponent, { oidcInitializationError });
        }
        return children;
    }
    function OidcInitializationGate(props) {
        const { fallback, children } = props;
        const [isReadyToRender, readyToRender] = (0, react_1.useReducer)(() => true, false);
        (0, react_1.useEffect)(() => {
            let isActive = true;
            dOidcCoreOrInitializationError.pr.then(() => {
                if (!isActive) {
                    return;
                }
                readyToRender();
            });
            return () => {
                isActive = false;
            };
        }, []);
        if (!isReadyToRender) {
            return fallback !== undefined ? fallback : null;
        }
        return children;
    }
    function withLoginEnforced(component) {
        const Component = component;
        function ComponentWithLoginEnforced(props) {
            const { hasResolved, value: oidcCore } = dOidcCoreOrInitializationError.getState();
            if (!hasResolved) {
                throw dOidcCoreOrInitializationError.pr;
            }
            if (oidcCore instanceof OidcInitializationError_1.OidcInitializationError) {
                throw oidcCore;
            }
            if (!oidcCore.isUserLoggedIn) {
                throw oidcCore.login({ doesCurrentHrefRequiresAuth: true });
            }
            return (0, react_1.createElement)(Component, props);
        }
        ComponentWithLoginEnforced.displayName = `${Component.displayName ?? Component.name ?? "Component"}WithLoginEnforced`;
        return ComponentWithLoginEnforced;
    }
    // @ts-expect-error
    return {
        bootstrapOidc,
        useOidc,
        getOidc,
        OidcInitializationGate,
        OidcInitializationErrorGate,
        enforceLogin,
        withLoginEnforced
    };
}
