// oidc-spa/react-spa/utilsBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.oidcSpaUtilsBuilder = void 0;
const createOidcSpaUtils_1 = require("./createOidcSpaUtils");
function createOidcSpaUtilsBuilder(params) {
    return {
        withAutoLogin: () => createOidcSpaUtilsBuilder({
            autoLogin: true,
            decodedIdTokenSchema: params.decodedIdTokenSchema,
            decodedIdToken_mock: params.decodedIdToken_mock
        }),
        withExpectedDecodedIdTokenShape: ({ decodedIdTokenSchema, decodedIdToken_mock }) => createOidcSpaUtilsBuilder({
            autoLogin: params.autoLogin,
            decodedIdTokenSchema,
            decodedIdToken_mock: decodedIdToken_mock
        }),
        createUtils: () => (0, createOidcSpaUtils_1.createOidcSpaUtils)({
            autoLogin: params.autoLogin,
            decodedIdTokenSchema: params.decodedIdTokenSchema,
            decodedIdToken_mock: params.decodedIdToken_mock
        })
    };
}
exports.oidcSpaUtilsBuilder = createOidcSpaUtilsBuilder({
    autoLogin: false,
    decodedIdToken_mock: undefined,
    decodedIdTokenSchema: undefined
});
