// oidc-spa/tools/Evt.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createEvt = createEvt;
const Deferred_1 = require("./Deferred");
function createEvt() {
    const listeners = [];
    let postCount = 0;
    const evt = {
        subscribe: next => {
            listeners.push(next);
            let isActive = true;
            return {
                unsubscribe: () => {
                    if (!isActive) {
                        return;
                    }
                    isActive = false;
                    const i = listeners.indexOf(next);
                    if (i >= 0) {
                        listeners.splice(i, 1);
                    }
                }
            };
        },
        waitFor: () => {
            const d = new Deferred_1.Deferred();
            const { unsubscribe } = evt.subscribe(data => {
                unsubscribe();
                d.resolve(data);
            });
            return d.pr;
        },
        post: (data) => {
            postCount++;
            const snapshot = listeners.slice();
            for (const l of snapshot) {
                try {
                    l(data);
                }
                catch { }
            }
        },
        get postCount() {
            return postCount;
        }
    };
    return evt;
}
