// oidc-spa/tools/lazySessionStorage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SESSION_STORAGE_GLOBAL_PREFIX = void 0;
exports.createLazySessionStorage = createLazySessionStorage;
const assert_1 = require("../tools/tsafe/assert");
exports.SESSION_STORAGE_GLOBAL_PREFIX = `oidc-spa:lazy-session-storage:`;
function createLazySessionStorage(params) {
    const { storageId } = params;
    const sessionStoragePrefix = `${exports.SESSION_STORAGE_GLOBAL_PREFIX}${storageId}:`;
    const getSessionStorageKey = (key) => `${sessionStoragePrefix}${key}`;
    const entries = [];
    for (let i = 0; i < sessionStorage.length; i++) {
        const key = sessionStorage.key(i);
        (0, assert_1.assert)(key !== null, "470498");
        if (!key.startsWith(sessionStoragePrefix)) {
            continue;
        }
        const value = sessionStorage.getItem(key);
        (0, assert_1.assert)(value !== null, "846771");
        sessionStorage.removeItem(key);
        entries.push({
            key: key.slice(sessionStoragePrefix.length),
            value
        });
    }
    let isPersistenceEnabled = false;
    const storage = {
        persistCurrentStateAndSubsequentChanges: () => {
            isPersistenceEnabled = true;
            for (let i = 0; i < storage.length; i++) {
                const key = storage.key(i);
                (0, assert_1.assert)(key !== null, "803385");
                const value = storage.getItem(key);
                (0, assert_1.assert)(value !== null, "777098");
                storage.setItem(key, value);
            }
        },
        get length() {
            return entries.length;
        },
        key: index => {
            const entry = entries[index];
            if (entry === undefined) {
                return null;
            }
            return entry.key;
        },
        removeItem: key => {
            const entry = entries.find(entry => entry.key === key);
            if (entry === undefined) {
                return;
            }
            sessionStorage.removeItem(getSessionStorageKey(entry.key));
            const index = entries.indexOf(entry);
            entries.splice(index, 1);
        },
        clear: () => {
            for (let i = 0; i < storage.length; i++) {
                const key = storage.key(i);
                (0, assert_1.assert)(key !== null, "290875");
                storage.removeItem(key);
            }
        },
        getItem: key => {
            const entry = entries.find(entry => entry.key === key);
            if (entry === undefined) {
                return null;
            }
            return entry.value;
        },
        setItem: (key, value) => {
            if (isPersistenceEnabled) {
                sessionStorage.setItem(getSessionStorageKey(key), value);
            }
            update: {
                const entry = entries.find(entry => entry.key === key);
                if (entry === undefined) {
                    break update;
                }
                entry.value = value;
                return;
            }
            entries.push({ key, value });
        }
    };
    return storage;
}
