// oidc-spa/tools/parseKeycloakIssuerUri.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parseKeycloakIssuerUri = parseKeycloakIssuerUri;
const keycloak_1 = require("../keycloak");
/**
 * Return undefined if the issuerUri doesn't match the expected shape of a Keycloak issuerUri
 *
 * Example:
 *
 * `parseKeycloakIssuerUri("https://auth.my-company.com/auth/realms/myrealm")` returns:
 * {
 *    origin: "https://auth.my-company.com",
 *    realm: "myrealm",
 *    kcHttpRelativePath: "/auth",
 *    adminConsoleUrl: "https://auth.my-company.com/auth/admin/myrealm/console",
 *    getAccountUrl: ({ thisAppDisplayName, backToAppFromAccountUrl }) =>
 *        `https://auth.my-company.com/auth/realms/myrealm/account?referrer=${thisAppDisplayName}&referrer_uri=${backToAppFromAccountUrl}`
 * }
 * */
function parseKeycloakIssuerUri(issuerUri) {
    if (!(0, keycloak_1.isKeycloak)({ issuerUri })) {
        return undefined;
    }
    const keycloakUtils = (0, keycloak_1.createKeycloakUtils)({ issuerUri });
    return {
        origin: keycloakUtils.issuerUriParsed.origin,
        realm: keycloakUtils.issuerUriParsed.realm,
        kcHttpRelativePath: keycloakUtils.issuerUriParsed.kcHttpRelativePath,
        adminConsoleUrl: keycloakUtils.adminConsoleUrl,
        adminConsoleUrl_master: keycloakUtils.adminConsoleUrl_master,
        getAccountUrl: ({ clientId, backToAppFromAccountUrl, locale }) => keycloakUtils.getAccountUrl({
            clientId,
            validRedirectUri: backToAppFromAccountUrl,
            locale
        })
    };
}
