// oidc-spa/esm/core/earlyInit.mjs
import { getStateData, getIsStatQueryParamValue } from "./StateData.mjs";
import { assert } from "../tools/tsafe/assert.mjs";
import { setBASE_URL_earlyInit } from "./earlyInit_BASE_URL.mjs";
import { isBrowser } from "../tools/isBrowser.mjs";
import { createEvt } from "../tools/Evt.mjs";
import { setGetRootRelativeOriginalLocationHref_earlyInit, getRootRelativeOriginalLocationHref_earlyInit } from "./earlyInit_rootRelativeOriginalLocationHref.mjs";
import { prModuleCreateOidc } from "./earlyInit_prModuleCreateOidc.mjs";
const IFRAME_MESSAGE_PREFIX = "oidc-spa:cross-window-messaging:";
let shouldLoadApp = undefined;
export function oidcEarlyInit(params) {
    if (shouldLoadApp !== undefined) {
        return { shouldLoadApp };
    }
    shouldLoadApp = oidcEarlyInit_nonMemoized(params).shouldLoadApp;
    return { shouldLoadApp };
}
function oidcEarlyInit_nonMemoized(params) {
    const { BASE_URL, sessionRestorationMethod, securityDefenses = {} } = params ?? {};
    if (!isBrowser) {
        return { shouldLoadApp: true };
    }
    const { shouldLoadApp } = handleOidcCallback();
    let exports_earlyInit;
    if (shouldLoadApp) {
        let evtIframeAuthResponse = undefined;
        {
            const _MessageEvent_prototype_data_get = (() => {
                const pd = Object.getOwnPropertyDescriptor(MessageEvent.prototype, "data");
                assert(pd !== undefined);
                const { get } = pd;
                assert(get !== undefined);
                return get;
            })();
            const _MessageEvent_prototype_origin_get = (() => {
                const pd = Object.getOwnPropertyDescriptor(MessageEvent.prototype, "origin");
                assert(pd !== undefined);
                const { get } = pd;
                assert(get !== undefined);
                return get;
            })();
            const _Event_prototype_stopImmediatePropagation_value = Event.prototype.stopImmediatePropagation;
            const origin = window.location.origin;
            window.addEventListener("message", event => {
                if (_MessageEvent_prototype_origin_get.call(event) !== origin) {
                    return;
                }
                const eventData = _MessageEvent_prototype_data_get.call(event);
                if (typeof eventData !== "string") {
                    return;
                }
                if (!eventData.startsWith(IFRAME_MESSAGE_PREFIX)) {
                    return;
                }
                _Event_prototype_stopImmediatePropagation_value.call(event);
                const authResponse = JSON.parse(eventData.slice(IFRAME_MESSAGE_PREFIX.length));
                (evtIframeAuthResponse ?? (evtIframeAuthResponse = createEvt())).post(authResponse);
            }, {
                capture: true,
                once: false,
                passive: false
            });
        }
        if (BASE_URL !== undefined) {
            setBASE_URL_earlyInit({ BASE_URL });
        }
        {
            const { enableBrowserRuntimeFreeze, enableDPoP, enableTokenSubstitution } = securityDefenses;
            enableDPoP?.();
            enableTokenSubstitution?.();
            enableBrowserRuntimeFreeze?.();
        }
        exports_earlyInit = {
            shouldLoadApp: true,
            getEvtIframeAuthResponse: () => {
                return (evtIframeAuthResponse ?? (evtIframeAuthResponse = createEvt()));
            },
            getRedirectAuthResponse: () => {
                return redirectAuthResponse === undefined
                    ? { authResponse: undefined }
                    : {
                        authResponse: redirectAuthResponse,
                        clearAuthResponse: () => {
                            redirectAuthResponse = undefined;
                        }
                    };
            },
            sessionRestorationMethod
        };
    }
    else {
        exports_earlyInit = {
            shouldLoadApp: false
        };
    }
    prModuleCreateOidc.then(({ registerExports_earlyInit }) => {
        registerExports_earlyInit(exports_earlyInit);
    });
    return { shouldLoadApp };
}
let redirectAuthResponse = undefined;
function handleOidcCallback() {
    const location_urlObj = new URL(window.location.href);
    const locationHrefAssessment = (() => {
        fragment: {
            const stateUrlParamValue = new URLSearchParams(location_urlObj.hash.replace(/^#/, "")).get("state");
            if (stateUrlParamValue === null) {
                break fragment;
            }
            if (!getIsStatQueryParamValue({ maybeStateUrlParamValue: stateUrlParamValue })) {
                break fragment;
            }
            return { hasAuthResponseInUrl: true, responseMode: "fragment" };
        }
        query: {
            const stateUrlParamValue = location_urlObj.searchParams.get("state");
            if (stateUrlParamValue === null) {
                break query;
            }
            if (!getIsStatQueryParamValue({ maybeStateUrlParamValue: stateUrlParamValue })) {
                break query;
            }
            if (location_urlObj.searchParams.get("client_id") !== null &&
                location_urlObj.searchParams.get("response_type") !== null &&
                location_urlObj.searchParams.get("redirect_uri") !== null) {
                // NOTE: We are probably in a Keycloakify theme and oidc-spa was loaded by mistake.
                break query;
            }
            return { hasAuthResponseInUrl: true, responseMode: "query" };
        }
        return { hasAuthResponseInUrl: false };
    })();
    if (!locationHrefAssessment.hasAuthResponseInUrl) {
        setGetRootRelativeOriginalLocationHref_earlyInit({
            rootRelativeOriginalLocationHref: location_urlObj.href.slice(location_urlObj.origin.length)
        });
        return { shouldLoadApp: true };
    }
    setGetRootRelativeOriginalLocationHref_earlyInit({
        rootRelativeOriginalLocationHref: location_urlObj.pathname
    });
    const { authResponse } = (() => {
        const authResponse = { state: "" };
        const searchParams = (() => {
            switch (locationHrefAssessment.responseMode) {
                case "fragment":
                    return new URLSearchParams(location_urlObj.hash.replace(/^#/, ""));
                case "query":
                    return location_urlObj.searchParams;
                default:
                    assert(false);
            }
        })();
        for (const [key, value] of searchParams) {
            authResponse[key] = value;
        }
        assert(authResponse.state !== "", "063965");
        return { authResponse };
    })();
    const stateData = getStateData({ stateUrlParamValue: authResponse.state });
    if (stateData === undefined) {
        history.replaceState({}, "", getRootRelativeOriginalLocationHref_earlyInit());
        return { shouldLoadApp: true };
    }
    switch (stateData.context) {
        case "iframe":
            if (parent !== top) {
                const errorMessage = [
                    "oidc-spa: For security reasons, refusing to post the auth response.",
                    "If you want your app to be framable use sessionRestorationMethod: 'full page redirect'."
                ].join(" ");
                alert(errorMessage);
                throw new Error(errorMessage);
            }
            parent.postMessage(`${IFRAME_MESSAGE_PREFIX}${JSON.stringify(authResponse)}`, location.origin);
            return { shouldLoadApp: false };
        case "redirect": {
            redirectAuthResponse = authResponse;
            const rootRelativeRedirectUrl = (() => {
                if (stateData.action === "login" && authResponse.error === "consent_required") {
                    return stateData.rootRelativeRedirectUrl_consentRequiredCase;
                }
                return stateData.rootRelativeRedirectUrl;
            })();
            history.replaceState({}, "", rootRelativeRedirectUrl);
            return { shouldLoadApp: true };
        }
        default:
            assert(false);
    }
}
