// oidc-spa/esm/core/OidcMetadata.mjs
import { assert } from "../tools/tsafe/assert.mjs";
import { getIsLikelyDevServer } from "../tools/isLikelyDevServer.mjs";
assert;
export const WELL_KNOWN_PATH = "/.well-known/openid-configuration";
function getSessionStorageKey(params) {
    const { issuerUri } = params;
    return `oidc-spa:openid-configuration:${issuerUri}`;
}
function readSessionStorage(params) {
    const { issuerUri } = params;
    const value = sessionStorage.getItem(getSessionStorageKey({ issuerUri }));
    if (value === null) {
        return undefined;
    }
    return JSON.parse(value);
}
function setSessionStorage(params) {
    const { issuerUri, oidcMetadata } = params;
    sessionStorage.setItem(getSessionStorageKey({ issuerUri }), JSON.stringify(oidcMetadata));
}
export async function fetchOidcMetadata(params) {
    const { issuerUri } = params;
    from_cache: {
        const oidcMetadata = readSessionStorage({ issuerUri });
        if (oidcMetadata === undefined) {
            break from_cache;
        }
        return oidcMetadata;
    }
    let oidcMetadata;
    try {
        const response = await fetch(`${issuerUri}${WELL_KNOWN_PATH}`, {
            headers: {
                Accept: "application/jwk-set+json, application/json"
            }
        });
        if (!response.ok) {
            throw new Error();
        }
        const obj = await response.json();
        {
            const { authorization_endpoint } = obj;
            if (typeof authorization_endpoint !== "string") {
                throw new Error();
            }
        }
        oidcMetadata = obj;
    }
    catch {
        return undefined;
    }
    if (!getIsLikelyDevServer()) {
        setSessionStorage({ issuerUri, oidcMetadata });
    }
    return oidcMetadata;
}
