// oidc-spa/esm/core/StateData.mjs
import { assert } from "../tools/tsafe/assert.mjs";
import { typeGuard } from "../tools/tsafe/typeGuard.mjs";
import { generateUrlSafeRandom } from "../tools/generateUrlSafeRandom.mjs";
const STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX = "b2lkYy1zcGEu";
const RANDOM_STRING_LENGTH = 32 - STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX.length;
export function generateStateUrlParamValue() {
    return `${STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX}${generateUrlSafeRandom({
        length: RANDOM_STRING_LENGTH
    })}`;
}
export function getIsStatQueryParamValue(params) {
    const { maybeStateUrlParamValue } = params;
    return (maybeStateUrlParamValue.startsWith(STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX) &&
        maybeStateUrlParamValue.length ===
            STATE_QUERY_PARAM_VALUE_IDENTIFIER_PREFIX.length + RANDOM_STRING_LENGTH);
}
export const STATE_STORE_KEY_PREFIX = "oidc.";
function getKey(params) {
    const { stateUrlParamValue } = params;
    return `${STATE_STORE_KEY_PREFIX}${stateUrlParamValue}`;
}
function getStateStore(params) {
    const { stateUrlParamValue } = params;
    const item = localStorage.getItem(getKey({ stateUrlParamValue }));
    if (item === null) {
        return undefined;
    }
    const obj = JSON.parse(item);
    assert(typeGuard(obj, obj instanceof Object && obj.data instanceof Object && typeof obj.data.context === "string"));
    return obj;
}
export function clearStateStore(params) {
    const { stateUrlParamValue } = params;
    localStorage.removeItem(getKey({ stateUrlParamValue }));
}
export function getStateData(params) {
    const { stateUrlParamValue } = params;
    const stateStore = getStateStore({ stateUrlParamValue });
    if (stateStore === undefined) {
        return undefined;
    }
    return stateStore.data;
}
