// oidc-spa/esm/vendor/server/tsafe.mjs
// node_modules/tsafe/esm/assert.mjs
var AssertionError = class extends Error {
  constructor(msg) {
    super(`Wrong assertion encountered` + (!msg ? "" : `: "${msg}"`));
    this.originalMessage = msg;
    Object.setPrototypeOf(this, new.target.prototype);
  }
};
var refOfIs = void 0;
function assert(condition, msg) {
  if (arguments.length === 0) {
    condition = true;
  }
  if (refOfIs !== void 0) {
    refOfIs = void 0;
    return;
  }
  if (!condition) {
    const error = new AssertionError(typeof msg === "function" ? msg() : msg);
    if (Error.captureStackTrace) {
      Error.captureStackTrace(error, assert);
    }
    throw error;
  }
}
var errorMessage = "Wrong usage of the `is` function refer to https://docs.tsafe.dev/is";
function is(value) {
  const ref = {};
  if (refOfIs !== void 0) {
    refOfIs = void 0;
    throw new Error(errorMessage);
  }
  refOfIs = ref;
  Promise.resolve().then(() => {
    if (refOfIs === ref) {
      throw new Error(errorMessage);
    }
  });
  return null;
}

// node_modules/tsafe/esm/id.mjs
var id = (x) => x;

// node_modules/tsafe/esm/isAmong.mjs
function isAmong(names, value) {
  for (const name of names) {
    if (name === value) {
      return true;
    }
  }
  return false;
}

// node_modules/tsafe/esm/exclude.mjs
function exclude(target) {
  const test = target instanceof Object ? (element) => target.indexOf(element) < 0 : (element) => element !== target;
  return function(element) {
    return test(element);
  };
}

// node_modules/tsafe/esm/Reflect.mjs
function Reflect() {
  return null;
}

// node_modules/tsafe/esm/objectFromEntries.mjs
function objectFromEntries(entries) {
  return Object.fromEntries(entries);
}
export {
  Reflect,
  assert,
  exclude,
  id,
  is,
  isAmong,
  objectFromEntries
};
