// oidc-spa/esm/keycloak/keycloak-js/keycloak-authz.mjs
/*
 *  Copyright 2016 Red Hat, Inc. and/or its affiliates
 *  and other contributors as indicated by the @author tags.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *  http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 *
 */
export class KeycloakAuthorization {
    constructor(keycloak) {
        this.keycloak = keycloak;
        this.rpt = null;
    }
    /**
     * This method enables client applications to better integrate with resource servers protected by a Keycloak
     * policy enforcer using UMA protocol.
     *
     * The authorization request must be provided with a ticket.
     *
     * @param authorizationRequest An AuthorizationRequest instance with a valid permission ticket set.
     * @returns A promise to set functions to be invoked on grant, deny or error.
     */
    authorize(authorizationRequest) {
        const then = async (onGrant, onDeny, onError) => {
            try {
                await this.initializeConfigIfNeeded();
            }
            catch (error) {
                handleError(error, onError?.bind(this));
                return;
            }
            if (authorizationRequest && authorizationRequest.ticket) {
                const request = new globalThis.XMLHttpRequest();
                request.open("POST", this.config.token_endpoint, true);
                request.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
                if (!this.keycloak.oidc.isUserLoggedIn) {
                    handleError(new Error("oidc-spa: The user is not currently logged in"));
                    return;
                }
                request.setRequestHeader("Authorization", "Bearer " + (await this.keycloak.oidc.getTokens()).accessToken);
                request.onreadystatechange = () => {
                    if (request.readyState === 4) {
                        const status = request.status;
                        if (status >= 200 && status < 300) {
                            const rpt = JSON.parse(request.responseText).access_token;
                            this.rpt = rpt;
                            onGrant.call(this, rpt);
                        }
                        else if (status === 403) {
                            if (onDeny) {
                                onDeny.call(this);
                            }
                            else {
                                console.error("Authorization request was denied by the server.");
                            }
                        }
                        else {
                            if (onError) {
                                onError();
                            }
                            else {
                                console.error("Could not obtain authorization data from server.");
                            }
                        }
                    }
                };
                let params = "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket&client_id=" +
                    this.keycloak.clientId +
                    "&ticket=" +
                    authorizationRequest.ticket;
                if (authorizationRequest.submitRequest !== undefined) {
                    params += "&submit_request=" + authorizationRequest.submitRequest;
                }
                const metadata = authorizationRequest.metadata;
                if (metadata) {
                    const includeName = metadata.responseIncludeResourceName ?? metadata.response_include_resource_name;
                    if (includeName !== undefined) {
                        params += "&response_include_resource_name=" + includeName;
                    }
                    const permissionsLimit = metadata.responsePermissionsLimit ?? metadata.response_permissions_limit;
                    if (permissionsLimit !== undefined) {
                        params += "&response_permissions_limit=" + permissionsLimit;
                    }
                }
                if (this.rpt &&
                    (authorizationRequest.incrementalAuthorization === undefined ||
                        authorizationRequest.incrementalAuthorization)) {
                    params += "&rpt=" + this.rpt;
                }
                request.send(params);
            }
        };
        return {
            then
        };
    }
    /**
     * Obtains all entitlements from a Keycloak server based on a given resourceServerId.
     *
     * @param resourceServerId The id (client id) of the resource server to obtain permissions from.
     * @param authorizationRequest An AuthorizationRequest instance.
     * @returns A promise to set functions to be invoked on grant, deny or error.
     */
    entitlement(resourceServerId, authorizationRequest = {}) {
        const then = async (onGrant, onDeny, onError) => {
            try {
                await this.initializeConfigIfNeeded();
            }
            catch (error) {
                handleError(error, onError?.bind(this));
                return;
            }
            const request = new globalThis.XMLHttpRequest();
            request.open("POST", this.config.token_endpoint, true);
            request.setRequestHeader("Content-type", "application/x-www-form-urlencoded");
            if (!this.keycloak.oidc.isUserLoggedIn) {
                handleError(new Error("oidc-spa: The user is not currently logged in"));
                return;
            }
            request.setRequestHeader("Authorization", "Bearer " + (await this.keycloak.oidc.getTokens()).accessToken);
            request.onreadystatechange = () => {
                if (request.readyState === 4) {
                    const status = request.status;
                    if (status >= 200 && status < 300) {
                        const rpt = JSON.parse(request.responseText).access_token;
                        this.rpt = rpt;
                        onGrant.call(this, rpt);
                    }
                    else if (status === 403) {
                        if (onDeny) {
                            onDeny.call(this);
                        }
                        else {
                            console.error("Authorization request was denied by the server.");
                        }
                    }
                    else {
                        if (onError) {
                            onError.call(this);
                        }
                        else {
                            console.error("Could not obtain authorization data from server.");
                        }
                    }
                }
            };
            let params = "grant_type=urn:ietf:params:oauth:grant-type:uma-ticket&client_id=" +
                this.keycloak.clientId;
            if (authorizationRequest.claimToken) {
                params += "&claim_token=" + authorizationRequest.claimToken;
                if (authorizationRequest.claimTokenFormat) {
                    params += "&claim_token_format=" + authorizationRequest.claimTokenFormat;
                }
            }
            params += "&audience=" + resourceServerId;
            const permissions = authorizationRequest.permissions ?? [];
            for (let i = 0; i < permissions.length; i++) {
                const resource = permissions[i];
                let permission = resource.id;
                if (resource.scopes && resource.scopes.length > 0) {
                    permission += "#";
                    for (let j = 0; j < resource.scopes.length; j++) {
                        const scope = resource.scopes[j];
                        if (permission.indexOf("#") !== permission.length - 1) {
                            permission += ",";
                        }
                        permission += scope;
                    }
                }
                params += "&permission=" + permission;
            }
            const metadata = authorizationRequest.metadata;
            if (metadata) {
                const includeName = metadata.responseIncludeResourceName ?? metadata.response_include_resource_name;
                if (includeName !== undefined) {
                    params += "&response_include_resource_name=" + includeName;
                }
                const permissionsLimit = metadata.responsePermissionsLimit ?? metadata.response_permissions_limit;
                if (permissionsLimit !== undefined) {
                    params += "&response_permissions_limit=" + permissionsLimit;
                }
            }
            if (this.rpt) {
                params += "&rpt=" + this.rpt;
            }
            request.send(params);
        };
        return { then };
    }
    async initializeConfigIfNeeded() {
        if (this.config) {
            return this.config;
        }
        if (this.configPromise) {
            return await this.configPromise;
        }
        if (!this.keycloak.didInitialize) {
            throw new Error("The Keycloak instance has not been initialized yet.");
        }
        this.configPromise = loadConfig(this.keycloak.authServerUrl, this.keycloak.realm);
        this.config = await this.configPromise;
        return this.config;
    }
}
/**
 * Obtains the configuration from the server.
 * @param serverUrl The URL of the Keycloak server.
 * @param realm The realm name.
 * @returns A promise that resolves when the configuration is loaded.
 */
async function loadConfig(serverUrl, realm) {
    const url = `${serverUrl}/realms/${encodeURIComponent(realm)}/.well-known/uma2-configuration`;
    try {
        return await fetchJSON(url);
    }
    catch (error) {
        throw new Error("Could not obtain configuration from server.", 
        //@ts-expect-error
        { cause: error });
    }
}
/**
 * Fetches the JSON data from the given URL.
 * @param url The URL to fetch the data from.
 * @returns A promise that resolves when the data is loaded.
 */
async function fetchJSON(url) {
    let response;
    try {
        response = await fetch(url);
    }
    catch (error) {
        throw new Error("Server did not respond.", 
        //@ts-expect-error
        { cause: error });
    }
    if (!response.ok) {
        throw new Error("Server responded with an invalid status.");
    }
    try {
        return (await response.json());
    }
    catch (error) {
        throw new Error("Server responded with invalid JSON.", 
        // @ts-expect-error
        { cause: error });
    }
}
function handleError(error, handler) {
    if (handler) {
        handler(error);
    }
    else {
        console.error(error);
    }
}
