// oidc-spa/esm/keycloak/keycloak-js/Keycloak.mjs
var __classPrivateFieldSet = (this && this.__classPrivateFieldSet) || function (receiver, state, value, kind, f) {
    if (kind === "m")
        throw new TypeError("Private method is not writable");
    if (kind === "a" && !f)
        throw new TypeError("Private accessor was defined without a setter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
        throw new TypeError("Cannot write private member to an object whose class did not declare it");
    return (kind === "a" ? f.call(receiver, value) : f ? f.value = value : state.set(receiver, value)), value;
};
var __classPrivateFieldGet = (this && this.__classPrivateFieldGet) || function (receiver, state, kind, f) {
    if (kind === "a" && !f)
        throw new TypeError("Private accessor was defined without a getter");
    if (typeof state === "function" ? receiver !== state || !f : !state.has(receiver))
        throw new TypeError("Cannot read private member from an object whose class did not declare it");
    return kind === "m" ? f : kind === "a" ? f.call(receiver) : f ? f.value : state.get(receiver);
};
var _Keycloak_instances, _Keycloak_state, _Keycloak_init, _Keycloak_login, _Keycloak_logout, _Keycloak_register, _Keycloak_accountManagement, _Keycloak_createAccountUrl, _Keycloak_isTokenExpired, _Keycloak_updateToken, _Keycloak_hasRealmRole, _Keycloak_hasResourceRole, _Keycloak_loadUserProfile, _Keycloak_loadUserInfo;
import { assert, is } from "../../tools/tsafe/assert.mjs";
import { isAmong } from "../../tools/tsafe/isAmong.mjs";
import { createOidc, OidcInitializationError } from "../../core/index.mjs";
import { Deferred } from "../../tools/Deferred.mjs";
import { decodeJwt } from "../../tools/decodeJwt.mjs";
import { createKeycloakUtils } from "../keycloakUtils.mjs";
import { createStatefulEvt } from "../../tools/StatefulEvt.mjs";
import { readExpirationTimeInJwt } from "../../tools/readExpirationTimeInJwt.mjs";
import { getHomeAndRedirectUri } from "../../core/homeAndRedirectUri.mjs";
/**
 * This module provides a drop-in replacement for `keycloak-js`,
 * designed for teams migrating to `oidc-spa` with minimal changes.
 *
 * While the import path is `oidc-spa/keycloak-js`, this is *not* a re-export or patch —
 * it is a full alternative implementation aligned with the `keycloak-js` API.
 */
export class Keycloak {
    /**
     * Creates a new Keycloak client instance.
     * @param config A configuration object or path to a JSON config file.
     *
     * NOTE oidc-spa: Currently not supporting GenericOidcConfig (providing explicitly authorization_endpoint ect)
     * But we could if with the __metadata parameter of oidc-spa.
     * I'm not seeing the usecase when ran against keycloak right now so not doing it.
     */
    constructor(params) {
        _Keycloak_instances.add(this);
        _Keycloak_state.set(this, void 0);
        /**
         * Called to initialize the adapter.
         * @param initOptions Initialization options.
         * @returns A promise to set functions to be invoked on success or error.
         */
        this.init = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_init).bind(this);
        /**
         * Response mode passed in init (default value is `'fragment'`).
         *
         * NOTE oidc-spa: Can only be fragment.
         */
        this.responseMode = "fragment";
        /**
         * Response type sent to Keycloak with login requests. This is determined
         * based on the flow value used during initialization, but can be overridden
         * by setting this value.
         *
         * NOTE oidc-spa: Can only be 'code'
         */
        this.responseType = "code";
        /**
         * Flow passed in init.
         *
         * NOTE oidc-spa: Can only be 'standard'
         */
        this.flow = "standard";
        /**
         * Redirects to login form.
         * @param options Login options.
         */
        this.login = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_login).bind(this);
        /**
         * Redirects to logout.
         * @param options Logout options.
         */
        this.logout = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_logout).bind(this);
        /**
         * Redirects to registration form.
         * @param options The options used for the registration.
         */
        this.register = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_register).bind(this);
        /**
         * Redirects to the Account Management Console.
         */
        this.accountManagement = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_accountManagement).bind(this);
        /**
         * Returns the URL to login form.
         * @param options Supports same options as Keycloak#login.
         *
         * NOTE oidc-spa: Not supported, please use login() method.
         */
        //createLoginUrl(options?: KeycloakLoginOptions): Promise<string>;
        /**
         * Returns the URL to logout the user.
         * @param options Logout options.
         *
         * NOTE oidc-spa: Not supported, please use logout() method.
         */
        //createLogoutUrl(options?: KeycloakLogoutOptions): string;
        /**
         * Returns the URL to registration page.
         * @param options The options used for creating the registration URL.
         *
         * NOTE oidc-spa: Not supported please user login({ action: "register" })
         */
        //createRegisterUrl(options?: KeycloakRegisterOptions): Promise<string>;
        /**
         * Returns the URL to the Account Management Console.
         * @param options The options used for creating the account URL.
         */
        this.createAccountUrl = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_createAccountUrl).bind(this);
        /**
         * Returns true if the token has less than `minValidity` seconds left before
         * it expires.
         * @param minValidity If not specified, `0` is used.
         */
        this.isTokenExpired = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_isTokenExpired).bind(this);
        /**
         * If the token expires within `minValidity` seconds, the token is refreshed.
         * If the session status iframe is enabled, the session status is also
         * checked.
         * @param minValidity If not specified, `5` is used.
         * @returns A promise to set functions that can be invoked if the token is
         *          still valid, or if the token is no longer valid.
         * @example
         * ```js
         * keycloak.updateToken(5).then(function(refreshed) {
         *   if (refreshed) {
         *     alert('Token was successfully refreshed');
         *   } else {
         *     alert('Token is still valid');
         *   }
         * }).catch(function() {
         *   alert('Failed to refresh the token, or the session has expired');
         * });
         */
        this.updateToken = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_updateToken).bind(this);
        /**
         * Clears authentication state, including tokens. This can be useful if
         * the application has detected the session was expired, for example if
         * updating token fails. Invoking this results in Keycloak#onAuthLogout
         * callback listener being invoked.
         *
         * NOTE oidc-spa: In this implementation we never end up in the kind of
         * state where calling this makes sense.
         * oidc-spa take more control and exposes less complexity to the user of the
         * adapter.
         */
        //clearToken(): void;
        /**
         * Returns true if the token has the given realm role.
         * @param role A realm role name.
         */
        this.hasRealmRole = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_hasRealmRole).bind(this);
        /**
         * Returns true if the token has the given role for the resource.
         * @param role A role name.
         * @param resource If not specified, `clientId` is used.
         */
        this.hasResourceRole = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_hasResourceRole).bind(this);
        /**
         * Loads the user's profile.
         * @returns A promise to set functions to be invoked on success or error.
         */
        this.loadUserProfile = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_loadUserProfile).bind(this);
        /**
         * @private Undocumented.
         */
        this.loadUserInfo = __classPrivateFieldGet(this, _Keycloak_instances, "m", _Keycloak_loadUserInfo).bind(this);
        const issuerUri = `${params.url.replace(/\/$/, "")}/realms/${params.realm}`;
        __classPrivateFieldSet(this, _Keycloak_state, {
            constructorParams: params,
            dInitialized: new Deferred(),
            initOptions: undefined,
            oidc: undefined,
            tokens: undefined,
            keycloakUtils: createKeycloakUtils({ issuerUri }),
            issuerUri,
            profile: undefined,
            userInfo: undefined,
            $onTokenExpired: createStatefulEvt(() => undefined)
        }, "f");
    }
    /**
     * Is true if the user is authenticated, false otherwise.
     */
    get authenticated() {
        if (!this.didInitialize) {
            return false;
        }
        const { oidc } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        return oidc.isUserLoggedIn;
    }
    /**
     * The user id.
     */
    get subject() {
        if (!this.didInitialize) {
            return undefined;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.subject when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        return tokens.decodedIdToken_original.sub;
    }
    /**
     * The realm roles associated with the token.
     */
    get realmAccess() {
        if (!this.didInitialize) {
            return undefined;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.realAccess when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        assert(is(tokens.decodedIdToken_original));
        return tokens.decodedIdToken_original.realm_access;
    }
    /**
     * The resource roles associated with the token.
     */
    get resourceAccess() {
        if (!this.didInitialize) {
            return undefined;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.resourceAccess when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        assert(is(tokens.decodedIdToken_original));
        return tokens.decodedIdToken_original.resource_access;
    }
    /**
     * The base64 encoded token that can be sent in the Authorization header in
     * requests to services.
     */
    get token() {
        if (!this.didInitialize) {
            return __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions?.token;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.token when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        return tokens.accessToken;
    }
    /**
     * The parsed token as a JavaScript object.
     */
    get tokenParsed() {
        if (!this.didInitialize) {
            const { token } = __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions ?? {};
            if (token === undefined) {
                return undefined;
            }
            return decodeJwt(token);
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.tokenParsed when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        return decodeJwt(tokens.accessToken);
    }
    /**
     * The base64 encoded refresh token that can be used to retrieve a new token.
     */
    get refreshToken() {
        if (!this.didInitialize) {
            return __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions?.refreshToken;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.refreshToken when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        return tokens.refreshToken;
    }
    /**
     * The parsed refresh token as a JavaScript object.
     */
    get refreshTokenParsed() {
        if (!this.didInitialize) {
            const { refreshToken } = __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions ?? {};
            if (refreshToken === undefined) {
                return undefined;
            }
            return decodeJwt(refreshToken);
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.refreshTokenParsed when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        if (tokens.refreshToken === undefined) {
            return undefined;
        }
        return decodeJwt(tokens.refreshToken);
    }
    /**
     * The base64 encoded ID token.
     */
    get idToken() {
        if (!this.didInitialize) {
            return __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions?.idToken;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.idToken when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        return tokens.idToken;
    }
    /**
     * The parsed id token as a JavaScript object.
     */
    get idTokenParsed() {
        if (!this.didInitialize) {
            const { idToken } = __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions ?? {};
            if (idToken === undefined) {
                return undefined;
            }
            return decodeJwt(idToken);
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.idTokenParsed when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        assert(is(tokens.decodedIdToken_original));
        return tokens.decodedIdToken_original;
    }
    /**
     * The estimated time difference between the browser time and the Keycloak
     * server in seconds. This value is just an estimation, but is accurate
     * enough when determining if a token is expired or not.
     */
    get timeSkew() {
        if (!this.didInitialize) {
            const { timeSkew } = __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions ?? {};
            if (timeSkew === undefined) {
                return null;
            }
            return timeSkew;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.timeSkew when keycloak.authenticated is false is a logical error in your application");
            return null;
        }
        assert(tokens !== undefined);
        return Math.ceil((tokens.getServerDateNow() - Date.now()) / 1000);
    }
    /**
     * Whether the instance has been initialized by calling `.init()`.
     */
    get didInitialize() {
        return __classPrivateFieldGet(this, _Keycloak_state, "f").oidc !== undefined;
    }
    /**
     * @private Undocumented.
     */
    get loginRequired() {
        const { initOptions } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        if (initOptions === undefined) {
            return false;
        }
        return initOptions.onLoad === "login-required";
    }
    /**
     * @private Undocumented.
     */
    get authServerUrl() {
        const { keycloakUtils: { issuerUriParsed } } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        return `${issuerUriParsed.origin}${issuerUriParsed.kcHttpRelativePath}`;
    }
    /**
     * @private Undocumented.
     */
    get realm() {
        const { keycloakUtils: { issuerUriParsed } } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        return issuerUriParsed.realm;
    }
    /**
     * @private Undocumented.
     */
    get clientId() {
        const { constructorParams } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        return constructorParams.clientId;
    }
    /**
     * @private Undocumented.
     */
    get redirectUri() {
        const { initOptions } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        if (initOptions === undefined) {
            return undefined;
        }
        return initOptions.redirectUri;
    }
    /**
     * @private Undocumented.
     */
    get sessionId() {
        if (!this.didInitialize) {
            return undefined;
        }
        const { oidc, tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        if (!oidc.isUserLoggedIn) {
            console.warn("Trying to read keycloak.sessionId when keycloak.authenticated is false is a logical error in your application");
            return undefined;
        }
        assert(tokens !== undefined);
        const { sid } = tokens.decodedIdToken_original;
        assert(typeof sid === "string");
        return sid;
    }
    /**
     * @private Undocumented.
     */
    get profile() {
        const { profile } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        return profile;
    }
    /**
     * @private Undocumented.
     */
    get userInfo() {
        const { userInfo } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        return userInfo;
    }
    /**
     * Called if there was an error while trying to refresh the token.
     *
     * NOTE oidc-spa: In oidc-spa an auth refresh error always triggers a page refresh.
     */
    //onAuthRefreshError?(): void;
    /**
     * Called if the user is logged out (will only be called if the session
     * status iframe is enabled, or in Cordova mode).
     *
     * NOTE oidc-spa: In oidc-spa a logout always triggers a page refresh.
     */
    //onAuthLogout?(): void;
    /**
     * Called when the access token is expired. If a refresh token is available
     * the token can be refreshed with Keycloak#updateToken, or in cases where
     * it's not (ie. with implicit flow) you can redirect to login screen to
     * obtain a new access token.
     */
    set onTokenExpired(value) {
        const { $onTokenExpired } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        $onTokenExpired.current = value;
    }
    get onTokenExpired() {
        const { $onTokenExpired } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        return $onTokenExpired.current;
    }
    /**
     * This property does not exist in the original keycloak-js module.
     * It let you access the underlying oidc-spa/core instance that power this adapter under the hood.
     * It can be useful for example to create an auto logout warning overlay.
     *
     * See: https://docs.oidc-spa.dev/v/v10/resources/migrating-from-keycloak-js
     */
    get oidc() {
        assert(this.didInitialize, "Cannot get keycloak.oidc before the init() method was called and have resolved.");
        const { oidc } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        return oidc;
    }
}
_Keycloak_state = new WeakMap(), _Keycloak_instances = new WeakSet(), _Keycloak_init = async function _Keycloak_init(initOptions = {}) {
    const { onLoad = "check-sso", redirectUri, enableLogging, scope, locale, autoLogoutParams } = initOptions;
    if (__classPrivateFieldGet(this, _Keycloak_state, "f").initOptions !== undefined) {
        if (JSON.stringify(__classPrivateFieldGet(this, _Keycloak_state, "f").initOptions) !== JSON.stringify(initOptions)) {
            throw new Error("Can't call init() multiple time with different params");
        }
        await __classPrivateFieldGet(this, _Keycloak_state, "f").dInitialized.pr;
        const { oidc } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(oidc !== undefined);
        return oidc.isUserLoggedIn;
    }
    __classPrivateFieldGet(this, _Keycloak_state, "f").initOptions = initOptions;
    const { constructorParams, issuerUri } = __classPrivateFieldGet(this, _Keycloak_state, "f");
    const autoLogin = onLoad === "login-required";
    let hasCreateResolved = false;
    const oidcOrError = await createOidc({
        BASE_URL: constructorParams.BASE_URL,
        sessionRestorationMethod: constructorParams.sessionRestorationMethod,
        issuerUri,
        clientId: __classPrivateFieldGet(this, _Keycloak_state, "f").constructorParams.clientId,
        autoLogin,
        postLoginRedirectUrl: redirectUri,
        debugLogs: enableLogging,
        scopes: scope?.split(" "),
        autoLogoutParams,
        extraQueryParams: !autoLogin || locale === undefined
            ? undefined
            : () => {
                if (hasCreateResolved) {
                    return {};
                }
                return {
                    ui_locales: locale
                };
            }
    })
        // NOTE: This can only happen when autoLogin is true, otherwise the error
        // is in oidc.initializationError
        .catch((error) => error);
    hasCreateResolved = true;
    if (oidcOrError instanceof OidcInitializationError) {
        this.onAuthError?.({
            error: oidcOrError.name,
            error_description: oidcOrError.message
        });
        await new Promise(() => { });
        assert(false);
    }
    const oidc = oidcOrError;
    if (oidc.isUserLoggedIn) {
        const tokens = await oidc.getTokens();
        const onNewToken = (tokens_new) => {
            __classPrivateFieldGet(this, _Keycloak_state, "f").tokens = tokens_new;
            this.onAuthRefreshSuccess?.();
        };
        onNewToken(tokens);
        oidc.subscribeToTokensChange(onNewToken);
    }
    __classPrivateFieldGet(this, _Keycloak_state, "f").oidc = oidc;
    __classPrivateFieldGet(this, _Keycloak_state, "f").dInitialized.resolve();
    this.onReady?.(oidc.isUserLoggedIn);
    onAuthSuccess_call: {
        if (!oidc.isUserLoggedIn) {
            break onAuthSuccess_call;
        }
        this.onAuthSuccess?.();
    }
    onAuthError_call: {
        if (oidc.isUserLoggedIn) {
            break onAuthError_call;
        }
        if (oidc.initializationError === undefined) {
            break onAuthError_call;
        }
        this.onAuthError?.({
            error: oidc.initializationError.name,
            error_description: oidc.initializationError.message
        });
    }
    onActionUpdate_call: {
        if (!oidc.isUserLoggedIn) {
            break onActionUpdate_call;
        }
        if (this.onActionUpdate === undefined) {
            break onActionUpdate_call;
        }
        const { backFromAuthServer } = oidc;
        if (backFromAuthServer === undefined) {
            break onActionUpdate_call;
        }
        const status = backFromAuthServer.result.kc_action_status;
        if (!isAmong(["success", "cancelled", "error"], status)) {
            break onActionUpdate_call;
        }
        const action = backFromAuthServer.extraQueryParams.kc_action;
        if (action === undefined) {
            break onActionUpdate_call;
        }
        this.onActionUpdate(status, action);
    }
    schedule_onTokenExpired_call: {
        if (!oidc.isUserLoggedIn) {
            break schedule_onTokenExpired_call;
        }
        const { $onTokenExpired } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        let clear = undefined;
        const next = (onTokenExpired) => {
            clear?.();
            if (onTokenExpired === undefined) {
                return;
            }
            let timer = undefined;
            const onNewToken = () => {
                if (timer !== undefined) {
                    clearTimeout(timer);
                }
                const { tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
                assert(tokens !== undefined);
                timer = setTimeout(() => {
                    onTokenExpired.call(this);
                }, Math.max(tokens.accessTokenExpirationTime - tokens.getServerDateNow() - 3000, 0));
            };
            onNewToken();
            const { unsubscribeFromTokensChange } = oidc.subscribeToTokensChange(onNewToken);
            clear = () => {
                if (timer !== undefined) {
                    clearTimeout(timer);
                }
                unsubscribeFromTokensChange();
            };
        };
        next($onTokenExpired.current);
        $onTokenExpired.subscribe(next);
    }
    return oidc.isUserLoggedIn;
}, _Keycloak_login = async function _Keycloak_login(options) {
    const { redirectUri, action, loginHint, acr, acrValues, idpHint, locale, doesCurrentHrefRequiresAuth } = options ?? {};
    if (!this.didInitialize) {
        await __classPrivateFieldGet(this, _Keycloak_state, "f").dInitialized.pr;
    }
    const { oidc, keycloakUtils } = __classPrivateFieldGet(this, _Keycloak_state, "f");
    assert(oidc !== undefined);
    const extraQueryParams_commons = {
        claims: acr === undefined
            ? undefined
            : JSON.stringify({
                id_token: {
                    acr
                }
            }),
        acr_values: acrValues,
        ui_locales: locale
    };
    if (oidc.isUserLoggedIn) {
        assert(action !== "register");
        assert(loginHint === undefined);
        assert(idpHint === undefined);
        assert(doesCurrentHrefRequiresAuth === undefined);
        await oidc.goToAuthServer({
            redirectUrl: redirectUri,
            extraQueryParams: {
                ...extraQueryParams_commons,
                kc_action: action,
                ui_locales: locale
            }
        });
        assert(false);
    }
    assert(action === undefined || action === "register");
    await oidc.login({
        redirectUrl: redirectUri,
        doesCurrentHrefRequiresAuth: doesCurrentHrefRequiresAuth ?? false,
        extraQueryParams: {
            ...extraQueryParams_commons,
            login_hint: loginHint,
            kc_idp_hint: idpHint
        },
        transformUrlBeforeRedirect: action !== "register" ? undefined : keycloakUtils.transformUrlBeforeRedirectForRegister
    });
    assert(false);
}, _Keycloak_logout = async function _Keycloak_logout(options) {
    if (!this.didInitialize) {
        await __classPrivateFieldGet(this, _Keycloak_state, "f").dInitialized.pr;
    }
    const { oidc, initOptions } = __classPrivateFieldGet(this, _Keycloak_state, "f");
    assert(oidc !== undefined);
    assert(initOptions !== undefined);
    assert(oidc.isUserLoggedIn, "The user is not currently logged in");
    const redirectUri = options?.redirectUri ?? initOptions.redirectUri;
    await oidc.logout({
        ...(redirectUri === undefined
            ? { redirectTo: "current page" }
            : { redirectTo: "specific url", url: redirectUri })
    });
    assert(false);
}, _Keycloak_register = async function _Keycloak_register(options) {
    return this.login({
        ...options,
        action: "register"
    });
}, _Keycloak_accountManagement = async function _Keycloak_accountManagement(options) {
    const { redirectUri, locale } = options ?? {};
    window.location.href = this.createAccountUrl({
        redirectUri,
        locale
    });
    return new Promise(() => { });
}, _Keycloak_createAccountUrl = function _Keycloak_createAccountUrl(options) {
    const { locale, redirectUri } = options ?? {};
    const { keycloakUtils, constructorParams } = __classPrivateFieldGet(this, _Keycloak_state, "f");
    return keycloakUtils.getAccountUrl({
        clientId: this.clientId,
        validRedirectUri: (() => {
            if (redirectUri !== undefined) {
                return redirectUri;
            }
            const { homeUrlAndRedirectUri } = getHomeAndRedirectUri({
                BASE_URL_params: constructorParams.BASE_URL
            });
            return homeUrlAndRedirectUri;
        })(),
        locale
    });
}, _Keycloak_isTokenExpired = function _Keycloak_isTokenExpired(minValidity = 0) {
    let accessTokenExpirationTime;
    if (!this.didInitialize) {
        const fakeAccessToken = this.token;
        if (fakeAccessToken === undefined) {
            throw new Error("isTokenExpired was called too early");
        }
        const time = readExpirationTimeInJwt(fakeAccessToken);
        assert(time !== undefined, "The initial token is not a JWT");
        accessTokenExpirationTime = time;
    }
    else {
        const { tokens } = __classPrivateFieldGet(this, _Keycloak_state, "f");
        assert(tokens !== undefined);
        accessTokenExpirationTime = tokens.accessTokenExpirationTime;
    }
    if (accessTokenExpirationTime > Date.now() + minValidity * 1000) {
        return false;
    }
    return true;
}, _Keycloak_updateToken = async function _Keycloak_updateToken(minValidity = 5) {
    if (!this.didInitialize) {
        await __classPrivateFieldGet(this, _Keycloak_state, "f").dInitialized.pr;
    }
    const { oidc } = __classPrivateFieldGet(this, _Keycloak_state, "f");
    assert(oidc !== undefined);
    assert(oidc.isUserLoggedIn, "updateToken called too early");
    if (!this.isTokenExpired(minValidity)) {
        return false;
    }
    await oidc.renewTokens();
    return true;
}, _Keycloak_hasRealmRole = function _Keycloak_hasRealmRole(role) {
    const access = this.realmAccess;
    return access !== undefined && access.roles.indexOf(role) >= 0;
}, _Keycloak_hasResourceRole = function _Keycloak_hasResourceRole(role, resource) {
    if (this.resourceAccess === undefined) {
        return false;
    }
    const access = this.resourceAccess[resource || this.clientId];
    return access !== undefined && access.roles.indexOf(role) >= 0;
}, _Keycloak_loadUserProfile = async function _Keycloak_loadUserProfile() {
    if (!this.didInitialize) {
        await __classPrivateFieldGet(this, _Keycloak_state, "f").dInitialized.pr;
    }
    const { oidc, keycloakUtils } = __classPrivateFieldGet(this, _Keycloak_state, "f");
    assert(oidc !== undefined);
    assert(oidc.isUserLoggedIn, "Can't load userProfile if user not authenticated");
    const { accessToken } = await oidc.getTokens();
    return (__classPrivateFieldGet(this, _Keycloak_state, "f").profile = await keycloakUtils.fetchUserProfile({ accessToken }));
}, _Keycloak_loadUserInfo = async function _Keycloak_loadUserInfo() {
    if (!this.didInitialize) {
        await __classPrivateFieldGet(this, _Keycloak_state, "f").dInitialized.pr;
    }
    const { oidc, keycloakUtils } = __classPrivateFieldGet(this, _Keycloak_state, "f");
    assert(oidc !== undefined);
    assert(oidc.isUserLoggedIn, "Can't load userInfo if user not authenticated");
    const { accessToken } = await oidc.getTokens();
    return (__classPrivateFieldGet(this, _Keycloak_state, "f").userInfo = await keycloakUtils.fetchUserInfo({ accessToken }));
};
