// oidc-spa/keycloak/keycloakUtils.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.createKeycloakUtils = createKeycloakUtils;
const toFullyQualifiedUrl_1 = require("../tools/toFullyQualifiedUrl");
const keycloakIssuerUriParsed_1 = require("./keycloakIssuerUriParsed");
function createKeycloakUtils(params) {
    const { issuerUri } = params;
    const issuerUriParsed = (0, keycloakIssuerUriParsed_1.parseKeycloakIssuerUri)({ issuerUri });
    const keycloakServerUrl = `${issuerUriParsed.origin}${issuerUriParsed.kcHttpRelativePath ?? ""}`;
    const getAdminConsoleUrl = (realm) => `${keycloakServerUrl}/admin/${encodeURIComponent(realm)}/console`;
    const realmUrl = `${keycloakServerUrl}/realms/${encodeURIComponent(issuerUriParsed.realm)}`;
    return {
        issuerUriParsed,
        adminConsoleUrl: getAdminConsoleUrl(issuerUriParsed.realm),
        adminConsoleUrl_master: getAdminConsoleUrl("master"),
        getAccountUrl: ({ clientId, locale, validRedirectUri }) => {
            const accountUrlObj = new URL(`${keycloakServerUrl}/realms/${issuerUriParsed.realm}/account`);
            accountUrlObj.searchParams.set("referrer", clientId);
            accountUrlObj.searchParams.set("referrer_uri", (() => {
                try {
                    return (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                        urlish: validRedirectUri,
                        doAssertNoQueryParams: true,
                        doOutputWithTrailingSlash: true
                    });
                }
                catch {
                    return (0, toFullyQualifiedUrl_1.toFullyQualifiedUrl)({
                        urlish: validRedirectUri,
                        doAssertNoQueryParams: false
                    });
                }
            })());
            if (locale !== undefined) {
                accountUrlObj.searchParams.set("kc_locale", locale);
            }
            return accountUrlObj.href;
        },
        fetchUserProfile: ({ accessToken }) => fetch(`${realmUrl}/account`, {
            headers: {
                Accept: "application/json",
                Authorization: `Bearer ${accessToken}`
            }
        }).then(r => r.json()),
        fetchUserInfo: ({ accessToken }) => fetch(`${realmUrl}/protocol/openid-connect/userinfo`, {
            headers: {
                Accept: "application/json",
                Authorization: `Bearer ${accessToken}`
            }
        }).then(r => r.json()),
        transformUrlBeforeRedirectForRegister: authorizationUrl => {
            const urlObj = new URL(authorizationUrl);
            urlObj.pathname = urlObj.pathname.replace(/\/auth$/, "/registrations");
            return urlObj.href;
        }
    };
}
